{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "18f1577d",
   "metadata": {},
   "source": [
    "# Differential Expression Analysis\n",
    "Salmon was used to obtain counts and eggnog-mapper was used to obtain functional annotation. \n",
    "\n",
    "We first read in packages from Bioconductor to import the data and to use the DESeq2 package. Bioconductor requires a specific version of R to run (likely the most recent) I had troubles initially but once I created a conda environemnt with the most recent R version, Bioconductor worked fine. \n",
    "\n",
    "#note about loading and installing tidyverse -it was working but stoped working, I tried installing it to conda which took forever! Conda found some conflicts and searched all packages in my environment-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "47cde0fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘gplots’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    space\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    space\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "\n",
      "Loading required package: registry\n",
      "\n",
      "Loading required package: rngtools\n",
      "\n",
      "Loading required package: cluster\n",
      "\n",
      "NMF - BioConductor layer [OK] | Shared memory capabilities [NO: bigmemory] | Cores 2/2\n",
      "\n",
      "  To enable shared memory capabilities, try: install.extras('\n",
      "NMF\n",
      "')\n",
      "\n",
      "\n",
      "Attaching package: ‘NMF’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    nrun\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#if (!require(\"BiocManager\", quietly = TRUE))\n",
    "  #install.packages(\"BiocManager\")\n",
    "#BiocManager::install(c(\"tximport\",\"DESeq2\"))        #I think I have already installed this an do not need to every time?\n",
    "#BiocManager::install(c(\"apeglm\"))\n",
    "library('tximport')\n",
    "library('DESeq2')\n",
    "#install.packages(c(\"dplyr\",\"tidyr\",\"ggplot2\",\"gplots\",\"NMF\",'RColorBrewer','pheatmap','hexbin'))\n",
    "library('tidyr')\n",
    "library('gplots')\n",
    "library('RColorBrewer')\n",
    "library('NMF')\n",
    "library('ggplot2')\n",
    "library(\"pheatmap\")\n",
    "library('hexbin')\n",
    "library('dplyr')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b71cb480",
   "metadata": {},
   "source": [
    "# 1. Using the tximport function, read in the files.  \n",
    "\n",
    "The tximport will create a list of matricies. We use the column names of the counts matrix to create rownames for the samples matrix. \n",
    "\n",
    "### Locate directory with files and name it as dir. \n",
    "Create a vector called files for all salmon quantification files. Create, or read in, a csv file with sample names referencing the files in the files vector. We will create the names of each item in vector files using the samples csv. Read in the files using file.path to feed in the dir variable before each file name in the vector files.  \n",
    "\n",
    " Note: you can only read in files associated with the same reference transcriptome together.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ee1d7557",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dir <-  \"/work/nclab/lucy/SAB/Assembly/04/salmon\"\n",
    "#list.files(dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2f7dcb4",
   "metadata": {},
   "source": [
    "### First, we construct a df of metadata for the transcriptome. \n",
    "This includes column 'id' with sample id's I have used in my personal notes, column 'isolate' with the isolate number (4,6,8, or 13), column 'treatment' with treatment for each, and column 'replicate'. Basically, the 2-3 columns are info from column 1 broken up. \n",
    "\n",
    "### Then, using file.path, point R to the location of all the salmon quant files for this transcriptome\n",
    "This will create a list of all the file location and names. Using the id column from sample metadata, name each file in the list. This name will better represent to us what we are looking at later. \n",
    "\n",
    "### Finally, check that all files do exist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1a90cb74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "counts_metadata <- data.frame(\n",
    "    id=c(\"4add_backA\",\"4add_backB\",\"4pFe21_9A\",\"4pFe21_9B\",\"4pFe21_9C\"),\n",
    "    isolate=rep_len(4,5),\n",
    "    treatment=c(rep_len(\"add_back\",2),rep_len(\"pFe21_9\",3)),\n",
    "    replicate=c('A','B','A','B','C'))\n",
    "\n",
    "\n",
    "\n",
    "files <- file.path(dir, list.files(dir, pattern=\".sf\",recursive = TRUE))\n",
    "\n",
    "names(files) <- paste0(\n",
    "    counts_metadata$id)\n",
    "\n",
    "#files\n",
    "\n",
    "all(file.exists(files))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cca29a4d",
   "metadata": {},
   "source": [
    "### Now we can read in our list of files using tximport. \n",
    "Ensure the rownames for counts_metadata is equal to the column names of the raw_counts file you just created and check out the top few rows of the data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3224d406",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "raw_counts <- tximport(files, type='salmon',txIn=TRUE, txOut = TRUE) #should this be true? transcript level data?\n",
    "\n",
    "rownames(counts_metadata) <- colnames(raw_counts$counts)\n",
    "\n",
    "all(rownames(counts_metadata)==colnames(raw_counts))\n",
    "\n",
    "#raw_counts[,rownames(counts_metadata)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126e2042",
   "metadata": {},
   "source": [
    "if I sum up all counts from the open reading frames from Salmon output, there should be less than th enumber of reads I began my assembly with because about 45% of reads were mapped for quantification. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f3fa7f45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>4add_backA</dt><dd>1283802.01</dd><dt>4add_backB</dt><dd>2350598.013</dd><dt>4pFe21_9A</dt><dd>1928237.983</dd><dt>4pFe21_9B</dt><dd>2310138.004</dd><dt>4pFe21_9C</dt><dd>1748057.992</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[4add\\textbackslash{}\\_backA] 1283802.01\n",
       "\\item[4add\\textbackslash{}\\_backB] 2350598.013\n",
       "\\item[4pFe21\\textbackslash{}\\_9A] 1928237.983\n",
       "\\item[4pFe21\\textbackslash{}\\_9B] 2310138.004\n",
       "\\item[4pFe21\\textbackslash{}\\_9C] 1748057.992\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "4add_backA\n",
       ":   1283802.014add_backB\n",
       ":   2350598.0134pFe21_9A\n",
       ":   1928237.9834pFe21_9B\n",
       ":   2310138.0044pFe21_9C\n",
       ":   1748057.992\n",
       "\n"
      ],
      "text/plain": [
       "4add_backA 4add_backB  4pFe21_9A  4pFe21_9B  4pFe21_9C \n",
       "   1283802    2350598    1928238    2310138    1748058 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>4add_backA</dt><dd>999999.999911</dd><dt>4add_backB</dt><dd>999999.999931</dd><dt>4pFe21_9A</dt><dd>1000000.000057</dd><dt>4pFe21_9B</dt><dd>999999.999975</dd><dt>4pFe21_9C</dt><dd>1000000.000007</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[4add\\textbackslash{}\\_backA] 999999.999911\n",
       "\\item[4add\\textbackslash{}\\_backB] 999999.999931\n",
       "\\item[4pFe21\\textbackslash{}\\_9A] 1000000.000057\n",
       "\\item[4pFe21\\textbackslash{}\\_9B] 999999.999975\n",
       "\\item[4pFe21\\textbackslash{}\\_9C] 1000000.000007\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "4add_backA\n",
       ":   999999.9999114add_backB\n",
       ":   999999.9999314pFe21_9A\n",
       ":   1000000.0000574pFe21_9B\n",
       ":   999999.9999754pFe21_9C\n",
       ":   1000000.000007\n",
       "\n"
      ],
      "text/plain": [
       "4add_backA 4add_backB  4pFe21_9A  4pFe21_9B  4pFe21_9C \n",
       "     1e+06      1e+06      1e+06      1e+06      1e+06 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "my_counts <- raw_counts$counts \n",
    "myTPM <-  raw_counts$abundance\n",
    "\n",
    "colSums(my_counts) #total number of reads mapping to my reference (the trranslated salmon pojbect) per same\n",
    "colSums(myTPM)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "536b724e",
   "metadata": {},
   "source": [
    "----------\n",
    "# 2. Create DESeq data frame\n",
    "\n",
    "We will: \n",
    "1. calculate summary stats on TPM matrix  \n",
    "2. generate a deseq data set\n",
    "2. filter by counts \n",
    "3. calculated DE and results\n",
    "4. log fold shrinkage of effect size (for visualization)\n",
    "3. pull out 100 most differentially expressed genes\n",
    "4. merge the matrix of count data with the eggnog functional annotations\n",
    "6. normalize the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4b4e89c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate summary stats on the my_TPM data matrix \n",
    "myTPM_stats  <-  transform(myTPM,\n",
    "                          SD=rowSds(myTPM),\n",
    "                          AVG=rowMeans(myTPM),\n",
    "                          MED=rowMedians(myTPM))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5882f268",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzdeXxU9b3/8e85M5NMFhMmCVup7EtQdhHBpQQEBK8somhEUBFbcCuUQm/B\nBVTQ23JRqVwUF1oBNYiyWNSgVHKlVJRVsJewBSmL4gJkJ5lkzu+P8+t0moRk5sw5M5PveT0f\nefCYOd+zfL9zJpk3n7OMommaAAAAQOOnRrsDAAAAMAfBDgAAQBIEOwAAAEkQ7AAAACRBsAMA\nAJAEwQ4AAEASBDsAAABJEOwAAAAkIXOwO3/+vBKcefPmCSEyMzMDJ7rd7rZt295111179uwJ\nXK1/tocffrjO7U6ZMkWfoVu3bhEYptWqqqr0VyP2txWZrmZkZCiKUlJSEuPrtLk63wxRfJ3/\n/Oc/X3PNNSkpKfofh6+++srqLS5fvrx///4pKSkpKSkDBgz405/+1OAiDzzwgN69//qv/7K6\newAs4ox2ByzkdDqvuuqqwCnff/99QUFBYmJi9+7dA6f/9Kc/9T/u3r17s2bNhBDnzp3Lz89f\nuXLlW2+9tXLlyuzs7Brrf/PNNxctWhQXFxc48cKFC6tXrzZxFMnJyaWlpeXl5ZGJVrAOu9K2\n9u3bd8sttwghRowY0bJlSyFEWlqadZvTNO2ee+5ZsWKFy+W69tprhRB//etfJ02atHXr1tde\ne+1iS33yyScvvfSSoih8HRHQqMkc7JKTk7dv3x445U9/+tOkSZM6depUY3qgxx9//NZbb9Uf\n//DDD3fdddeHH344ZcqUYcOGBf4t7tu3786dO9977z3/zLq1a9cWFhZeeeWVO3bsMHU0iBUL\nFiwoLy+Pj4+P8XWitmi9zhs2bPB6vXPmzFmwYEEENrdixYoVK1Y0b948Ly8vMzNTCHHgwIGs\nrKzly5dff/3148ePr71ISUnJ5MmTO3To0L59+48++igCnQRgEZkPxYYvIyNjxYoV8fHxRUVF\nubm5gU333HOPEKL20Q19yt133x2ZHiLypkyZMn36dJfLFePrRG3Rep1PnDghhGjXrl1kNvfM\nM88IIX73u9/pqU4I0bVr16efftrfVNvMmTOPHz/+6quvJiQkRKaTACxCsGtARkZGp06dhBAF\nBQWB07t06TJgwIDc3Nxvv/3WP/HEiRN/+ctffvazn3Xo0CHI9R8/fvzhhx/u2rVrcnLyJZdc\n0rZt25EjR7755ptCCP2wSGlpqRAiISHBf/Kff4tbt26dPn16nz59mjZtGhcX16pVq9tvv33n\nzp01NqEoitPpFEK8/fbbAwYMSE5OTklJGTZsWJ1lyz179owePTotLS0pKalPnz6vvvpqnd0O\nddOrVq0aMGCAfoLRhQsXQtrWxQS/eHFx8dNPP33FFVekpKQkJCRcfvnl8+bNKy4u9s+Qn5+v\nKEqzZs28Xm+NZaurq1u0aKEoyt///nd9Su3ztBp8NRrclXWe+3Xs2LEpU6a0a9cuPj7e4/EM\nGjRIf2MECmnnBgppyPW8S+vh79sf//jHK664IikpqXnz5pMmTfruu++EEJWVlQsWLMjMzExI\nSGjVqtWsWbMqKipqrKHBHecX5JvBwL6rMZZQX+d58+YpivLKK68IIX7+85/r+33ChAn+GULa\ny3X+HtWQn59/8ODBuLi42267LXD67bff7nQ6v/rqq6NHj9ZYZPPmzcuWLZs6derAgQPrHw6A\nRkCzkz/+8Y9CiJ49e9bZ2qVLFyHEmjVrakzX/5/9u9/9LnC2jz/++OWXXxZC/P73v/fPOX/+\nfCHEH//4xw8//FAIcfnll9ffn/z8/NTUVCFEZmbmuHHjsrOzr7322uTk5IEDB2qatmPHjrlz\n5+rVhUceeWTuPxUXF+uLX3HFFQ6Ho1u3biNGjBg1alTnzp2FEC6Xa/369YFbEUI4HI7HH39c\nUZSuXbsOHz780ksvFULEx8fv2rUrcM6PPvpIP0rVs2fPiRMnDho0SFXVX/7yl/rMgXOGtOn/\n/M//FEJ06dJl0KBBP/3pT0tLS0PaVp2CX/zYsWN6NE9PT7/++utvvPFG/RzKbt26/fjjj/7Z\n9NMxa/Rf07SNGzcKIfr27eufkp6eLoTw74VgXo0Gd2XtdX766acpKSlCiLZt2956661ZWVn6\nR/vEiRN9Pl+NVziYnVtbkEOu/11aD//eV1W1R48eQ4cO1U9muPzyy4uLi7Oystxu91VXXTVw\n4ED9pMPs7OzAxYPccVoobwYD+y7M13nLli1z587t3bu3EGLkyJH6fn/33Xf11pD2cp2/R7Xp\nubBPnz61m/Rzi2v8iSsqKmrTps2ll15aVFSkadro0aOFEM8880w9gwIQywh2/1JnsNu3b5+q\nqkKI999/P3C2jz/+uLCwUK8i+Gfu2LFjUlJScXFxkMHu/vvvF0LMnj07cGJZWdlnn33mf5qU\nlCSEKC8vr73422+/ffr06cApb775pqqqzZo1C5xfT/Aej+fjjz/Wp1RWVur/mx81apR/tqKi\noubNmwshnn32Wf/ETz75RP/QrfEBGdKmk5OTN23apE/RP65C2lZtwS9eXV3dp08fIcT9999f\nUlKiTywpKbn99tuFEBMmTPDP+eKLLwohbr755hrbGjdunBBiyZIl/im1w0GQr0Y9u7LGOktK\nSlq0aCGE+PWvf11VVaVP3LVrlz7bSy+95F8wyJ1bpyCHHMy7tE563zIyMrZu3apP+f777/WD\ng926devevfvRo0f16V9++aV+BHD//v36lOB3XEjvJcP7LpzXWdO0yZMnCyFeeeWVwImh7uXa\nv0d10k/jGzNmTO2mG2+8UQixcOHCwIk///nPA/++EeyAxo5g9y81gt25c+f+/Oc/6wdV27dv\nX1lZGTib/vf9zjvvFEJ88cUXmqZ9+umnQoh77rlH07Qgg93YsWOFEFu2bKlnnnrSQJ30i+/8\nHwDaPz8VXnjhhcDZ9MMxKSkp/k8IvQAZWJrS6Xd1CaaKdrFNz507t8acYW4r+MXXrl0rhOjX\nr191dXXgnMXFxU2bNnU4HP7az7lz59xut8vl+v777/2znT17Nj4+Pi4uLrBEVDsc1Kn2qxF8\nsNMH2LFjR//nvW7x4sX6dP+UIHdunYIccjDv0jrpfXvxxRcDJy5ZskSfvmPHjsDpd911lxBi\n8eLF+tPgd1xI7yXD+y6c11m7SLALdS/X/j2q0+zZs4UQd955Z+0mPYk+/vjj/in6dRITJ070\nTyHYAY0d59jVNG7cOP08GI/HM3LkyKNHj7Zu3Xr9+vV1nnAdeAmF/q8+JUj9+vUTQkybNm3j\nxo3l5eUGeltZWfnJJ58sWbJk/vz58+bNmzdvnn4C08GDB2vMqf+99mvfvn1CQkJRUZH/fKP/\n/d//FULoUTXQxIkTw9x07avwQt2W4cU/+OADIcTYsWP1sqtfcnLyVVddVV1dvWvXLn1KkyZN\nxowZ4/V6A89wysnJqaioGDVqVIM3pwj+1Qh+gBMnTnQ4HIHT7733XiHEkSNHTp06FTi9wZ1b\npyCHHOa79Kabbgp8qh/rbN68ed++fQOn6/9fOn36tP40+B0X5ntJmPpLFJJQ93KdV7PWpmma\nEEJRlAbnLCoqmjx5cvPmzZ9//vnguw0gxsl8uxNj/Pexi4+Pb9GixXXXXXfbbbclJibWOfP1\n11/funXrt95666mnnlqzZk379u1/9rOfBb+t6dOnb9myZdOmTSNHjnS5XN26dRs0aNDEiRN7\n9eoVzOJr1qx56KGH9A+hGoqKigKfqqoaeK8+XUpKSnl5eUVFxSWXXCKEOHnypKjrwr06L+UL\nftNCiDZt2tSYEtK2agt+cf2Sl9/+9re//e1v61zVjz/+6H98zz335OTkvP766/rpWUKI119/\nXQQR1kN6NYKhf6K3b9++xvTk5OTmzZufOXPm1KlTrVq10icGs3MvJpghh/MuVVXV30+d3p/W\nrVvX7rAQwn9BQPA7Lsz3krm/RCEJaS+Lun6P6qT3pM6sqU/0d3XGjBknTpxYs2aNpTfVAxBh\nBLuaAu9j1yBFUe6+++6nnnrq7rvvLi4unjVrVjD/UfaLj4/Pzc39/PPPP/jgg23btm3fvn3P\nnj3PPvvsY4899uSTT9a/7K5du7Kzs+Pj45csWTJs2LBWrVrpl1vOmTNHP4xSo59BdiyY2ULa\ntMPhuNhtw0J6rYwtXl1dLYS49dZbL7/88jpnuOyyy/yPhw4d+tOf/nT37t379+/v3r37wYMH\nP//88xYtWgwfPryeTYT0agSpnqJL7RUGv3NrC2bI4bxLL9a3GnW42kLaccLoe8miX6IghbSX\n6/k9qqFt27ZCiH/84x+1m/S7rugzCCHWr1/vdDqXLFniPz4uhNC/EuPll1/Ozc3t3r37Cy+8\nEMxGAcQOgl247rnnnvnz52/cuFEPeQbWcNVVV+nXJ1ZWVr7zzjv33nvv/Pnzb7/99ot9pOlW\nrlzp8/keffTRBx98MHD64cOHDfRBCKHXBo4dO1Zj+tdff236poPfVpiL65cu6rfJaHC1qqpO\nnDjxmWee+dOf/rRo0SL92PqECRNqHCmrwfQdIf75VSi1b0tRWlqq15ZqlMEMC37Ixt6lhgW/\n48J5L1mx74Jn0V7WLzr56quvysvLA29KV1JScuDAASFEYKm1qqpKPyJcw7Fjx44dO1ZVVWWg\nAwCii3PswtW+ffsRI0akp6ePHDmy9gGmkMTFxY0fP37o0KGapu3bt88/UQhR+y+s/qe/xtGZ\n77//fvPmzca2rt/CqvY9tFatWmX6poPfVpiLjxgxQgjx1ltvBXlymH4I8o033qisrNTX1uBx\n2OBfjYvtytr0Aa5atUovXPnpwatjx45mBTsR+pDrfJeaLvgdF857yfRfopBYtJczMzM7d+5c\nWVn59ttvB05fvXp1VVVVt27dOnbsqE/54Ycfap92HXjxxF//+ldDIwMQTQQ7E7z//vs//PDD\nhg0bQl3wxRdfrFEb+Pbbb3fv3i0CDpfo/6333yrWT7/F1/Lly/23dS0pKZk0adL58+dDHoAQ\nQojs7OymTZt+8cUXf/jDH/wTP/30U/3aPXM3Hfy2wlz8tttu6969+6FDh+68887AW0kLIY4c\nOfLcc8/VmL9z585XX331mTNnZs2adfLkyb59+zZYkQr+1bjYrqxt/PjxLVq0OHz48COPPOLz\n+fSJ+/fvf+KJJ4QQM2fObHANwWtwyMG8S00X/I4L571k+i9RSKzby/qFsb/97W/913/k5+fP\nmTNHnxhuvwHEuMhcfBsjjN2guM7Z/LezqlOQtzvp2bOnEKJTp05jx469++67R4wYoV+lEXin\n1scff1wI4fF4xo0bN3ny5MmTJxcWFmqadvr06aZNmwohWrVqddttt91yyy3p6ektWrSYNGmS\nEOKpp57yr0EI4XA4am9dv/tX4K0uPvzwQ72q1KtXr4kTJw4ePFhV1dq3jQh/08Fv62KCX/zY\nsWNdu3YVQiQmJg4YMOD2228fOnSo/9rM2msODASBt6/zq3HLjOBfjYvtytrr1DTt008/1U9y\n79ixY3Z29tChQ/Xrsuu8dW3tTtbeufWof8jBvEvrVGffPvvsMyHEVVddVWO6fi7XtGnT/FOC\n33HBvxkM77swX+c6b3eihb2XL8bn8+kXBbtcrsGDBw8ePFhf7aRJkxpcltudAI0dwe5fIh/s\n/vznP0+dOrVXr1761xm1bt162LBha9asCbx3V0VFxezZszt16qR/dAkhvvnmG73pxIkTd999\nd5s2beLj41u3bv2LX/zi9OnTc+fODeczaefOnTfddFOTJk0SEhJ69uz54osv6l85VeMDMsxN\nh7St8BcvKytbvHjxdddd5/F4XC5Xy5Yt+/btO3PmzG3bttVerX7faSFEjdvX+dUOYUG+GvXs\nyjrvr3b06NGf//znbdq0cblcqampAwcOXLVqVY1bppkS7OofcjDv0jqFGey0UHZckG8Gw/vO\nomCnhbeX6+Hz+V599dV+/folJyfr94hZvnx5MAsS7IDGTtEMXbUHAACAWMM5dgAAAJIg2AEA\nAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJg\nBwAAIAmCHQAAgCQIdgAAAJJwRrsD1iotLa2qqjJ3nYmJiUKIsrIyc1fbKCQmJtpz4ElJSdXV\n1RcuXIh2RyJNVdW4uDh7DjwhIaGqqqqioiLafYk0h8PhdDrtOXC32+31eisrK6Pdl0hzOp2q\nqhobeGpqqun9QTgkD3ZVVVVer9fcdTocDk3TTF9to6Cqqg0HriiK0+m05073f9RFuyORpoeb\n6upqG45dCOFyuew5cKfTWVlZacOx2/lzTT4cigUAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbAD\nAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAE\nwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJOGMdgcQHZlZ2f7H\n+Xk5UewJAAAwCxU7OwpMdbWfAgCARoqKnb1cLMPp0yndAQDQqFGxs5EGK3OU7gAAaNSo2NlC\n8ImN0h0AAI0XFTv5GajDUboDAKAxIthJznBEI9sBANDoEOwAAAAkQbADAACQBMEOAABAEgQ7\nAAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJ\nEOwkl5+XE+EFAQBAtDij3QFYTo9omVnZIc0PAAAaHSp2dhFkXCPVAQDQeFGxs5H6S3dEOgAA\nGjsqdrZTZ4Aj1QEAIAEqdnYUWLoj0gEAIA2CnX0R6QAAkAyHYgEAACRBsAMAAJAEwQ4AAEAS\nBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAA\nAEkQ7AAAACRBsAMAAJAEwQ4AAEASiqZp0e6Dhbxer6qaHF4dDocQorq62tzVNgoOh8O2A9c0\nzefzRbsjUaCqqj0HbtudriiKoij2HLiqquz0UOmfiYgdzmh3wFplZWVer9fcdaalpWmadu7c\nOXNX2yh4PB4bDlxRlPT09KqqqsLCwmj3JdIcDkdycrI9B+7xeCorK4uLi6Pdl0hzuVxut9ue\nA09NTS0vLy8rK4t2XyLN7Xarqmps4BkZGab3B+HgUCwAAIAkCHYAAACSINgBAABIgmAHAAAg\nCYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0A\nAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQI\ndgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAA\nkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgB\nAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiC\nYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJgBwAA\nIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYId\nAACAJAh2AAAAknBGZjP5+fmzZ8/WNG39+vX+iTt37ly5cuXJkydTU1OHDBlyxx13KIoSThMA\nAICdRaJiV1RUtHDhwt69ewdOPHjw4Pz58y+77LJnn312woQJa9eufeONN8JpAgAAsDnLK3aa\npi1atGjIkCFut3v37t3+6WvXrm3VqtWUKVOEEG3atPnmm282bNgwbty4+Ph4Y01WDwQAACDG\nWV6xy8nJqaqqys7OrjH9wIEDffr08T/t06fPhQsXCgoKDDcBAADYnLUVuy+//DI3N/f555+v\ncRqcpmnnz5/3eDz+Kfrjs2fPGmvyT9m+ffszzzzjf/rEE090797d3EGpqurftN04HA57DlwI\n4XQ6bTh2RVEURbHhwHVxcXE2HLttd7r+OZWQkGDDQ0D6TrfhwKVkYcXu3LlzixYtmjZtmg3/\nQAAAAESehRW7Y8eOnT9//sknn9SfapqmadqYMWNuu+228ePHN2nS5Ny5c/6Z9cdpaWmKohho\n8k/p37//hg0b/E8LCwsD5zdFWlqapmmmr7ZR8Hg8Nhy4oijp6elVVVWFhYXR7kukORyO5ORk\new7c4/FUVlYWFxdHuy+R5nK53G63PQeemppaXl5eVlYW7b5EmtvtVlXV2MAzMjJM7w/CYWHF\n7rLLLnvhhRcW/9Po0aNVVV28ePF//Md/CCG6du0aeC3F7t273W53+/btDTfFiMysmmcTAgAA\nRIaFwc7tdrcJoB+QbdOmTWpqqhBi7Nixp06dWrZs2fHjx7ds2bJu3bpRo0bpB/iNNUVdZla2\nnur8DwAAACIpQjcorq1Lly6PPPLIqlWrNm3alJqaevPNN48fPz6cpuiqneQys7Lz83Ki0hkA\nAGBPiqZp0e6DhQoLC71er7nrrH2OXf31OZninZ3PsfN6vfY81czO59hVVFTY81QzO59jV1ZW\nxjl2IeEcu1jDd8WGJZijrhyWBQAAkUGwMy74xMZZdwAAIAIIdkY073mDgaBGtgMAAJYi2AEA\nAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJg\nBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYGfEmS83GVgqPy/H9J4AAAD4OaPdgcZKT2mZWdnB\nzwwAAGApKnZhCSaxkeoAAEBkULELVz2lOyIdAACIJCp25qid4Uh1AAAgwqjYmcZfuiPSAQCA\nqKBiZzJSHQAAiBaCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABI\ngmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcA\nACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmC\nHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACA\nJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYA\nAACSINgBAABIQtE0Ldp9sJDX61VVk8Orw+EQQlRXV5u72kbB4XDYduCapvl8vmh3JApUVbXn\nwG270xVFURTFngNXVZWdHir9MxGxwxntDlirrKzM6/Wau860tDRN086dO2fuahsFj8djw4Er\nipKenl5VVVVYWBjtvkSaw+FITk6258A9Hk9lZWVxcXG0+xJpLpfL7Xbbc+Cpqanl5eVlZWXR\n7kukud1uVVWNDTwjI8P0/iAcHIoFAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBDgAA\nQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7\nAAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJ\nEOwAAAAkQbADAACQBMEOAABAEs5od8AuMrOy/Y/z83Ki2BMAACArKnaREJjqaj8FAAAwBcHO\ncnXGuMysbOIdAAAwF8HOQg2mN7IdAAAwEcHOKkGGNkp3AADALAQ78xnIamQ7AAAQPoJdrCDb\nAQCAMBHsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJ\nEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDszJeflxOxpQAAAPwIdpYIKaXl5+WQ6gAAQPgI\ndlYJMqsR6QAAgFmc0e6AzPTQlpmVXU8rAACAWajYWa7OAEeqAwAApqNiFwmBpTsiHQAAsAjB\nLnKIdAAAwFIcigUAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDs\nAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASTkvXvnXr1vfee+/U\nqVMVFRXp6enXXXdddna2y+XSW3fu3Lly5cqTJ0+mpqYOGTLkjjvuUBQlnCYAAAA7szbYORyO\nIUOG/OQnP4mLizty5Mjrr79eVFT04IMPCiEOHjw4f/78ESNGzJgx4+jRo0uXLvX5fBMmTDDc\nBAAAYHPWBrurr77a/7hLly7Hjx/ft2+f/nTt2rWtWrWaMmWKEKJNmzbffPPNhg0bxo0bFx8f\nb6zJ0oEAAADEPmuDnZ/P5/v666/37t3bu3dvfcqBAwcGDhzon6FPnz6rV68uKCjo2rWrsSZ9\nSlVVVVlZWeB2rThQqyiKbY//2nDg/iHbduy2Hbiw8djtPHB7jt2eA5eS5cHO6/WOGzdO0zRN\n04YNG/aLX/xCCKFp2vnz5z0ej382/fHZs2eNNfmnbN26ddasWf6nS5cu7devnxXjSk9Pt2K1\nsc+2A3e5XLYdu20HHh8fb9ujAbYdeEJCQkJCQrR7ER22HbhkLA92Tqdz8eLFXq/38OHDq1at\nSklJueuuu6zbXFpaWmCSS0xM9Hq95m7C5XJpmlZVVWXuahsFp9Npz4Gz06Pdi0hTFMXpdPp8\nvurq6mj3JdIURVFV1Z4Dt+1OV1VVCOHz+Qws678gEjHC8mCnKEqbNm2EEB07dlRVdenSpWPH\njk1OTm7SpMm5c+f8s+mP09LSFEUx0OSf0rNnz6VLl/qfFhYWFhYWmjuitLQ0TX//Es4AACAA\nSURBVNNMX22j4PF4bDhwRVHS09OrqqpsOHaHw5GcnGzPgXs8Hq/XW1xcHO2+RJrL5XK73fYc\neGpq6oULFwLP57EJt9utqqqxgWdkZJjeH4Qjovexq6qq8pc9unbtunv3bn/T7t273W53+/bt\nDTcBAADYnLXB7uWXX87Lyztw4MDf//73devWrVixom/fvk2aNBFCjB079tSpU8uWLTt+/PiW\nLVvWrVs3atQo/awOY00AAAA2p2iaZt3aV6xY8fnnn3/33XeqqjZr1mzgwIEjR47057AdO3as\nWrXqxIkT+q2Gx48f778kx1hTbYWFhaafY6cfig08ImwfHo/HhgPXD8V6vV57HpG086HYiooK\nex6RtPOh2LKyMg7FhoRDsbHG2mAXdQQ7cxHsot2XSCPY2TPfEOyi3ZdII9jJhO+KBQAAkATB\nDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAKAOhw4dmjdv3r59+6LdkRAQ7AAAAOpw6NCh\nJ554onEFO8u/K9ZWMrOy9Qf5eTnR7QkAALAhKnam8ae6Go8BAECoqqqqnn/++SuuuCIpKemS\nSy7p0aPH3Llz/a3nz5//9a9/3a5du/j4+ObNm995551Hjhzxtz7//POKouzcuTNwhWPGjElO\nTvY/feeddxRFeffdd3/3u9917tw5Pj6+devWCxYs8H9xw7x580aOHCmEmDhxoqIoiqJkZWVZ\nOmRTULEzR+0kp0+hdAcAQKiqqqpuuummTZs2DRw48PHHH09JScnPz1+zZs0TTzwhhCgtLf3Z\nz362f//+O++88+qrrz58+PCLL7744YcffvbZZ126dAlpQ7/5zW86d+78hz/8oUmTJq+++uqj\njz6anp4+depUIcQ999wTHx8/Z86cOXPmDB06VAihf9l9jCPYhav+4lxmVjbZDgCAkCxZsmTT\npk0PP/zw4sWL/d8I7/P59AeLFi3av3//ggUL5syZo08ZMWLEDTfcMG3atNzc3JA2lJaW9sEH\nH+ib6Nev36effvqHP/xBD3Zt27bt3r27EKJr166Nolan41BsWII55JqZlc2RWQAAgrdq1aqE\nhISnn37an+qEEKr6/0PLu+++m5ycPGPGDH/TsGHDBgwY8PHHHxcVFYW0If0wq3/9ffv2PXr0\nqD9BNkYEO4NCjWtkOwAAgnTo0KGOHTsGnhIXqKCgoEOHDm63O3Bi9+7dfT7f119/HdKGLr30\n0sCnKSkplZWVxcXFIfY3hhDsjGje8wYDS5HtAAAIhqZpgbW6kFqFEHW2VlVVBTmn//qJxohg\nBwAAYkuXLl0OHz5cUlJSZ2uHDh2OHDly4cKFwIlfffWVqqpt27YVQqSlpQkhzp49GzhD4GWz\nQao/PsYmgh0AAIgtEyZMKC8vf+yxxwIn+gtpY8eOLSkpef755/1Nmzdv/tvf/jZkyJCUlBQh\nhH5tbOCFFGvXrj148GCo3bjkkktErYAY47gqFgAAxJYHH3xw48aNzz///N69e0eMGJGSknL4\n8OFNmzZ99dVXQoiZM2e+8847s2fP/vvf/+6/3YnH41m8eLG+eL9+/fr377948eLi4uJu3brt\n27dvw4YN3bt3LygoCKkbPXv2dLvdL7zwQlxcXJMmTZo1azZ48GDzR2sqKnYAACC2uFyuDz/8\n8Pe///2PP/44d+7c3/zmN5s3bx43bpzempSUtHXr1l/96ldbt26dNm3aqlWrbr755i+++CIz\nM9O/hrfffnvUqFE5OTmPPPLIiRMntmzZ0r59+1C7kZqa+uabbyYlJU2fPv2OO+548sknTRuh\nZZRGfYZggwoLC71er7nrTEtLa9ZjmLFlG/s97Twez7lz56Ldi0hTFCU9Pd3r9RYWFka7L5Hm\ncDiSk5PtOXCPx1NRUdGoL44zxuVyud1uew48NTW1rKysrKws2n2JNLfbraqqsYFnZGSY3h+E\ng4odAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYBdRmVnZ0e6C\n5ewwRgAAYhPBzogzX24yfKthuXOPPrrMrGy5hwkAQGziu2KNy8/LMRZfMrOyG/tXUNRW+6WQ\ncpgAAEuVV4mj532mr9bjVlolK6avNgYR7MKiBxcD8U5fRJrcc7FXQLJhAgCs9m2p76ltFaav\ndnAb55ReLtNXG4M4FGsCOx+WDeaoqwTDBABEiCY0zWf6j9DMrwLGJip25rBn6S748TbqYQIA\nIkrTot2DRoyKnZnsU7ozdnlEoxsmACDCNKFZJNojixCCncnsk+0AALCEfuTU3B9hl2DHoVgA\nABBDrKiu2adiR7ADAAAxQ7PmHDu75DqCHQAAiCGaZskVrFwVCwAAEHlU7MJAsAMAADHEknPs\nbJPsCHYAACB2aJbcTJiLJwAAACJNs+YKVoIdAABAhGlCWHHxBIdiAQAAooGLJ8JAsAMAADGE\nGxSHg2AHAABihzUXT9imZEewAwAAMYOLJ8JDsDNTZla24WXz83LC2ajhxY3Jz8sxMNgIdxIA\n0CjZJoRZgWBnGsOpLsxIF/g4kslJ31bwoybVAQCCYNF97PhKMQQtioW62hNjsHRHpAMABM+a\niydMX2WMItiFK7qFuou1xk7pjlQHAAiNJSnMLsmOYGdc7BTq6pwt6qU7Ih0AIGRWXTxh/ipj\nE8HOoFgr1NU5f+SznYjSxRwAAFlYdLsTzrHDxTXveYOxBSOW6vxLRT5gEekAAOHgHLtwEOwA\nAECssOq7YrkqFgAAIArsU16zAMEOAADEDE2zpLpmm7BIsAMAALGE252EgWAHAABiCd8VGwY1\n2h0AAAD4F80SQW169erVAwYMSEtLc7vdHTt2fOSRRyoqKvyt77//fq9evdxu96WXXjp37lyf\nz2ddk2FU7AAAQOyw6D52QSU7p9M5adKkzp07u93unTt3zp49+4cffli2bJkQYvv27aNHj546\nderKlSv37NkzderU6urq+fPnW9EUDoIdAACIGVZ980RQ67zlllv8j/v37//VV1998skn+tOF\nCxd27tx5yZIlQoju3bsfPnz4ueeemzNnTmJioulN4QyUQ7EAACCWaJr5PyFePFFdXb1nz56P\nP/74+uuv16ds27Zt+PDh/hmGDx9eWlq6Z88eK5rCQcUOAADECpdDPHJj+8ApXxw7/9H//RDS\nSjKS46YObB04pbIq2MO7FRUVSUlJPp9P07T77rvvhRdeEEL4fL4zZ860aNHCP5v++PTp06Y3\nhTTS2gh2AAAgVnirtfkbD4e5kh+KK2qsZES3ZkO6ZgSzbFxc3N69ey9cuLBz585HH320adOm\nTz/99MVmVhQlYk1BItgBAIBYoVn0XbFBH4pVFKVbt25CiL59+zocjilTpsyaNcvj8TRv3vzb\nb7/1z6Y/btmypaqq5jaFM0whfbBTVdXpNHmMzXoMM7ys4c4c+es7Ha+9NZJbrJOiKKa/no2F\nPceuqqptBy7sutMdDodtBy6s+dSIER2vvfXIX9+ps0lV1RgauGbN97oaCosVFRWaplVWVgoh\nrrnmmtzc3GeffVZvys3NTUpK6t27txVN4YiNvWiZuLg4/Q+0WVr2HmFswW/2fBjmpr/Z82FI\nWw9/i7UpipKUlGT6ahsFh8Nhw7ErimLbgQshnE6nPceuqqo9By6EiIuL0xOeZPSPD71AUPvT\nQf8vXMwMXIviDYp/+ctf9uvXr127dj6f7/PPP3/yySdvvPHG5s2bCyFmzZp17bXXPvTQQ1Om\nTNm7d++iRYtmzJihX8FqelM4lJAKntXV1TGz44NSWFjo9XpNWVVmVrbhZfPzckzpQ/A9MXeL\nfh6P59y5c1asOZYpipKenu71egsLC6Pdl0hzOBzJycn2HLjH46moqCguLo52XyLN5XK53W57\nDjw1NbWsrKysrCzafTFTnR8ZNT4j3G63qqrGBp6REdSJa8E7fKbk/hXhXhla2409Wsy4oVOD\ns82ePXvDhg3Hjx93OBxt27a94447pk2b5g9bGzdufPTRRw8cONC0adN777133rx5/uKR6U2G\nNRzszp49u3jx4o0bNx48eLC0tDQpKalLly4jR46cNm2ax+MJc/NWMyvYGU51FgWsevpj0RZ1\nBLto9yXSCHb2zDcEu2j3xTT1f3j5Py9iLdhNfX23uesUQtzYo8Wvh3c2fbUxqIFDsV9++eUN\nN9xw5swZIcQll1zSqlWroqKi3bt37969+5VXXsnNze3evXtE+hlNsZbq/Guu3TFLUx0AoLEI\n5pMrMys7Rj81bPO9rlaor+JXXl5+yy23fP/99zNmzDhy5EhRUdHJkyeLiooOHTo0ffr0b775\n5tZbbw38DjX5ZGZlG0t1+Xk5EfhtqbGJGP39BABEVvCfXIY/5ixk0TfF2iYs1lexW7169dGj\nR//nf/7ngQceCJzeqVOn5557rl27dtOmTVuzZs2ECRMs7mR0xGChLrrbAgDEOGOfXG37j/nH\nF++Z3hnjYuaq2Maovorde++917Zt26lTp9bZ+tBDD7Vu3XrDhg3WdCzKYu5/MAAA2INFJbto\nDytC6qvY7du37/rrr7/YBRqqqg4ZMuTTTz+1pmMAAMCWLLndifmrjE31BbszZ860adOmnhla\nt2793Xffmd0lAABgU5rQtKh+80RjV1+wKy0tTUhIqGeGpKQkG14SDwAArKJxjl1Y6gt2wURm\nK2I1AACwLUuihW3iSgP3sVuzZk1+fv7FWvfv3292fwAAgJ1pllTsOBSr++KLL7744ovIdAUA\nAMCSc+yo2AkhduzYEbF+AAAACGGjw6ZWqC/Y9e3bN2L9AAAAEEJoFhyKpWIHAAAQcZpF97Gz\nS7Cr75sn8vLyjh8/Xs8Mu3fvfumll8zuEgAAsC1rvnjCNhdP1BfsBg0atHLlSv1xQUGB2+1e\nt25d4AwffPDB/fffb2HvAACA3WiaBT/RHlSkBHso1ufzVVRUVFdXW9obAABgZ5pF59jZJtnV\nV7Gzs/y8nPy8HGPLZmZlG1hE/zG2RQAAhBDGPrm+3r7e9J6ER7PmxxYIdvWJTLYLnJl4BwAI\nR0iFiXCqGJax5hw7Lp6AzvCbPpiIdrF5yHYAgHAE88kVe5FOCPHPq2I5x84ogl1QrCjd1Z/e\nyHYAgHDUU5iIyUJdACuCnW2SXQMXT6xfv/7rr78WQhQVFQkhXnrppdzcXH/r3r17rexbbNF/\nB4ydPyf+PRoGuZLaCwIAEJL8vJwaHzqx/7Gi+bhBsXENBLtdu3bt2rXL//Qvf/mLxf2JdbV/\nQ4KUmZVtLBr6FwQAwIDAT5/G8IGi8c0T4eC7YkP23b6PmvUYZmBBw0dXyXYAgDA1qs8RK0IY\nwY7vigUAABHGV4qFh++KBQAAMcQ+h02tUN9VsReCE7G+AgAA2Vlxr5Ng72O3fPnyIUOGNGvW\nLDk5uXfv3q+99lpg6/vvv9+rVy+3233ppZfOnTvXF3CRh+lNhtVXsUtISAhmFSRrAABgCqu+\nUiy4rLJixYrrrrvuV7/6VWpq6rvvvnvfffd5vd6pU6cKIbZv3z569OipU6euXLlyz549U6dO\nra6unj9/vhVN4VDqGaqiKG63u3///g6Ho55VbN68OcxOWKewsNDr9Zq7zrS0NGMXT4QjRk56\n9Xg8586di3YvIk1RlPT0dK/XW1hYGO2+RJrD4UhOTrbnwD0eT0VFRXFxcbT7Emkul8vtdttz\n4KmpqWVlZWVlZdHuS6S53W5VVY0NPCMjw9zO5J/88a6FG8xdpxBi9IDOj2RfG+pSgwYNUlVV\nvyXILbfccuDAgf/7v//Tmx577LHnnnvuu+++S0xMNL0pnJHWdyi2Q4cOFy5cOHTo0FVXXbVs\n2bLNFxHO5gEAAP5Fs+Q7xYxdFHvhwoVmzZrpj7dt2zZ8+HB/0/Dhw0tLS/fs2WNFUzjqC3aH\nDx/+5JNPBg0a9Nxzz3Xq1Gnw4MFvvPFGeXl5mJsEAACok6KIVunJgT+piXGhnk7nVJUaK0mM\nD/lq0eXLl+/atWvGjBlCCJ/Pd+bMmRYtWvhb9cenT582vcnAixaovnEqijJo0KBBgwadP3/+\nzTfffO211yZMmNCkSZPx48dPnjy5T58+YW4bAAAgkENVru/VLnDKgX98/8XBUyGtJDE+rsZK\nMlJCO765evXqBx988PXXX7/yyivrmU1RlIg1BSmoANukSZMHHnjggQce2Lt372uvvfbGG28s\nXbp04cKFM2fODHPzAAAAflXVvtc/CvdwZGFJeY2VjLk6M/jFX3rppRkzZrz11ltjxozRp6iq\n2rx582+//dY/j/64ZcuWpjeFMtA61HcotraOHTv26tWrQ4cOQoiSkpIwtw0AAPDvNE3zmf8T\n9El2Tz755KxZs9577z1/qtNdc801ubm5/qe5ublJSUm9e/e2oikcwQa7bdu23XvvvS1atLjv\nvvtcLterr75KuQ4AAJhM/+YJ83+C2vj06dOfeuqphQsXZmRk7N27d+/evQcOHNCbZs2adejQ\noYceemj//v0rV65ctGjR9OnT9StYTW8KR323OxFCfPvttytWrFi+fPnBgwebNWs2ceLEyZMn\nd+3aNcytRgy3OzEXtzuJdl8ijdud2POuH9zuJNp9ibTYut3JP74f//Tb5q5TCHHztZc9NmFQ\ng7NlZGT8+OOPgVM6dOhw5MgR/fHGjRsfffTRAwcONG3a9N577503b56qqhY1GVZfsBs9evQH\nH3ygadqwYcMmT548atQol8sV5vYizKJgp2la8543mLva+hHsoohgZ8+BE+yi3ZFII9jFULBb\nsNrcdQo92E0cbPpqY1B9F0+89957brd7zJgxrVq1+uyzzz777LM6Z/vv//5va/oW0/LzcjKz\nsiOzoQhsBQCAWKBZ841W9vmSrAauir1w4UJOTgPBwp7BTvwzclka70h1AACb0YQw/yvFhLE7\nFDdC9QW7HTt2RKwfjZdFpTsiHQDApiwprxHshOjbt2/E+tGomV66I9UBAGxKs+RQrH2OxYb8\nDRu4GFNKd0Q6AIDd2SaEWSHcq2oRKMxYRqoDANieZpFojytCCHYmI5wBABAWzWfBj12CHYdi\nAQBADLHkdidcPAEAABBpmiY0C253QsUOAAAg8qw5H45gBwAAEHmW3O7E/FXGJoIdAACIFZZ9\npZhdkh3BDgAAxBLbhDArEOwAAEDMsOiec7YJiwQ7AAAQUyy4KtY2J9kR7AAAQCyhYhcGgh0A\nAIgh1tyg2C4IdgAAIJZQsQsDwQ4AAMQOSy6e4CvFYERmVraxBfPzcsztCQAAjZLGDYrDQrAz\nDakOAIDwaXxXbBgIdiYg0gEAYBq+KzYMBLtwkeoAADCLJjQrKnZ8pRgaRqQDAMB8VOzCQLAz\niFQHAID5NGvuY2eXXEewM6R5zxuMLUiqAwCgIVTsjFOj3QEAAAA/TWjW/ATh888/Hzt2bLt2\n7RRFue+++2q0vv/++7169XK73ZdeeuncuXN9Pp91TYYR7AAAQAzRLBDkpktLSzt27LhgwYKO\nHTvWaNq+ffvo0aOvvfbaHTt2LFiwYOHChY8//rhFTeHgUCwAAIgl0ftKscGDBw8ePFgI8fvf\n/75G08KFCzt37rxkyRIhRPfu3Q8fPvzcc8/NmTMnMTHR9KZwBmptxW7z5s2PPfbYxIkTb7vt\ntunTp3/88ceBrTt37pw2bdott9xy7733vvnmm4GB2lgTAABo7DTNZ/5P2OfYbdu2bfjw4f6n\nw4cPLy0t3bNnjxVN4bC2YvfJJ59cfvnlo0ePTkxM/Nvf/vbCCy9UVVWNGDFCCHHw4MH58+eP\nGDFixowZR48eXbp0qc/nmzBhguEmAADQ2Dkd6pRbrg+csv/IP/6291BIK2lySdLtNwwInNI0\nLSWcXvl8vjNnzrRo0cI/RX98+vRp05vC6aewOtg9/fTT/seXXXbZsWPHtm3bpge7tWvXtmrV\nasqUKUKINm3afPPNNxs2bBg3blx8fLyxJksHAgAAIqDa53v/r/9WtSouLQ+13lZcXl5jJUP7\ndzehc7UoihKxpiBF9By7ysrKZs2a6Y8PHDgwcOBAf1OfPn1Wr15dUFDQtWtXY036lOLi4pMn\nT/pnSE9Pj4uLs3ZUoXA6G/dJjYqiNPYhGGbPsauqatuBC7vudIfDYduBCyFUVbXh2FVVjZ2B\naz7txOnvwlxJVaWvxkpKy8rDWaGqqs2bN//222/9U/THLVu2NL0pnH6KSAa7zZs3Hzly5Be/\n+IUQQtO08+fPezwef6v++OzZs8aa/FN27tw5a9Ys/9OlS5f269fPwlGFqEmTJtHuQrgkGIIx\nTqfTtmO37cDj4uJi6n+GkWTbgbvdbrfbHe1eREeMDFyz5uu/wl/lNddck5ub++yzz+pPc3Nz\nk5KSevfubUVTOCIU7LZu3frSSy/96le/6tSpk6UbatWq1dixY/1PPR7PhQsXzN1EOId9Te9M\nhMXHx1dUVES7F1Hgdrt9Pl9lZWW0OxJpiqK4XC57Djw+Pr66utrr9Ua7L5GmqqrD4bDnwOPi\n4qqqqqqqqqLdl0jTy7TGBm5BHNSieIPi8vLygwcP6g/Onj27d+9eRVF69uwphJg1a9a11177\n0EMPTZkyZe/evYsWLZoxY4Z+BavpTeGIRLD78MMPX3vttZkzZ/bv31+foihKkyZNzp07559H\nf5yWlmasyT+lc+fOc+bM8T8tLCwsKSkxdzjh/EfW9M5EmMvlauxDMEBRFLfbXV1dbcOxOxyO\n5ORkew48Pj6+qqrKhmN3uVxut9ueA4+Li6usrCwrK4t2XyLN7Xarqmps4JbU+aJ3v4uDBw/6\na2aHDh1at26dw+HQI2///v3Xr1//6KOPvvLKK02bNp05c+a8efP0OU1vCoflwS4nJ2ft2rWP\nPfaYHnj9unbtunv37smTJ+tPd+/e7Xa727dvb7gJAABIwJpDsUGts1evXvXMedNNN910002R\naTLM2vvYvfLKK6tXr540adIll1xSUFBQUFBw4sQJvWns2LGnTp1atmzZ8ePHt2zZsm7dulGj\nRulHOY01AQCARk8TUfxKMQlYW7HLy8urrq5+8cUX/VNatGjx8ssvCyG6dOnyyCOPrFq1atOm\nTampqTfffPP48eP1eYw1RUyzHsOMLZifl2NuTwAAkA9fPRAOa4PdG2+8UU/rlVdeeeWVV5rY\nFAGZWdnGFiTVAQAQBGuqa7YJizFx05pGgUgHAEAEaMJnwToJdghAqgMAIEIsqa4R7CCEINIB\nABBhlhyKNX+VsYlgVx9SHQAAEaVpllw8wTl2Nmc40glSHQAA4dDMP8fOPiU7gh0AAIgVln1X\nLMEOAAAg8jjHLgwEOwAAEDssOceO250AAABEBbc7MY5gBwAAYoZmzflwdsl1BDsAABBTuEFx\nGAh2AAAgllCxCwPBDgAAxA5rLp7gdicAAABRwA2Kw0CwAwAAMcSaW5MQ7AAAACJM4xy7sBDs\nAABA7OAGxWFRo92BGJWflxPtLgAAYEuaZsmPPRDsLspwtsvMyjawiP5jbIsAAMjDkmAX7UFF\nCsGuPuFku+BTWuCcxDsAgJ1pQmiaz/wf2yQ7gl0D8vNyrCvdXSzGke0AADalcSg2LAS7oFhR\nuqs/vVG6AwDYkkUHYgl2+Hcmlu6CD21kOwCA7VCuCwPBLjThl+5CzWqU7gAA9sKh2DAQ7EL2\n3b6PDC9LRAMAoH6aJaI9qkjhBsUAACCm8JVixhHsAABA7LCovEawAwAAiKxLkhKv69fL9NV2\natfa9HXGJoIdAACIFe1at1r32sJo96IR4+IJAAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAA\nAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsjzny5KcJbzMzKjvAWAQBAo0OwMyg/LyfCW8zMyibe\nAQCAehDsjMvPy4lKvIvwFgEAQGNBsAsXpTsAABAjCHYmoHQHAABiAcHONJTuAABAdBHszETp\nDgAARBHBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQ\nBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATBzmSZWdnGFszPyzG3JwAAwG4IdmYKM9UZyHb5\neTkkQgAAoHNGuwOSMBzpxL/nOf1xkGsj0gEAgEAEOxOYfvg1Py+n/nUS6QAAQG0Eu7CYVai7\nWGud6yfVAQCAOhHsjIvAdRI1SndEOgAAUA+CnRHNe95geNlQw5m/dEeqAwAA9eOq2IgyHM5I\ndQAAoEEEOwAAAElIfijW6XSqqsnhVVEUw8vGx8eb2JPIUxSlsQ/BMFVVbTh2VVVtO3Bh153u\ncDhsO3AhhNPptOHY9c9KGw5cSpIHO/1jKdq9+Bens3G/4IqiNPYhGGbPsSuKYtuBCyFUVbXh\n2PU/m/YcuGCno/GTfC9WVlZ6vV5z1xnO/2lKS0tN7EnkxcXFNfYhGKAoSkJCQnV1tQ3H7nA4\nHA6HPQfudrurqqpsOHaXy+V2u+058Pj4+MrKyrKysmj3JdLcbreqqsYGnpCQYHp/EI4YqmYB\nAAAgHAQ7AAAASRDsAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwS5kzXoM\nM7xsZla2iT0BAAAIJPk3T5jLlFimryQ/Lyf8VQEAAASiYhcsc4ttlO4AAIDpCHZBsSKHZWZl\nE+8AAICJCHYNsDp+ke0AAIBZCHb1iUzqonQHAABMQbCrW+TDFtkOAACEiWAHAAAgCYIdAACA\nJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYA\nAACSINgBAABIgmBXt/y8HOm3CAAAJEOwu6hIJi1SHQAACJ8z2h2IaXreyszKtnoTAAAA4aNi\n1zDrshepDgAAmIiKXVBML90R6QAAgOmo2IXArDRGqgMAAFYg2IUmPy/n9Tl/KAAAFopJREFU\nu30fhbkGszoDAAAQiGAXUaQ6AABgHYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQI\ndgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYJdyJr1GGZ42cys\nbGNLGVsQAADYijPaHWhMTElX+kry83JC3WioCwIAALuhYhcsc2tmwaytzkIdpTsAAHAxVOwa\nZlGWqr8CV89GKd0BAIA6UbFrgNUVsjprckHW86zpEQAAaKyo2F1UxJJTYAUupI1SugMAAIGo\n2NUt8vUww5e+UroDAAA6gh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDY\nAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJp6VrP3To0Lvvvnv06NHvvvtu\n6NChDz/8cGDrzp07V65cefLkydTU1CFDhtxxxx2KooTTZKL8vJzMrGzTV2vFFvPzckzvDAAA\naIysrdhduHChZcuWEydObNmyZY2mgwcPzp8//7LLLnv22WcnTJiwdu3aN954I5wm00UyMOnb\nys/LCXWjpDoAAOBnbcWuR48ePXr0EEKsXbu2RtPatWtbtWo1ZcoUIUSbNm2++eabDRs2jBs3\nLj4+3liTFf3XY5OlpbvaySzI0h2RDgAA1GBtsKvHgQMHBg4c6H/ap0+f1atXFxQUdO3a1ViT\nPuX06dPbt2/3z9C3b9+0tLQwu/r19vVt+48JcyUXW3M90+vZ6MUWtJqiKG63OyqbjjpVVW04\ndlVVbTtwIYTD4bDh2B0Oh20HLoRwOp02HLvL5bLzn3fJRCfYaZp2/vx5j8fjn6I/Pnv2rLEm\n/5SDBw8+/fTT/qdLly5t3bp1+B3+4avNQoiMbkPCX1XgChucp84tBrOsdZKTk6O49ShyOBy2\nHbttB+50Om07dtsOPC4uLi4uLtq9iA7bDlwyUavYWaRLly5z5szxP23RokVJSYlZKzerdPf1\n9vVB9qpG6U5/auKIQpWYmFhWVhatrUdRcnJydXV1eXl5tDsSaaqqxsfH23PgiYmJVVVVFy5c\niHZfIs3hcLhcLnsOPCEhobKysrKyMtp9iTS9Ymds4Lb9P0DMik6wUxSlSZMm586d80/RH6el\npRlr8k/5yU9+MnbsWP/TwsJCc/885eflpKWlNesxzPDiQohQu6SfdZeflxP1P7UJCQlR70Pk\nKYqSnJzs8/lsOHY7f8YnJiZWV1fbcOwul0tVVXsOPCEhwZ5pXghheKcT7GJN1O5j17Vr1927\nd/uf7t692+12t2/f3nCT3LhUAgAANMjaYFdZWVlQUFBQUFBZWVlSUlJQUHDs2DG9aezYsadO\nnVq2bNnx48e3bNmybt26UaNG6Re3GmsCAACwOUXTNOvWXlBQMH369MApqqquX///r+jcsWPH\nqlWrTpw4od9qePz48f5bDRtrqq2wsNDr9Zo7qPAPxTZeHo8n8FC4TSiKkp6e7vV6CwsLo92X\nSNMvGbHnwD0eT0VFRXFxcbT7Emkul8vtdttz4KmpqWVlZTY8k9jtdquqamzgGRkZpvcH4bA2\n2EUdwc5cBLto9yXSCHb2zDcEu2j3JdIIdjLhu2IBAAAkQbADAACQBMEOAABAEgQ7AAAASRDs\nAAAAJEGwAwAAkATBDgAAQBIEOwAAAEkQ7AAAACRBsAuZ4a+dEEJkZmWb2BMAAIBAzmh3oDEx\nJZbpK2ns3y0GAABiEBW7YJlbbKN0BwAATEfFrmEWhTBKdwAAwFxU7BpgdWmN0h0AADALFbuL\niljkonQHAABMQcWubpEvpFG6AwAAYSLYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiC\nYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDY1S0/L0f6LQIA\nAMkQ7C4qkkmLVAcAAMLnjHYHYpqetzKzsq3eBAAAQPio2DXMuuxFqgMAACaiYhcU00t3RDoA\nAGA6KnYhMCuNkeoAAIAVqNiFJj8vJy0trVmPYYYXN7c/AAAAflTsAAAAJEGwAwAAkATBDgAA\nQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7\nAAAASRDsAAAAJOGMdgcan2Y9hhleNjMrWwiRn5djXndiiz5AIfUYAQCIWVTsQpCZle0PLmGu\nJ/yVxKDAcZn1WgEAgOAR7IJlbkyRLPRcLMZJNkwAAGIch2IbZlE6keawbP2vjzTDBAAg9lGx\na4DVNadGXdMK/nhrox4mAACNBRW7i4pYFmmkNa1QX59GOkwAABoRKnZ1i3yFqXHVtBpXbwEA\nsAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYId\nAACAJAh2AAAAkpD8u2Lj4uLi4uKi3YtgJf2/9u4/psry/+P4dd/n8EvAW2lRQeqmAskUHQ0t\nW6PP5o8lH7FcmJqrLVu2sVIbf9VqbZ/m1lYtVjklXUswWcuTltXXRoNytrXOyK9RQip6VMa+\nahwISPBwzv394+xzjykcbg6H++a+7ufjr3Pu+76832+upq/e5wfp6XaXMAZVVSdY5NTvcTQe\nj8e5xcdNURTXNi6E8Hq9LuxdVVXXNi6ESE5Oju6+q3g8HlVVXdi4lCQPdrquRyIRu6swKxwO\n213CGHRdn2CRU7/H0Uy8dydSVdW1jQu3brpwceNCiEgk4sLeVVV1Z+NSkjzYhUKhUCgUx8LW\npnqLf899a1P9wMCAlXeMQ1paWrTIuH8+U7/H2ymKkpGREYlEnFj8BHk8nqSkJHc2Pm3atHA4\n7MLek5KSVFV1Z+NpaWlDQ0Mu7F0IEfemZ2RkJLwYTATvsRtVa1N9a1O9Zfey5kYJNN6arfx5\nAgDgTgS7MUx2FnF03DFfvHN7BADAQQh2Y5u87CVH3IndhaOTKwAAzkKwMyux6USyuDNaOzL1\nCADA1EewG4dExRRZ487wviRLrgAAOILkn4pNuNam+qysrOyiVXEvT2w9U430DQIAMJUxsQMA\nAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATB\nDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJCE1+4CHGbBvzYLIRRF1fVIHMvve2SjEKK1qT6OVXEs\nBAAArsLEbhyiqS5KUVRFifOnZwQ1M1cOv9j8QgAA4EIEO7OGpzrDRLLdmCltxAvMLAQAAO7E\nS7FjGzHSGaLZLu5XZkd8ddVM5uNlWQAAcAsmdmOIneoMCRzdmRzIMboDAAC3YGI3KpORzjDx\n0V0cQY3RHQAAMDCxG9l4U53Bgk9UJGohAACQDMEOAABAEgQ7AAAASRDsAAAAJEGwAwAAkATB\nDgAAQBIEOwAAAEkQ7AAAACRBsAMAAJAEwQ4AAEASBDsAAABJEOwAAAAkQbADAACQBMFuZGca\nP41voaLa8CO975GN1t8UAABMNQS7UZ1p/HRc8U5R1WiqUz1e1eOdtLpGRrYDAAAEuzGYzHa3\nD+psyXbEOwAA3IxgN7bYoztjUHc7RncAAMBKBDuzRsx2Zt5Rx+gOAABYg2A3DsNHdzEGdbdj\ndAcAACxAsBu3//vf/4nvo69kOwAAMKkIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACS\nINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJLw2l2A\nw1T+lCHEzX/954gQovG1x8a1NhIempyiRtXaVG/xHQEAgI2Y2I1D5U8Zw59G450ZkfCQxamu\ntameVAcAgNswsTPllkhnMDO6Y1AHAACswcRubKOlOsNoozsGdQAAwEpM7GIZM9IZbh/dMagD\nAAAWY2I3KvOpzhCNdwzqAACALQh2I4sj1UWZ/0QFAABAYjnvpVi/319bW3vlyhVN01asWLFp\n0yZFUewuCgAAwH4Om9i1tbW9+eabhYWF77777pYtW3w+38GDB+0uCgAAYEpw2MTO5/Pl5uZu\n27ZNCDFnzpzOzs6jR49WVFSkpKTYXRoAAIDNHDaxO3PmTHFxsfG0uLh4YGCgvb3dxpIAAACm\nCCdN7HRd7+7unjlzpnEk+rirq8s40t7e/vXXXxtP16xZk5OTY2WRtkhPT7fmRqqqWnavqcbj\n8biwd0VRXNu4EMLr9bqwd1VVXdu4ECI5OdmF79v2eDyKorhw06XkpGBnRiAQ+OSTT4yny5Yt\nmzdvXlx/0kCiSrJAWlqalPeaUlRVdW3vrm3c4/G4tnfXNu71er1e2f5lNCkpKcnuEpAATvrP\nV1GUGTNmBINB40j0cVZWlnFk8eLFu3fvNp7OmjWrp6cnrrs56U178fY4bpmZmb29vdbca0rR\nNG1oaKi/v9/uQqzm8XhSU1Nd2LiqqpmZmaFQ6J9//rG7Fqt5vd7k5GR3Np6enj44ODgw4KT/\nsU+I5ORkVVXja1zTtITXg4lwUrATQixYsKC5uXnr1q3Rp83NzampqXPnzjUuyMrKWrp0qfG0\np6cnFArFdSsnBbt4exw3Xdctu9fUEX1dxp29RyKRlJQUFzbu8XiEEJFIxIW9CyG8Xq87GxdC\nhMNhF/bu8Xjc+VeclBz24Yn169d3dHTs3bs3EAg0NjZ+8cUX5eXlfCQWAABAOG5iV1BQ8Oqr\nr9bV1R0/flzTtMcff3zz5s12FwUAADAlOCzYCSFKSkpKSkrsrgIAAGDKcdhLsQAAABgNwW5k\nHy7vi3tha1N9HAtbm+rjXhjHKgAAIB/nvRRrmWi2q/wpY1zXR7U21d/3yEaTC4cns7gXAgAA\nMLEbg8nR3e2XmUxdt19mcnRHqgMAALdgYje22KO7GMkvmr1Gm8DFTmYxRndEOgAAMCImdmaN\nGODMzPNGzGEmZ3IjzvPGXAgAANyJid04DB/djevTFcNHd+NNZsNHd6Q6AAAQA8Fu3A79O1nX\n9WG/sdasuGMZeQ4AAJjBS7EAAACSINgBAABIgmAHAAAgCYIdAACAJAh2AAAAkiDYAQAASIJg\nBwAAIAmCHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAg\nCYIdAACAJAh2AAAAkiDYAQAASIJgBwAAIAmCHQAAgCQUXdftrsFhqqurvV5vZWWl3YXAIuFw\n+K233po9e/aWLVvsrgUW6erq2rNnz8KFC8vLy+2uBRa5cOHCoUOHHnroodLSUrtrAeLHxG7c\nvvnmm+PHj9tdBawTDod9Pt+PP/5odyGwTl9fn8/n8/v9dhcC61y7ds3n8/3xxx92FwJMCMEO\nAABAEgQ7AAAASRDsAAAAJMGHJwAAACTBxA4AAEASBDsAAABJEOwAAAAk4bW7ACfx+/21tbVX\nrlzRNG3FihWbNm1SFMXuomDKn3/+efjw4fPnz1+9enXlypUvvvji8LMxdjbhp2CZhoaGH374\n4eLFi4ODgzk5OWVlZStXrjTOsulSOnHixJdfftnR0TE4OHjHHXc8/PDDGzduTEpKip5l0+EG\nnjfeeMPuGpyhra3t9ddfX758eWVl5axZsw4cOBAKhYqKiuyuC6Z0dHT09fWVlpZevHgxOzt7\n2bJlxqkYO5vwU7DSvn37CgsLo3lucHCwtrZ2xowZeXl5gk2XV0dHx7333rtmzZpVq1ZlZ2d/\n9tlnf/3119KlSwWbDtdgYmeWz+fLzc3dtm2bEGLOnDmdnZ1Hjx6tqKhISUmxuzSMraioKPoX\nrs/nu+VUjJ1N+Cmr23a3Xbt2GY8LCwsvXLhw8uTJRx99VLDp8lq+fLnxuKCgIBAInD59OvqU\nTYdL8B47s86cOVNcXGw8LS4uHhgYaG9vt7EkJESMnU34Kdjo5s2bmqZFH7Pp0otEIu3t7adO\nnVq8eHH0CJsOl2BiZ4qu693d3TNnzjSORB93dXXZVxQSIMbOJvzU5HeDUTU0NJw7d+75558X\nbLrsQqFQRUWFruu6rq9atYpNh9sQ7ABI7sSJE3v27Nm5c2f0DXaQm9frra6uDoVCZ8+eraur\nmz59+tNPP213UYB1CHamKIoyY8aMYDBoHIk+zsrKsq8oJECMnU34qcnvBiP49ttv9+/fX1VV\n9cADD0SPsOlyUxRlzpw5Qoj58+erqrp79+7169dnZGSw6XAJ3mNn1oIFC5qbm42nzc3Nqamp\nc+fOtbEkJESMnU34KVisvr7+448/fu2114xUF8Wmu8TQ0JCu60NDQ4JNh2vwdSdmZWdn+3y+\nnp6eO++889dffz1w4MC6deuGv3MWU9nNmzcDgUAwGDxx4kRaWlpubq7x/pgYO5vwU7DSRx99\ndOTIkeeeey4nJycYDAaDwb6+vujnJ9h0WdXU1PT29t64cePq1asnT548ePDgkiVLVq9eLdh0\nuIai67rdNTjGL7/8UldXd/ny5eh3UW7evJnvonSK9vb2HTt2DD+iquqRI0eij2PsbMJPwTJP\nPfVUb2/v8CN33313TU1N9DGbLqUDBw78/PPPV69eVVU1Ozu7tLR07dq1xvePsOlwA4IdAACA\nJHiPHQAAgCQIdgAAAJIg2AEAAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHwDqdnZ1V\nVVULFy7MzMzUNC0/P//JJ588ePCgcYHf71f+S1VVTdPy8vKeeOKJ2tragYEBGysHAEfgC4oB\nWOTcuXMPPvhgMBgsKysrKSnxeDznz5///vvvNU07depU9Bq/319SUrJ06dKKigohRH9//6VL\nlxoaGi5dupSfn//5558vWrTI1iYAYErz2l0AALfYtWvX9evX9+3bt3Xr1uHH29rabrly0aJF\nVVVVxtNIJPLBBx/s2LFj9erVLS0tWVlZVpQLAA7ES7EALHL27FkhxLp16245XlBQEHuhqqov\nvfTSjh07Ojs7q6urJ6s+AHA+gh0Ai8yfP18IUVdXF9/yyspKIcSxY8cSWRMAyIVgB8Air7zy\niqZpO3fuzM/P37Zt2969e1taWswvnzdvXmZm5u2v2wIADAQ7ABbJy8s7ffr0yy+/rChKTU3N\nCy+8sGjRoqKiopMnT5r8E6ZPn97f3x8Ohye1TgBwLoIdAOvMnj37nXfeaWtr6+7u/u677559\n9tnff/+9rKzs8uXLZpb//fff6enpHo9nsusEAIci2AGwgaZpK1eu3L9/f1VVVU9Pz6FDh8Zc\ncu7cud7e3jE/aQEAbkawA2Cn+++/XwhhZmL34YcfCiHWrl076TUBgGMR7ABY5NixYz09PcOP\n6LoendUVFhbGWBiJRN5///3q6up77rln+/btk1slADgZX1AMwCLvvffehg0bVqxYUVxcrGna\ntWvXvvrqq5aWlqKiomeeeWb4lb/99tvbb78thLhx40YgEGhoaAgEAnl5eYcPH545c6ZN5QOA\nA/ArxQBYxO/3+3y+xsbGixcvXr9+fdq0afn5+Y899tj27dszMjKMa0pKSqKPFUVJT0+/6667\nlixZUl5evmHDhtTUVPvKBwAHINgBAABIgvfYAQAASIJgBwAAIAmCHQAAgCQIdgAAAJIg2AEA\nAEiCYAcAACAJgh0AAIAkCHYAAACSINgBAABIgmAHAAAgCYIdAACAJP4fCzaHGL2faZIAAAAA\nSUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot the summary stats of the TPM data\n",
    "ggplot(myTPM_stats, aes(SD, MED)) +\n",
    "geom_hex(mapping = ) + labs(title=\"TPM standard deviation vs median for 04\")\n",
    "\n",
    "#we see a heteroscedasticity of the transcritps here. The transcripts expressed at\n",
    "#the highest levels have higher varience. This is why we will represent the data \n",
    "#on a log2 scale later\n",
    "\n",
    "#I am betting the transcripts with high expression and low variance are those which \n",
    "#are important diatoms in general and might be interesting to compare between \n",
    "#diatoms and haptophytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1e9ea8c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "using counts and average transcript lengths from tximport\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds <- DESeqDataSetFromTximport(\n",
    "  raw_counts,\n",
    "  colData=counts_metadata,\n",
    "  design = ~ treatment)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d3e1b6b",
   "metadata": {},
   "source": [
    "### Let's examine the dds object\n",
    "\n",
    "Use `assayNmes()` to print out the matrixes listed in the dds object. Let's rename the levels of the treatment column to more understandable terms from \"add_back\" and \"pFe12.9\" to \"Iron_ammendment\" and \"Low_Iron\". We are renaming the string here so we <b>must preserve the order of the levels in the original data!!!</b>. First, just print the levels of the dds then we can rename them. \n",
    "\n",
    "Use `relevel()` to set the Low_Iron treatment to the reference level. For the other organisms, we will treat High_Iron as the reference level. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e5e4c5e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 5 rows and 4 columns\n",
       "                    id   isolate treatment   replicate\n",
       "           <character> <numeric>  <factor> <character>\n",
       "4add_backA  4add_backA         4  add_back           A\n",
       "4add_backB  4add_backB         4  add_back           B\n",
       "4pFe21_9A    4pFe21_9A         4  pFe21_9            A\n",
       "4pFe21_9B    4pFe21_9B         4  pFe21_9            B\n",
       "4pFe21_9C    4pFe21_9C         4  pFe21_9            C"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'add_back'</li><li>'pFe21_9'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'add\\_back'\n",
       "\\item 'pFe21\\_9'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'add_back'\n",
       "2. 'pFe21_9'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"add_back\" \"pFe21_9\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>Iron_ammendment</li><li>Iron_ammendment</li><li>Low_Iron</li><li>Low_Iron</li><li>Low_Iron</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'Low_Iron'</li><li>'Iron_ammendment'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item Iron\\_ammendment\n",
       "\\item Iron\\_ammendment\n",
       "\\item Low\\_Iron\n",
       "\\item Low\\_Iron\n",
       "\\item Low\\_Iron\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'Low\\_Iron'\n",
       "\\item 'Iron\\_ammendment'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. Iron_ammendment\n",
       "2. Iron_ammendment\n",
       "3. Low_Iron\n",
       "4. Low_Iron\n",
       "5. Low_Iron\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'Low_Iron'\n",
       "2. 'Iron_ammendment'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] Iron_ammendment Iron_ammendment Low_Iron        Low_Iron       \n",
       "[5] Low_Iron       \n",
       "Levels: Low_Iron Iron_ammendment"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#dds\n",
    "#assayNames(dds) # <--this will show us the order of the count matrixes stored in the DeSeq object\n",
    "\n",
    "#head(adssay(dds),2) # <-- prints the second matrix, angTxLength\n",
    "\n",
    "#colSums(assay(dds)) #<-- sums the transcripts for each sample\n",
    "\n",
    "#rowRanges(dds) #shows the ranges of the first and last five contigs\n",
    "\n",
    "#extract the metadata from the rowRanges function\n",
    "#seqinfo(rowRanges(dds)) # <--we have none because this is a denovo analysis\n",
    "\n",
    "#metadata provided to dds object when read into tximport\n",
    "colData(dds)\n",
    "\n",
    "#if we want to rename the collomn isolate to culture, we would do so by:\n",
    "#dds$culture  <- dds$isolate\n",
    "\n",
    "#let's rename the treatement levels\n",
    "#first, we must know the order of the levels to rename them properly!!\n",
    "levels(dds$treatment)\n",
    "levels(dds$treatment)  <- c(\"Iron_ammendment\",\"Low_Iron\")\n",
    "\n",
    "#set Low Iron as the reference treatment(level) \n",
    "dds$treatment  <- relevel(dds$treatment, \"Low_Iron\")\n",
    "dds$treatment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb634f1b",
   "metadata": {},
   "source": [
    "### How many millions of contigs were mapped to the Salmon index for each sample?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e95f3842",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_mapped <- round(colSums(assay(dds))/1e6,1)\n",
    "head(counts_mapped)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "288664f9",
   "metadata": {},
   "source": [
    "### Pref-filtering Contigs with minimal expression\n",
    "We keep the contigs which at least two samples have 10 reads total. It is reccomended that number of samples (here set to 2) is that of the smalled group size. In this case there is an n=2 for Iron_ammendment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d667e20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "keep <- rowSums(counts(dds)>=10)>=2 #we use 2 at the end because the group\n",
    "                                                                            #with smallest n for this organism is 2\n",
    "dds <- dds[keep,]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9c260c4",
   "metadata": {},
   "source": [
    "# 3. Functional annotation\n",
    "------\n",
    "## Read in the eggnog-mapper functional annotations\n",
    "\n",
    "At this point, the variables/data we are working with and will move forward with are:\n",
    "- counts : dataframe of salmon count data, now containing eggnog output asociated with contigs\n",
    "- counts_metadata: experimental set up \n",
    "- dds: DESeq2 dataframe (list of matrixes)\n",
    "- res: results dataframe from DESeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87c423ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir2  <- \"/work/nclab/lucy/SAB/Annotation/eggnog/04\"\n",
    "emap <- read.csv(file.path(dir2,\"04.emapper.annotations\"),\n",
    "                 sep = \"\\t\",\n",
    "                 comment.char = \"#\",\n",
    "                 header = FALSE,\n",
    "                 na.strings = \"-\")[-c(1:4),]\n",
    "\n",
    "colnames(emap) <- c(\n",
    "        'query', 'seed_ortholog', 'evalue', 'score', 'eggNOG_OGs', 'max_annot_lvl', 'COG_category', 'Description', \n",
    "        'Preferred_name', 'GOs', 'EC', 'KEGG_ko', 'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction', \n",
    "        'KEGG_rclass',  'BRITE', 'KEGG_TC', 'CAZy', 'BiGG_Reaction', 'PFAMs')\n",
    "\n",
    "emap <- emap[, c(\"query\", \"seed_ortholog\", \"KEGG_ko\", \"GOs\",\"PFAMs\")]\n",
    "\n",
    "kegg <- emap[,c('query','KEGG_ko')]\n",
    "\n",
    "pfams <- emap[,c('query','PFAMs')]\n",
    "str(pfams)\n",
    "\n",
    "Go  <- emap[,c(\"query\",\"GOs\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0284be17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3ce0a654",
   "metadata": {},
   "source": [
    "# 4. Data transformation for visualizations\n",
    "------------\n",
    "- VST : variance stablizing tranformations\n",
    "- rlog : regularized logarithm \n",
    "\n",
    "The point of these two transformations, the VST and the rlog, is to <b>remove the dependence of the variance on the mean</b>, particularly the high variance of the logarithm of count data when the mean is low.\n",
    "\n",
    "Both VST and rlog perform log2 scale transformations, normalied with respoect to library size or other normalization factors. We can set these transformations to `blind`, which would re-estimate the dispersions using only an intercept, ignorming sample information specified by the design formula. Using `blind = TRUE` compares samples in a wholly unbiased manor from information about treatment groups. If you expect to see differences in the treatements, it is best to use `blind = FALSE`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db2ff16a",
   "metadata": {},
   "outputs": [],
   "source": [
    "vsd <- vst(dds,blind=FALSE)\n",
    "\n",
    "rld <- rlog(dds, blind=FALSE)\n",
    "\n",
    "counts_norm <- (assay(vsd))\n",
    "counts_norm <- as.data.frame(counts_norm) %>%\n",
    "tibble::rownames_to_column(\"query\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d6b87f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "colData(vsd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c684428",
   "metadata": {},
   "source": [
    "## 4.2 Add functional annotations to normalized counts \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a996f562",
   "metadata": {},
   "outputs": [],
   "source": [
    "nrow(counts_norm)\n",
    "nrow(vsd)\n",
    "nrow(dds)\n",
    "nrow(emap) ##??????how is the rows of emap > res?? Is this because res only has diff expressed contigs? \n",
    "#str(counts_norm)\n",
    "vsd_merged  <- left_join(counts_norm, emap, 'query')\n",
    "\n",
    "#head(vsd_merged)\n",
    "nrow(vsd_merged)==nrow(counts_norm)\n",
    "(nrow(vsd_merged))==(nrow(vsd))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b2eb3ea",
   "metadata": {},
   "source": [
    "## 4.3 Visualize effects of normalization\n",
    "Plot the first vs second sample of the dds object agains the vsd and rld normalized counts. We first must `estimteSizeFactors` on the dds to account for sequencing depth (this is automatically done with `vsd()`). Then, normalize the dds counts using the log2 function, we add 1 at the end to avoid taking the log of 0; and specify `normalized=TRUE` in the non-vsd plot. Create a df of all three normaization methods with a column at the end specifying the type of transformation performed. Plot the transformed counts using hexbin from ggplot. \n",
    "\n",
    "https://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#quantifying-with-salmon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fefbb3ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "dds_est <- estimateSizeFactors(dds)\n",
    "df <- bind_rows(\n",
    "                            as.data.frame(log2(counts(dds_est,  normalized = TRUE)[, 1:2]  + 1)) %>%\n",
    "                                    mutate(transformation = \"log2(x+1)\"),\n",
    "                            as.data.frame(assay(vsd)[, 1:2]) %>%\n",
    "                                     mutate(transformation = \"vst\"),\n",
    "                            as.data.frame(assay(rld)[, 1:2]) %>%\n",
    "                                    mutate(transformation = \"rlog\"))\n",
    "\n",
    "colnames(df)[1:2] <- c(\"x\",\"y\")\n",
    "\n",
    "lvls <- c(\"log2(x+1)\", \"vst\", \"rlog\")\n",
    "\n",
    "df$transformation  <- factor(df$transformation, levels=lvls)\n",
    "\n",
    "ggplot(df, aes(x=x, y=y)) + geom_hex(bins=80) + coord_fixed() + facet_wrap(~transformation)\n",
    "\n",
    "#meanSdPlot(assay(ntd))\n",
    "#meanSdPlot(assay(rld))\n",
    "#meanSdPlot(assay(vsd))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1a9dad6",
   "metadata": {},
   "source": [
    "Note how the bins are shifted up for the vst plot. This enables us to better vissualize the distances between smaller values in the pca plots. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "006cc9f3",
   "metadata": {},
   "source": [
    "## 4.4 Sample Distances\n",
    "Which samples are similare which are different? Does this fit with our expectations? \n",
    "\n",
    "Using the `dist()` calcualte the Euclidean distances between samples, using the VSD data to ensure equal contribution among contigs. `dist()` expects rows to be different samples columns to be the contigs, so transpose the matrix using `t()`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "380b0494",
   "metadata": {},
   "outputs": [],
   "source": [
    "vsd_dist <- dist(t(assay(vsd)))\n",
    "vsd_dist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a675b6e5",
   "metadata": {},
   "source": [
    "Plot a heatmap of the distance matrix, arranging the rows and columns by the distances. Provide the sample distances from the vsd_dist to the `clustering_distance` argument to cluster sample which are must similar. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8108f7dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "vsd_dist_mat <- as.matrix(vsd_dist)\n",
    "rownames(vsd_dist_mat) <- paste(vsd$id)\n",
    "\n",
    "colors <- colorRampPalette(rev(brewer.pal(9,\"Blues\")))(255)\n",
    "pheatmap(vsd_dist_mat,\n",
    "        clustering_distance_rows = vsd_dist,\n",
    "        clustering_distance_cols = vsd_dist,\n",
    "        col = colors)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46397684",
   "metadata": {},
   "source": [
    "Here we see how similare the samples are to one annother. As expected, the two treatment types separate, but we see replicate C of low iron is different from the rest. \n",
    "\n",
    "We can also calculate a distance matrix from the original dds object using the Poisson Distance. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5864004e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#install.packages('PoiClaClu')\n",
    "library('PoiClaClu')\n",
    "poisd <- PoissonDistance(t(counts(dds)))\n",
    "\n",
    "poisd_mat <- as.matrix(poisd$dd)\n",
    "rownames(poisd_mat) <- dds$id\n",
    "pheatmap(poisd_mat, \n",
    "        clustering_distance_rows = poisd$dd,\n",
    "        clustering_distance_cols = poisd$dd,\n",
    "        col = colors)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40ade9c4",
   "metadata": {},
   "source": [
    "## 4.5 The PCA plot also shows sample-to-sample distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26c96bcc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plotPCA(vsd, intgroup=c(\"treatment\",\"replicate\"))\n",
    "\n",
    "\n",
    "#using ggplot to customize PCA plots\n",
    "pcaData <- plotPCA(vsd, intgroup=c(\"treatment\", \"replicate\"), returnData=TRUE)\n",
    "\n",
    "percentVar <- round(100 * attr(pcaData, \"percentVar\"))\n",
    "\n",
    "ggplot(pcaData, aes(PC1, PC2, color=treatment, shape=replicate)) +\n",
    "  geom_point(size=3) +\n",
    "  xlab(paste0(\"PC1: \",percentVar[1],\"% variance\")) +\n",
    "  ylab(paste0(\"PC2: \",percentVar[2],\"% variance\")) + \n",
    "  coord_fixed()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85f663d4",
   "metadata": {},
   "source": [
    "# Tests of log2 fold change above or below a threshold\n",
    "It is also possible to provide thresholds for constructing Wald tests of significance. Two arguments to the results function allow for threshold-based Wald tests: lfcThreshold, which takes a numeric of a non-negative threshold value, and altHypothesis, which specifies the kind of test. Note that the alternative hypothesis is specified by the user, i.e. those genes which the user is interested in finding, and the test provides p values for the null hypothesis, the complement of the set defined by the alternative. The altHypothesis argument can take one of the following four values, where \\(\\beta\\) is the log2 fold change specified by the name argument, and \\(x\\) is the lfcThreshold.\n",
    "\n",
    "greaterAbs - \\(|\\beta| > x\\) - tests are two-tailed\n",
    "lessAbs - \\(|\\beta| < x\\) - p values are the maximum of the upper and lower tests\n",
    "greater - \\(\\beta > x\\)\n",
    "less - \\(\\beta < -x\\)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c002c1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "par(mfrow=c(2,2),mar=c(2,2,1,1))\n",
    "ylim <- c(-2.5,2.5)\n",
    "resGA <- results(dds, lfcThreshold=.5, altHypothesis=\"greaterAbs\")\n",
    "resLA <- results(dds, lfcThreshold=.5, altHypothesis=\"lessAbs\")\n",
    "resG <- results(dds, lfcThreshold=.5, altHypothesis=\"greater\")\n",
    "resL <- results(dds, lfcThreshold=.5, altHypothesis=\"less\")\n",
    "drawLines <- function() abline(h=c(-.5,.5),col=\"dodgerblue\",lwd=2)\n",
    "plotMA(resGA, ylim=ylim); drawLines()\n",
    "plotMA(resLA, ylim=ylim); drawLines()\n",
    "plotMA(resG, ylim=ylim); drawLines()\n",
    "plotMA(resL, ylim=ylim); drawLines()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da23aa6d",
   "metadata": {},
   "source": [
    "# Heat map with annotations\n",
    "\n",
    "this is taken from code from https://github.com/UVa-R-Users-Group/meetup/blob/master/2015-02-19-heat-maps/heatmaps.R\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a18d3fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#using gplots::heatmaps.2()\n",
    "\n",
    "\n",
    "#using 2 pallets\n",
    "pal1  <- brewer.pal(name=\"Blues\", n=9)\n",
    "pal2 <- colorRampPalette(pal1)(50)\n",
    "\n",
    "heatmap.2(as.matrix(vsd), col=pal2, \n",
    "          hclustfun=hclust2,\n",
    "          distfun=dist2,\n",
    "          scale=\"row\",\n",
    "          sepcolor=\"black\",\n",
    "          dendrogram=\"both\",\n",
    "          key=TRUE,\n",
    "          symkey=FALSE,\n",
    "          denstiy.info=\"none\",\n",
    "          trace=\"none\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "009a9b9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "select <- order(rowMeans(counts(dds,normalized=TRUE)),\n",
    "                decreasing=TRUE)\n",
    "\n",
    "df <- as.data.frame(colData(dds)[,c(\"treatment\",\"replicate\")])\n",
    "\n",
    "pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=FALSE,\n",
    "         cluster_cols=FALSE, annotation_col=df)\n",
    "\n",
    "pheatmap((vsd_merged[,2:6], cluster_rows=FALSE, show_rownames=FALSE,\n",
    "         cluster_cols=TRUE, annotation_col=df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c736a9f6",
   "metadata": {},
   "source": [
    "# 5 Differential expression analysis\n",
    "------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cd2a5e5",
   "metadata": {},
   "source": [
    "## 5.1 DE extract results\n",
    "\n",
    "All de steps are calculated in on step, DESeq(). The results are extracted with results() giving the logfold2, p-values, and adjusted p-values. Here we identify the name of the comparison we wish to do, this is becuase DEseq automatically sets the control group to the group which is first alphabetically. \n",
    "\"\n",
    "With no additional arguments to results, the log2 fold change and Wald test p value will be for the last variable in the design formula, and if this is a factor, the comparison will be the last level of this variable over the reference level\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ef384f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "dds <- DESeq(dds)\n",
    "res <- results(dds, name=\"treatment_Iron_ammendment_vs_Low_Iron\",\n",
    "             tidy=TRUE)\n",
    "\n",
    "head(res)\n",
    "str(res)\n",
    "\n",
    "\n",
    "sum(res$padj < 0.05, na.rm=T) #sum up how many adjusted p-values are significant"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13c3d91c",
   "metadata": {},
   "source": [
    "## 5.2 Merging the eggnog-mapper annotations file with the results file of the DESeq analysis \n",
    "\n",
    "First, rename the 'row' column to query in the res file. We will merge with left_join dplyr function, this should only include rows from the emap dataframe which are also found in the res dataframe; meaning we would not loose any unannotated contigs. To assure this is the case, count the number of rows for each dataframe before merging, and checked that the merged output hase equal rows to the res datatable. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52fa3bf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(res)[1]  <- \"query\" #rename the 'row' column to 'query' so it can be merged with eggnog output\n",
    "\n",
    "nrow(res)\n",
    "nrow(emap) ##??????how is the rows of emap > res?? Is this because res only has diff expressed contigs? \n",
    "\n",
    "res_merged  <- left_join(res, emap, 'query')\n",
    "\n",
    "head(res_merged,2)\n",
    "\n",
    "(nrow(res_merged))==(nrow(res))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9912e54f",
   "metadata": {},
   "source": [
    "## 5.3 Order the results by smallest p value and extract top 100 differentially expressed contigs\n",
    "\n",
    "Use sumary() to view summary stats of various results data frames.\n",
    "Create a table of the top 100 most differentially expressed contigs and call it top_100. Filter the mail results table by those which have significant expression, p-adjust < 0.05. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0abf8b41",
   "metadata": {},
   "outputs": [],
   "source": [
    "resOrdered <- res_merged[order(res_merged$pvalue),]\n",
    "\n",
    "top_100  <- resOrdered[c(1:100),]\n",
    "head(top_100)\n",
    "#top 100 diff expressed contigs\n",
    "#summary(top_100)\n",
    "\n",
    "#unfiltered results\n",
    "#summary(res_merged)\n",
    "\n",
    "#all contigs with signigifcant adjusted p-values\n",
    "res_Sig  <-filter(res_merged, padj < 0.1, na.rm = TRUE)\n",
    "summary(res_Sig)\n",
    "\n",
    "#order the significant congits by log2FoldChange to got those with the strongest DOWN regulation\n",
    "res_Sig_down <-  res_Sig[order(res_Sig$log2FoldChange), ]\n",
    "res_100_down <- res_Sig_down[c(1:100), ]\n",
    "\n",
    "#order the significant congits by log2FoldChange to got those with the strongest UP regulation\n",
    "res_Sig_up <-  res_Sig[order(res_Sig$log2FoldChange, decreasing = TRUE), ]\n",
    "res_100_up <- res_Sig_up[c(1:100), ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90cb7627",
   "metadata": {},
   "source": [
    "## 5.4 Log Fold Change Shrinkage\n",
    "lfcShrink is helpful to <b>visualize</b> and <b>rank</b> genes. Here, we use the apelglm method to measure effect size shrinkage. We provide the function with the dds object and the name or number of the coefficient we want to shrink. \n",
    "\n",
    "We then order the results like earlier and extract the top 100 contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03331203",
   "metadata": {},
   "outputs": [],
   "source": [
    "resultsNames(dds)\n",
    "resLFC <- lfcShrink(dds,coef='treatment_Iron_ammendment_vs_Low_Iron',\n",
    "                    type='apeglm')\n",
    "resLFCdf <- as.data.frame(resLFC) %>% tibble::rownames_to_column(\"query\")\n",
    "\n",
    "\n",
    "LFC_merged  <- left_join(resLFCdf, emap, 'query')\n",
    "\n",
    "(nrow(LFC_merged))==(nrow(resLFC))\n",
    "\n",
    "LFC_ordered <- LFC_merged[order(LFC_merged$pvalue),]\n",
    "topLFC_100  <- LFC_ordered[c(1:100),]\n",
    "head(topLFC_100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ec75ab6",
   "metadata": {},
   "source": [
    "## 5.6 Plotting the DE data\n",
    "\n",
    "### 5.6.1 Plot Counts\n",
    "\n",
    "Here we plot counts of specific contigs, grouped by the intgroup variable. The contig may be selected by rowname (contig id) or numberic index. Here, we <b>plotted the contig with the most significant change between treatments</b>. To customize these plots using ggplot, add the argument `retunData = TRUE` to output a data frame. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ad376c6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "plotCounts(dds, gene=which.min(res$padj), intgroup=\"treatment\", main = \"Contig with smalled pvalue\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3da2e0a1",
   "metadata": {},
   "source": [
    "For a particular gene, a log2 fold change of -1 for condition treated vs untreated means that the treatment induces a multiplicative change in observed gene expression level of \\(2^{-1} = 0.5\\) compared to the untreated condition. If the variable of interest is continuous-valued, then the reported log2 fold change is per unit of change of that variable.\n",
    "\n",
    "<b>Note on p-values set to NA</b>: some values in the results table can be set to NA for one of the following reasons:\n",
    "\n",
    "If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p value and adjusted p value will all be set to NA.\n",
    "If a row contains a sample with an extreme count outlier then the p value and adjusted p value will be set to NA. These outlier counts are detected by Cook’s distance. Customization of this outlier filtering and description of functionality for replacement of outlier counts and refitting is described below\n",
    "If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p value will be set to NA. Description and customization of independent filtering is described below\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff7900e",
   "metadata": {},
   "source": [
    "### 5.6.2 MA-plots \n",
    "Showing the distribution of estimated coefficients in the model. \n",
    "point out some of the most differentially expressed genes in ppt with arrow. Look for genes most differentially expressed and for those I am interested in to see where they are. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387a4ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotMA(resLFC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b4898aa",
   "metadata": {},
   "source": [
    "# Gene Clustering\n",
    "\n",
    "### We can cluster the genes like we clustered the samples by distance. In this case, we cluser contigs which have the highest variance across samples\n",
    "\n",
    "Subset the top 20 most variable contigs and call it top_var, using the normalized count data vsd. Use `rowVars` to estimate variance across rows. \n",
    "\n",
    "Center the contigs value across samples to show the amount a contigs deviates in a sample from the average across all samples.  This will generate a heatmap of relative VST-transformed values across samples. Treatment and replicate are shown with color bars at the top of the heatmap "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8ba665b6",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': error in evaluating the argument 'x' in selecting a method for function 'rowVars': error in evaluating the argument 'x' in selecting a method for function 'assay': object 'vsd' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': error in evaluating the argument 'x' in selecting a method for function 'rowVars': error in evaluating the argument 'x' in selecting a method for function 'assay': object 'vsd' not found\nTraceback:\n",
      "1. head(order(rowVars(assay(vsd)), decreasing = TRUE), 20)",
      "2. order(rowVars(assay(vsd)), decreasing = TRUE)",
      "3. standardGeneric(\"order\")",
      "4. eval(quote(list(...)), env)",
      "5. eval(quote(list(...)), env)",
      "6. eval(quote(list(...)), env)",
      "7. rowVars(assay(vsd))",
      "8. assay(vsd)",
      "9. .handleSimpleError(function (cond) \n . .Internal(C_tryCatchHelper(addr, 1L, cond)), \"object 'vsd' not found\", \n .     base::quote(eval(expr, envir, enclos)))",
      "10. h(simpleError(msg, call))",
      "11. .handleSimpleError(function (cond) \n  . .Internal(C_tryCatchHelper(addr, 1L, cond)), \"error in evaluating the argument 'x' in selecting a method for function 'assay': object 'vsd' not found\", \n  .     base::quote(h(simpleError(msg, call))))",
      "12. h(simpleError(msg, call))",
      "13. .handleSimpleError(function (cond) \n  . .Internal(C_tryCatchHelper(addr, 1L, cond)), \"error in evaluating the argument 'x' in selecting a method for function 'rowVars': error in evaluating the argument 'x' in selecting a method for function 'assay': object 'vsd' not found\", \n  .     base::quote(h(simpleError(msg, call))))",
      "14. h(simpleError(msg, call))"
     ]
    }
   ],
   "source": [
    "top_var <- head(order(rowVars(assay(vsd)), decreasing = TRUE),20)\n",
    "\n",
    "mat_var <- assay(vsd)[top_var,] #extract count values for the most differentially expressed contigs\n",
    "#mat_var <- mat_var - rowMeans(mat_var) #center each cell in matrix by the avereage across samples (setting scale to true?)\n",
    "anno <- as.data.frame(colData(vsd)[, c(\"treatment\",\"replicate\")])\n",
    "\n",
    "pheatmap(mat_var,  annotation_col = anno, show_rownames = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37e1c8dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "pfams  <- filter(pfams, complete.cases(PFAMs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0d1f1f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "vsd_df <- as.data.frame(assay(vsd)) %>% tibble::rownames_to_column('query')\n",
    "#newrows <- c(pfams$query)\n",
    "vsd_df <- vsd_df %>% filter(query %in% pfams$query)  \n",
    "vsd_df <- left_join(vsd_df, pfams, by='query')\n",
    "head(vsd_df)\n",
    "\n",
    "vsd_df <- tibble::column_to_rownames(vsd_df, var= \"query\")\n",
    "vsd_mat <- vsd_df[,-6] %>% as.matrix()\n",
    "#vsd_mat = matrix of contigs which have pfam values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fe73915",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_var <- head(order(rowVars(vsd_mat), decreasing = TRUE),20)\n",
    "\n",
    "mat_var <- vsd_mat[top_var,] #extract count values for the most differentially expressed contigs\n",
    "mat_var <- mat_var - rowMeans(mat_var) #center each cell in matrix by the avereage across samples\n",
    "anno <- as.data.frame(colData(vsd)[, c(\"treatment\",\"replicate\")])\n",
    "\n",
    "pheatmap(mat_var,  annotation_col = anno, show_rownames = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1b0bb7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "order(vsd_mat,decreasing = TRUE)\n",
    "#we need \n",
    "left_join(vsd_var, pfams,by='query')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cc819ab",
   "metadata": {},
   "source": [
    "## Examine ratio of small p-values for binned mean normalize count of contigs\n",
    "\n",
    "As displayed in the MA plot, weakly expressed genes are not obserbed by differential expression analysis. This s because low read counts have high Poisson noise, uncertainties from low sampling rates drown out any biological effect. Use the results table subjected to the threshold to show what this looks like in a case wehen there are few tests with small p values. We will use `quantile()`  to bin contigs by base mean and use `cut` to rename the levels of the bins by the median point. Then we will calculate the ration of the p-value < 0.05 for each bin. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "194761c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "qs <- c(0, quantile(res$baseMean[res$baseMean > 0], 0:6/6))\n",
    "\n",
    "bins <- cut(res$baseMean, qs)\n",
    "levels(bins) <- paste(\"~\", round(signif ( ( qs[-1] + qs[-length(qs) ]) / 2, 2 ) ) )\n",
    "fractionSig <- tapply(res$pvalue, bins, function(p)\n",
    "    mean(p < 0.05, na.rm=TRUE))\n",
    "barplot(fractionSig, xlab=\"mean normalized count\", ylab=\"fraction of small p values\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7139bad5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9670455",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd584919",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe4e1968",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
