{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2eeabfee",
   "metadata": {},
   "source": [
    "# Differential gene expression\n",
    "### Prepare environment\n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "da4d6982",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ────────────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "── \u001b[1mConflicts\u001b[22m ──────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiobase\u001b[39m::combine(), \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcount()\u001b[39m        masks \u001b[34mmatrixStats\u001b[39m::count()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in library(\"Glimma\", quietly = T): there is no package called ‘Glimma’\n",
     "output_type": "error",
     "traceback": [
      "Error in library(\"Glimma\", quietly = T): there is no package called ‘Glimma’\nTraceback:\n",
      "1. library(\"Glimma\", quietly = T)"
     ]
    }
   ],
   "source": [
    " \n",
    "library('tximport', quietly=T)\n",
    "library('DESeq2',quietly=T)\n",
    "library('ashr',quietly=T)\n",
    "library('tibble',quietly=T)\n",
    "library('tidyverse',quietly=T)\n",
    "library('Glimma',quietly=T)\n",
    "library('RCurl',quietly=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbdaeac3",
   "metadata": {},
   "source": [
    "# This notebook will create and save two data types:\n",
    "1. matrix of significant differentially expressed orf's from the results of the differential expression analysis\n",
    "2. matrix of VSD normalized counts ordered by variance across samples\n",
    "## Both of these files will be created for transcript-level data and \"gene-level\" data using only transcripts with Kegg annotations and summing by Kegg annotation. \n",
    "#### The second method will result in data with rownames as Kegg annotations, meaning each annotation appears once in the matrix. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c10b8cd6",
   "metadata": {},
   "source": [
    "# 1. Read in Salmon Counts\n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "020b7d25-9166-4aad-b31a-b6522d94eaf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         id isolate treatment rep\n",
      "1 add_backA      04  Add_Back   A\n",
      "2 add_backB      04  Add_Back   B\n",
      "3  pFe21_9A      04  Low_Iron   A\n",
      "4  pFe21_9B      04  Low_Iron   B\n",
      "5  pFe21_9C      04  Low_Iron   C\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         id isolate treatment rep\n",
      "1 add_backB      08  Add_Back   B\n",
      "2 add_backC      08  Add_Back   C\n",
      "3  pFe21_9A      08  Low_Iron   A\n",
      "4    pFe19A      08 High_Iron   A\n",
      "5    pFe19B      08 High_Iron   B\n",
      "6    pFe19C      08 High_Iron   C\n",
      "7  pFe21_9B      08  Low_Iron   B\n",
      "8  pFe21_9C      08  Low_Iron   C\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         id isolate treatment rep\n",
      "1 add_backA      06  Add_Back   A\n",
      "2 add_backB      06  Add_Back   B\n",
      "3 add_backC      06  Add_Back   C\n",
      "4    pFe19A      06 High_Iron   A\n",
      "5    pFe19B      06 High_Iron   B\n",
      "6    pFe19C      06 High_Iron   C\n",
      "7  pFe21_9A      06  Low_Iron   A\n",
      "8  pFe21_9B      06  Low_Iron   B\n",
      "9  pFe21_9C      06  Low_Iron   C\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         id isolate treatment rep\n",
      "1 add_backA      13  Add_Back   A\n",
      "2 add_backB      13  Add_Back   B\n",
      "3 add_backC      13  Add_Back   C\n",
      "4    pFe19A      13 High_Iron   A\n",
      "5    pFe19B      13 High_Iron   B\n",
      "6    pFe19C      13 High_Iron   C\n",
      "7  pFe21_9A      13  Low_Iron   A\n",
      "8  pFe21_9B      13  Low_Iron   B\n",
      "9  pFe21_9C      13  Low_Iron   C\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 8 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>add_backB</th><th scope=col>add_backC</th><th scope=col>pFe21_9A</th><th scope=col>pFe19A</th><th scope=col>pFe19B</th><th scope=col>pFe19C</th><th scope=col>pFe21_9B</th><th scope=col>pFe21_9C</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>NODE_10000_length_1734_cov_7.947622_g6638_i0.p1</th><td> 34.000</td><td> 15.000</td><td> 9.000</td><td> 5.000</td><td> 7.000</td><td>15.000</td><td> 10.000</td><td>  1.000</td></tr>\n",
       "\t<tr><th scope=row>NODE_10001_length_1733_cov_212.025904_g6639_i0.p1</th><td>677.000</td><td>263.878</td><td>49.000</td><td>37.000</td><td>30.000</td><td>58.000</td><td>108.000</td><td>141.237</td></tr>\n",
       "\t<tr><th scope=row>NODE_10002_length_1733_cov_138.864458_g1003_i6.p1</th><td>165.923</td><td> 88.899</td><td>28.753</td><td>14.298</td><td>16.228</td><td>44.130</td><td> 12.042</td><td> 26.511</td></tr>\n",
       "\t<tr><th scope=row>NODE_10003_length_1733_cov_40.357229_g5975_i1.p2</th><td>  6.055</td><td>  1.363</td><td> 1.001</td><td> 1.098</td><td> 0.000</td><td> 3.387</td><td>  9.037</td><td>  4.809</td></tr>\n",
       "\t<tr><th scope=row>NODE_10003_length_1733_cov_40.357229_g5975_i1.p1</th><td>  4.000</td><td>  2.000</td><td> 4.000</td><td> 1.000</td><td> 0.000</td><td> 7.000</td><td>  4.000</td><td>  1.000</td></tr>\n",
       "\t<tr><th scope=row>NODE_10004_length_1733_cov_25.593373_g6640_i0.p1</th><td> 70.000</td><td> 40.000</td><td>11.000</td><td>11.000</td><td> 5.000</td><td> 9.000</td><td> 34.000</td><td> 17.000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 8 of type dbl\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & add\\_backB & add\\_backC & pFe21\\_9A & pFe19A & pFe19B & pFe19C & pFe21\\_9B & pFe21\\_9C\\\\\n",
       "\\hline\n",
       "\tNODE\\_10000\\_length\\_1734\\_cov\\_7.947622\\_g6638\\_i0.p1 &  34.000 &  15.000 &  9.000 &  5.000 &  7.000 & 15.000 &  10.000 &   1.000\\\\\n",
       "\tNODE\\_10001\\_length\\_1733\\_cov\\_212.025904\\_g6639\\_i0.p1 & 677.000 & 263.878 & 49.000 & 37.000 & 30.000 & 58.000 & 108.000 & 141.237\\\\\n",
       "\tNODE\\_10002\\_length\\_1733\\_cov\\_138.864458\\_g1003\\_i6.p1 & 165.923 &  88.899 & 28.753 & 14.298 & 16.228 & 44.130 &  12.042 &  26.511\\\\\n",
       "\tNODE\\_10003\\_length\\_1733\\_cov\\_40.357229\\_g5975\\_i1.p2 &   6.055 &   1.363 &  1.001 &  1.098 &  0.000 &  3.387 &   9.037 &   4.809\\\\\n",
       "\tNODE\\_10003\\_length\\_1733\\_cov\\_40.357229\\_g5975\\_i1.p1 &   4.000 &   2.000 &  4.000 &  1.000 &  0.000 &  7.000 &   4.000 &   1.000\\\\\n",
       "\tNODE\\_10004\\_length\\_1733\\_cov\\_25.593373\\_g6640\\_i0.p1 &  70.000 &  40.000 & 11.000 & 11.000 &  5.000 &  9.000 &  34.000 &  17.000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 8 of type dbl\n",
       "\n",
       "| <!--/--> | add_backB | add_backC | pFe21_9A | pFe19A | pFe19B | pFe19C | pFe21_9B | pFe21_9C |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| NODE_10000_length_1734_cov_7.947622_g6638_i0.p1 |  34.000 |  15.000 |  9.000 |  5.000 |  7.000 | 15.000 |  10.000 |   1.000 |\n",
       "| NODE_10001_length_1733_cov_212.025904_g6639_i0.p1 | 677.000 | 263.878 | 49.000 | 37.000 | 30.000 | 58.000 | 108.000 | 141.237 |\n",
       "| NODE_10002_length_1733_cov_138.864458_g1003_i6.p1 | 165.923 |  88.899 | 28.753 | 14.298 | 16.228 | 44.130 |  12.042 |  26.511 |\n",
       "| NODE_10003_length_1733_cov_40.357229_g5975_i1.p2 |   6.055 |   1.363 |  1.001 |  1.098 |  0.000 |  3.387 |   9.037 |   4.809 |\n",
       "| NODE_10003_length_1733_cov_40.357229_g5975_i1.p1 |   4.000 |   2.000 |  4.000 |  1.000 |  0.000 |  7.000 |   4.000 |   1.000 |\n",
       "| NODE_10004_length_1733_cov_25.593373_g6640_i0.p1 |  70.000 |  40.000 | 11.000 | 11.000 |  5.000 |  9.000 |  34.000 |  17.000 |\n",
       "\n"
      ],
      "text/plain": [
       "                                                  add_backB add_backC pFe21_9A\n",
       "NODE_10000_length_1734_cov_7.947622_g6638_i0.p1    34.000    15.000    9.000  \n",
       "NODE_10001_length_1733_cov_212.025904_g6639_i0.p1 677.000   263.878   49.000  \n",
       "NODE_10002_length_1733_cov_138.864458_g1003_i6.p1 165.923    88.899   28.753  \n",
       "NODE_10003_length_1733_cov_40.357229_g5975_i1.p2    6.055     1.363    1.001  \n",
       "NODE_10003_length_1733_cov_40.357229_g5975_i1.p1    4.000     2.000    4.000  \n",
       "NODE_10004_length_1733_cov_25.593373_g6640_i0.p1   70.000    40.000   11.000  \n",
       "                                                  pFe19A pFe19B pFe19C pFe21_9B\n",
       "NODE_10000_length_1734_cov_7.947622_g6638_i0.p1    5.000  7.000 15.000  10.000 \n",
       "NODE_10001_length_1733_cov_212.025904_g6639_i0.p1 37.000 30.000 58.000 108.000 \n",
       "NODE_10002_length_1733_cov_138.864458_g1003_i6.p1 14.298 16.228 44.130  12.042 \n",
       "NODE_10003_length_1733_cov_40.357229_g5975_i1.p2   1.098  0.000  3.387   9.037 \n",
       "NODE_10003_length_1733_cov_40.357229_g5975_i1.p1   1.000  0.000  7.000   4.000 \n",
       "NODE_10004_length_1733_cov_25.593373_g6640_i0.p1  11.000  5.000  9.000  34.000 \n",
       "                                                  pFe21_9C\n",
       "NODE_10000_length_1734_cov_7.947622_g6638_i0.p1     1.000 \n",
       "NODE_10001_length_1733_cov_212.025904_g6639_i0.p1 141.237 \n",
       "NODE_10002_length_1733_cov_138.864458_g1003_i6.p1  26.511 \n",
       "NODE_10003_length_1733_cov_40.357229_g5975_i1.p2    4.809 \n",
       "NODE_10003_length_1733_cov_40.357229_g5975_i1.p1    1.000 \n",
       "NODE_10004_length_1733_cov_25.593373_g6640_i0.p1   17.000 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#fixing tximport function \n",
    "\n",
    "pattern='[[:alpha:]]+([[:digit:]]{2}|_)(_9[[:alpha:]]|[[:alpha:]]*)'\n",
    "\n",
    "read.in <- function(org){\n",
    "    dir <- paste(\"/work/nclab/lucy/SAB/Assembly/\",org,\"/salmon\",sep='')\n",
    "    files <- file.path(dir,list.files(dir,pattern=\".sf\",recursive=TRUE))\n",
    "    \n",
    "    names(files)=str_extract(files,pattern)\n",
    "    names(files)=str_replace(names(files),'oFe', 'pFe') #correct a sample for 08 from oFe to pFe\n",
    "\n",
    "    if (all(file.exists(files)) == FALSE) {\n",
    "        print(\"ERROR IN FILE NAMES, not all files exist\")\n",
    "        print(paste(\"Directory:\", dir, sep=\"/n\"))\n",
    "        print(paste(\"Files:\", files, sep='/n'))\n",
    "    }\n",
    "    \n",
    "    raw_counts <- tximport(files, type='salmon', txOut = TRUE) \n",
    "}\n",
    "\n",
    "create.metadata=function(org){\n",
    "    dir <- paste(\"/work/nclab/lucy/SAB/Assembly/\",org,\"/salmon\",sep='')\n",
    "    files <- file.path(dir,list.files(dir,pattern=\".sf\",recursive=TRUE))\n",
    "    \n",
    "    id=str_extract(files,pattern)\n",
    "    id=str_replace(id,'oFe', 'pFe')\n",
    "    metadata=data.frame('id'=id,'isolate'=org,'treatment'=str_extract(id,'[[:alpha:]]+(19|21_9|_back)'),'rep'=str_extract(id, 'A|B|C'))\n",
    "    metadata$treatment=str_replace_all(metadata$treatment, c('pFe19'='High_Iron', 'pFe21_9'='Low_Iron','add_back'='Add_Back'))\n",
    "    metadata\n",
    "    print(metadata)                    \n",
    "}\n",
    "counts_4=read.in('04')\n",
    "metadata_4=create.metadata('04')\n",
    "\n",
    "counts_8=read.in('08')\n",
    "metadata_8=create.metadata('08')\n",
    "\n",
    "counts_6=read.in('06')\n",
    "metadata_6=create.metadata('06')\n",
    "\n",
    "counts_13=read.in('13')\n",
    "metadata_13=create.metadata('13')\n",
    "head(counts_8$counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cba9d37e-169b-42c5-a63d-ee1f660d6ad1",
   "metadata": {},
   "source": [
    "### wrong and old read in function\n",
    "read.in <- function(org,meta){\n",
    "    dir <- paste(\"/work/nclab/lucy/SAB/Assembly/\",org,\"/salmon\",sep='')\n",
    "    \n",
    "    counts_metadata <- meta\n",
    "    files <- file.path(dir,list.files(dir,pattern=\".sf\",recursive=TRUE))\n",
    "    \n",
    "    names(files) <- paste0(counts_metadata$id)\n",
    "\n",
    "    if (all(file.exists(files)) == FALSE) {\n",
    "        print(\"ERROR IN FILE NAMES, not all files exist\")\n",
    "        print(paste(\"Directory:\", dir, sep=\"/n\"))\n",
    "        print(paste(\"Files:\", files, sep='/n'))\n",
    "    }\n",
    "    \n",
    "    raw_counts <- tximport(files, type='salmon', txOut = TRUE) #should this be true? transcript level data?\n",
    "    if (all(rownames(counts_metadata)==colnames(raw_counts)) == FALSE){\n",
    "        print(\"ERROR MISMATCH IN METADATA AND COUNTS\")\n",
    "        print(paste(\"Metadata rows:\", rownames(counts_metadata, sep='/n')))\n",
    "        print(paste(\"Counts columns:\", colnames(raw_counts), sep=\"\\n\"))\n",
    "    }\n",
    "    \n",
    "    rownames(counts_metadata) <- colnames(raw_counts$counts)\n",
    "    print(colnames(raw_counts$counts))\n",
    "    raw_counts\n",
    "}\n",
    "\n",
    "counts_metadata4 <- data.frame(\n",
    "    row.names = c(\"4add_backA\",\n",
    "         \"4add_backB\",\n",
    "         \"4pFe21_9A\",\n",
    "         \"4pFe21_9B\",\n",
    "         \"4pFe21_9C\"),\n",
    "    id=c(\"4add_backA\",\n",
    "         \"4add_backB\",\n",
    "         \"4pFe21_9A\",\n",
    "         \"4pFe21_9B\",\n",
    "         \"4pFe21_9C\"),\n",
    "    isolate=rep_len(4,5),\n",
    "    treatment=c(rep_len(\"Iron_ammendment\",2),rep_len(\"Low_Iron\",3)),\n",
    "    replicate=c('A','B','A','B','C'))\n",
    "\n",
    "counts_metadata8 <- data.frame(\n",
    "    row.names = c(\"8add_backB\",\n",
    "         \"8add_backC\",\n",
    "         \"8pFe21_9A\",\n",
    "         \"8pFe19_A\",\n",
    "         \"8pFe19_B\",\n",
    "         \"8pFe19_C\",\n",
    "         \"8pFe21_9B\",\n",
    "         \"8pFe21_9C\"),\n",
    "    id=c(\"8add_backB\",\n",
    "         \"8add_backC\",\n",
    "         \"8pFe21_9A\",\n",
    "         \"8pFe19_A\",\n",
    "         \"8pFe19_B\",\n",
    "         \"8pFe19_C\",\n",
    "         \"8pFe21_9B\",\n",
    "         \"8pFe21_9C\"),\n",
    "    isolate=rep_len(8,8),\n",
    "    treatment=c(rep_len(\"Iron_ammendment\",2), \"Low_Iron\", rep_len(\"High_Iron\",3),rep_len(\"Low_Iron\",2)),\n",
    "    replicate=c('B','C','A','A','B','C','B','C'))\n",
    "\n",
    "counts_metadata6 <- data.frame(\n",
    "    row.names = c(\n",
    "        \"6add_backA\",\n",
    "         \"6add_backB\",\n",
    "         \"6add_backC\",\n",
    "         \"6pFe21_9A\",\n",
    "         \"6pFe21_9B\",\n",
    "         \"6pFe21_9C\",\n",
    "         \"6pFe19_A\", \n",
    "         \"6pFe19_B\", \n",
    "         \"6pFe19_C\"),\n",
    "    id=c(\n",
    "        \"6add_backA\",\n",
    "         \"6add_backB\",\n",
    "         \"6add_backC\",\n",
    "         \"6pFe21_9A\",\n",
    "         \"6pFe21_9B\",\n",
    "         \"6pFe21_9C\",\n",
    "         \"6pFe19_A\", \n",
    "         \"6pFe19_B\", \n",
    "         \"6pFe19_C\"),\n",
    "    isolate=rep_len(6,9),\n",
    "    treatment=c(rep_len(\"Iron_ammendment\",3),rep_len(\"Low_Iron\",3), rep_len(\"High_Iron\",3)),\n",
    "    replicate=c(rep(c('A','B','C'),3)))\n",
    "\n",
    "counts_metadata13 <- data.frame(\n",
    "    row.names = c(\"13add_backA\",\n",
    "         \"13add_backB\",\n",
    "         \"13add_backC\",\n",
    "         \"13pFe21_9A\",\n",
    "         \"13pFe21_9B\",\n",
    "         \"13pFe21_9C\",\n",
    "        \"13pFe19_A\", \n",
    "         \"13pFe19_B\",\n",
    "         \"13pFe19_C\"),\n",
    "    id=c(\"13add_backA\",\n",
    "         \"13add_backB\",\n",
    "         \"13add_backC\",\n",
    "         \"13pFe21_9A\",\n",
    "         \"13pFe21_9B\",\n",
    "         \"13pFe21_9C\",\n",
    "        \"13pFe19_A\", \n",
    "         \"13pFe19_B\",\n",
    "         \"13pFe19_C\"),\n",
    "    isolate=rep_len(13,9),\n",
    "    treatment=c(rep_len(\"Iron_ammendment\",3),rep_len(\"Low_Iron\",3), rep_len(\"High_Iron\",3)),\n",
    "    replicate=c(rep(c('A','B','C'),3)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eb40913",
   "metadata": {},
   "source": [
    "# 2. Create DeSeq object\n",
    "---------\n",
    "## Create deseq2 object normalize the counts object for plotting later, and run the differential expression\n",
    "We should have four kinds of objects for each sample now, dds (deseq object), normalized dds, differential expression analysis object (de), and the differential expression results (res)\n",
    "## 2.1 create dds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "21c648de-7e0c-477f-93ad-70b5d3a52b4c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "using counts and average transcript lengths from tximport\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 40907\n",
      "[1] 27159\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "using counts and average transcript lengths from tximport\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 37958\n",
      "[1] 23714\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "using counts and average transcript lengths from tximport\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 53342\n",
      "[1] 35097\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "using counts and average transcript lengths from tximport\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 48089\n",
      "[1] 29014\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "List of length 2\n",
       "names(2): counts avgTxLength"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dds <- function(raw_counts, metadata, n){\n",
    "    dds <- DESeqDataSetFromTximport( raw_counts,\n",
    "                             colData=metadata,\n",
    "                             design=~treatment)\n",
    "    dds$treatment <- relevel(dds$treatment, ref = \"Low_Iron\")\n",
    "    keep <- rowSums(counts(dds) >=5) >= n #filter out rows with too low expression\n",
    "    print(nrow(dds))\n",
    "    dds <- dds[keep, ]\n",
    "    print(nrow(dds))\n",
    "    dds\n",
    "}\n",
    "\n",
    "dds4 <- dds(counts_4, metadata_4, 2)\n",
    "dds8 <- dds(counts_8, metadata_8, 2)\n",
    "dds6 <- dds(counts_6, metadata_6, 3)\n",
    "dds13 <- dds(counts_13, metadata_13, 3)\n",
    "head(assays(dds13))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8c311708-d821-40ba-94a2-bee71a3919c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "DataFrame with 9 rows and 4 columns\n",
       "                   id     isolate treatment         rep\n",
       "          <character> <character>  <factor> <character>\n",
       "add_backA   add_backA          13 Add_Back            A\n",
       "add_backB   add_backB          13 Add_Back            B\n",
       "add_backC   add_backC          13 Add_Back            C\n",
       "pFe19A         pFe19A          13 High_Iron           A\n",
       "pFe19B         pFe19B          13 High_Iron           B\n",
       "pFe19C         pFe19C          13 High_Iron           C\n",
       "pFe21_9A     pFe21_9A          13 Low_Iron            A\n",
       "pFe21_9B     pFe21_9B          13 Low_Iron            B\n",
       "pFe21_9C     pFe21_9C          13 Low_Iron            C"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "vsd.norm <- function(dds){\n",
    "    vst(dds,blind=FALSE)\n",
    "    }\n",
    "\n",
    "#full vsd deseq2 objects:\n",
    "vsd4 <- vsd.norm(dds4)\n",
    "vsd8 <- vsd.norm(dds8)\n",
    "vsd6 <- vsd.norm(dds6)\n",
    "vsd13 <- vsd.norm(dds13)\n",
    "colData(vsd13)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0e3b6ef-43fb-44f9-91c4-de42e2b3d142",
   "metadata": {},
   "source": [
    "## Below are flavodoxin genes to look for with kegg annotations to make sure they are not lost later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "12e89890-7158-42b6-8936-241f59c39f25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>orfs</th><th scope=col>add_backA</th><th scope=col>add_backB</th><th scope=col>add_backC</th><th scope=col>pFe19A</th><th scope=col>pFe19B</th><th scope=col>pFe19C</th><th scope=col>pFe21_9A</th><th scope=col>pFe21_9B</th><th scope=col>pFe21_9C</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>NODE_19011_length_1508_cov_60.425087_g9146_i1.p1</td><td>42</td><td>60</td><td>3</td><td>116</td><td>103</td><td>64</td><td>38</td><td>35</td><td>22</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " orfs & add\\_backA & add\\_backB & add\\_backC & pFe19A & pFe19B & pFe19C & pFe21\\_9A & pFe21\\_9B & pFe21\\_9C\\\\\n",
       " <chr> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t NODE\\_19011\\_length\\_1508\\_cov\\_60.425087\\_g9146\\_i1.p1 & 42 & 60 & 3 & 116 & 103 & 64 & 38 & 35 & 22\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 10\n",
       "\n",
       "| orfs &lt;chr&gt; | add_backA &lt;int&gt; | add_backB &lt;int&gt; | add_backC &lt;int&gt; | pFe19A &lt;int&gt; | pFe19B &lt;int&gt; | pFe19C &lt;int&gt; | pFe21_9A &lt;int&gt; | pFe21_9B &lt;int&gt; | pFe21_9C &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| NODE_19011_length_1508_cov_60.425087_g9146_i1.p1 | 42 | 60 | 3 | 116 | 103 | 64 | 38 | 35 | 22 |\n",
       "\n"
      ],
      "text/plain": [
       "  orfs                                             add_backA add_backB\n",
       "1 NODE_19011_length_1508_cov_60.425087_g9146_i1.p1 42        60       \n",
       "  add_backC pFe19A pFe19B pFe19C pFe21_9A pFe21_9B pFe21_9C\n",
       "1 3         116    103    64     38       35       22      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>orfs</th><th scope=col>add_backA</th><th scope=col>add_backB</th><th scope=col>add_backC</th><th scope=col>pFe19A</th><th scope=col>pFe19B</th><th scope=col>pFe19C</th><th scope=col>pFe21_9A</th><th scope=col>pFe21_9B</th><th scope=col>pFe21_9C</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>NODE_32738_length_942_cov_54.683544_g18313_i0.p1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>102</td><td>121</td><td>97</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " orfs & add\\_backA & add\\_backB & add\\_backC & pFe19A & pFe19B & pFe19C & pFe21\\_9A & pFe21\\_9B & pFe21\\_9C\\\\\n",
       " <chr> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t NODE\\_32738\\_length\\_942\\_cov\\_54.683544\\_g18313\\_i0.p1 & 0 & 0 & 0 & 0 & 0 & 0 & 102 & 121 & 97\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 10\n",
       "\n",
       "| orfs &lt;chr&gt; | add_backA &lt;int&gt; | add_backB &lt;int&gt; | add_backC &lt;int&gt; | pFe19A &lt;int&gt; | pFe19B &lt;int&gt; | pFe19C &lt;int&gt; | pFe21_9A &lt;int&gt; | pFe21_9B &lt;int&gt; | pFe21_9C &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| NODE_32738_length_942_cov_54.683544_g18313_i0.p1 | 0 | 0 | 0 | 0 | 0 | 0 | 102 | 121 | 97 |\n",
       "\n"
      ],
      "text/plain": [
       "  orfs                                             add_backA add_backB\n",
       "1 NODE_32738_length_942_cov_54.683544_g18313_i0.p1 0         0        \n",
       "  add_backC pFe19A pFe19B pFe19C pFe21_9A pFe21_9B pFe21_9C\n",
       "1 0         0      0      0      102      121      97      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>orfs</th><th scope=col>add_backA</th><th scope=col>add_backB</th><th scope=col>add_backC</th><th scope=col>pFe19A</th><th scope=col>pFe19B</th><th scope=col>pFe19C</th><th scope=col>pFe21_9A</th><th scope=col>pFe21_9B</th><th scope=col>pFe21_9C</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>NODE_34623_length_873_cov_1933.596250_g19622_i0.p1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>3619</td><td>3036</td><td>2551</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " orfs & add\\_backA & add\\_backB & add\\_backC & pFe19A & pFe19B & pFe19C & pFe21\\_9A & pFe21\\_9B & pFe21\\_9C\\\\\n",
       " <chr> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t NODE\\_34623\\_length\\_873\\_cov\\_1933.596250\\_g19622\\_i0.p1 & 0 & 0 & 0 & 0 & 0 & 0 & 3619 & 3036 & 2551\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 10\n",
       "\n",
       "| orfs &lt;chr&gt; | add_backA &lt;int&gt; | add_backB &lt;int&gt; | add_backC &lt;int&gt; | pFe19A &lt;int&gt; | pFe19B &lt;int&gt; | pFe19C &lt;int&gt; | pFe21_9A &lt;int&gt; | pFe21_9B &lt;int&gt; | pFe21_9C &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| NODE_34623_length_873_cov_1933.596250_g19622_i0.p1 | 0 | 0 | 0 | 0 | 0 | 0 | 3619 | 3036 | 2551 |\n",
       "\n"
      ],
      "text/plain": [
       "  orfs                                               add_backA add_backB\n",
       "1 NODE_34623_length_873_cov_1933.596250_g19622_i0.p1 0         0        \n",
       "  add_backC pFe19A pFe19B pFe19C pFe21_9A pFe21_9B pFe21_9C\n",
       "1 0         0      0      0      3619     3036     2551    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#isip2a:\n",
    "\n",
    "d13=as.data.frame(assay(dds13)) %>% rownames_to_column('orfs')\n",
    "\n",
    "filter(d13, orfs=='NODE_19011_length_1508_cov_60.425087_g9146_i1.p1') \n",
    "filter(d13, orfs=='NODE_32738_length_942_cov_54.683544_g18313_i0.p1')\n",
    "filter(d13, orfs=='NODE_34623_length_873_cov_1933.596250_g19622_i0.p1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "63870167",
   "metadata": {},
   "outputs": [],
   "source": [
    "ko4_def <- read.csv(\"ko4_def.csv\")\n",
    "ko8_def <- read.csv(\"ko8_def.csv\")\n",
    "ko6_def <- read.csv(\"ko6_def.csv\")\n",
    "ko13_def <- read.csv(\"ko13_def.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1291043e-861e-4bcd-8c74-e22ecccb3c98",
   "metadata": {},
   "source": [
    "# possible place for ko error where important genes dissapear"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "75b9676d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                              orfs ko_iteration  ko_id\n",
      "1 NODE_10051_length_2055_cov_12.485368_g5395_i0.p1        ko:_1 K02014\n",
      "2 NODE_10060_length_2054_cov_21.521454_g5168_i2.p1        ko:_1 K14683\n",
      "3 NODE_10077_length_2052_cov_55.663466_g5412_i0.p1        ko:_1 K03320\n",
      "4 NODE_10077_length_2052_cov_55.663466_g5412_i0.p1        ko:_2 K07573\n",
      "5 NODE_10189_length_2039_cov_48.328077_g5484_i0.p1        ko:_1 K02575\n",
      "6 NODE_10218_length_2035_cov_11.646279_g5502_i0.p2        ko:_1 K21867\n",
      "                   symbol\n",
      "1               TC.FEV.OM\n",
      "2       SLC34A, NPT, nptA\n",
      "3           amt, AMT, MEP\n",
      "4            CSL4, EXOSC1\n",
      "5  NRT2, narK, nrtP, nasA\n",
      "6    AKT, KAT, GORK, SKOR\n",
      "                                                                  name\n",
      "1                          iron complex outermembrane recepter protein\n",
      "2  solute carrier family 34 (sodium-dependent phosphate cotransporter)\n",
      "3                                     ammonium transporter, Amt family\n",
      "4                                       exosome complex component CSL4\n",
      "5             MFS transporter, NNP family, nitrate/nitrite transporter\n",
      "6                                                    potassium channel\n",
      "DataFrame with 6 rows and 3 columns\n",
      "                                                          ko_id\n",
      "                                                    <character>\n",
      "NODE_10000_length_2060_cov_11.205335_g5364_i0.p1         K02014\n",
      "NODE_10001_length_2060_cov_5.581278_g5365_i0.p1          K14683\n",
      "NODE_10002_length_2059_cov_37.920947_g5366_i0.p1 K03320; K07573\n",
      "NODE_10002_length_2059_cov_37.920947_g5366_i0.p2         K02575\n",
      "NODE_10003_length_2059_cov_28.617321_g5366_i1.p1         K21867\n",
      "NODE_10004_length_2059_cov_25.532729_g4469_i1.p1         K02575\n",
      "                                                                 symbol\n",
      "                                                            <character>\n",
      "NODE_10000_length_2060_cov_11.205335_g5364_i0.p1              TC.FEV.OM\n",
      "NODE_10001_length_2060_cov_5.581278_g5365_i0.p1       SLC34A, NPT, nptA\n",
      "NODE_10002_length_2059_cov_37.920947_g5366_i0.p1  amt, AMT, MEP;  CSL..\n",
      "NODE_10002_length_2059_cov_37.920947_g5366_i0.p2  NRT2, narK, nrtP, n..\n",
      "NODE_10003_length_2059_cov_28.617321_g5366_i1.p1   AKT, KAT, GORK, SKOR\n",
      "NODE_10004_length_2059_cov_25.532729_g4469_i1.p1  NRT2, narK, nrtP, n..\n",
      "                                                                   name\n",
      "                                                            <character>\n",
      "NODE_10000_length_2060_cov_11.205335_g5364_i0.p1  iron complex outerm..\n",
      "NODE_10001_length_2060_cov_5.581278_g5365_i0.p1   solute carrier fami..\n",
      "NODE_10002_length_2059_cov_37.920947_g5366_i0.p1  ammonium transporte..\n",
      "NODE_10002_length_2059_cov_37.920947_g5366_i0.p2  MFS transporter, NN..\n",
      "NODE_10003_length_2059_cov_28.617321_g5366_i1.p1      potassium channel\n",
      "NODE_10004_length_2059_cov_25.532729_g4469_i1.p1  MFS transporter, NN..\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in .local(x, ..., value = value): 6 rows in value to replace 27159rows\n",
     "output_type": "error",
     "traceback": [
      "Error in .local(x, ..., value = value): 6 rows in value to replace 27159rows\nTraceback:\n",
      "1. `mcols<-`(`*tmp*`, value = new(\"DFrame\", rownames = c(\"NODE_10000_length_2060_cov_11.205335_g5364_i0.p1\", \n . \"NODE_10001_length_2060_cov_5.581278_g5365_i0.p1\", \"NODE_10002_length_2059_cov_37.920947_g5366_i0.p1\", \n . \"NODE_10002_length_2059_cov_37.920947_g5366_i0.p2\", \"NODE_10003_length_2059_cov_28.617321_g5366_i1.p1\", \n . \"NODE_10004_length_2059_cov_25.532729_g4469_i1.p1\"), nrows = 6L, \n .     elementType = \"ANY\", elementMetadata = new(\"DFrame\", rownames = NULL, \n .         nrows = 3L, elementType = \"ANY\", elementMetadata = NULL, \n .         metadata = list(), listData = list(type = c(NA, NA, NA\n .         ), description = c(NA, NA, NA))), metadata = list(), \n .     listData = list(ko_id = c(\"K02014\", \"K14683\", \"K03320; K07573\", \n .     \"K02575\", \"K21867\", \"K02575\"), symbol = c(\" TC.FEV.OM\", \" SLC34A, NPT, nptA\", \n .     \" amt, AMT, MEP;  CSL4, EXOSC1\", \" NRT2, narK, nrtP, nasA\", \n .     \" AKT, KAT, GORK, SKOR\", \" NRT2, narK, nrtP, nasA\"), name = c(\" iron complex outermembrane recepter protein\", \n .     \" solute carrier family 34 (sodium-dependent phosphate cotransporter)\", \n .     \" ammonium transporter, Amt family;  exosome complex component CSL4\", \n .     \" MFS transporter, NNP family, nitrate/nitrite transporter\", \n .     \" potassium channel\", \" MFS transporter, NNP family, nitrate/nitrite transporter\"\n .     ))))",
      "2. `mcols<-`(`*tmp*`, value = new(\"DFrame\", rownames = c(\"NODE_10000_length_2060_cov_11.205335_g5364_i0.p1\", \n . \"NODE_10001_length_2060_cov_5.581278_g5365_i0.p1\", \"NODE_10002_length_2059_cov_37.920947_g5366_i0.p1\", \n . \"NODE_10002_length_2059_cov_37.920947_g5366_i0.p2\", \"NODE_10003_length_2059_cov_28.617321_g5366_i1.p1\", \n . \"NODE_10004_length_2059_cov_25.532729_g4469_i1.p1\"), nrows = 6L, \n .     elementType = \"ANY\", elementMetadata = new(\"DFrame\", rownames = NULL, \n .         nrows = 3L, elementType = \"ANY\", elementMetadata = NULL, \n .         metadata = list(), listData = list(type = c(NA, NA, NA\n .         ), description = c(NA, NA, NA))), metadata = list(), \n .     listData = list(ko_id = c(\"K02014\", \"K14683\", \"K03320; K07573\", \n .     \"K02575\", \"K21867\", \"K02575\"), symbol = c(\" TC.FEV.OM\", \" SLC34A, NPT, nptA\", \n .     \" amt, AMT, MEP;  CSL4, EXOSC1\", \" NRT2, narK, nrtP, nasA\", \n .     \" AKT, KAT, GORK, SKOR\", \" NRT2, narK, nrtP, nasA\"), name = c(\" iron complex outermembrane recepter protein\", \n .     \" solute carrier family 34 (sodium-dependent phosphate cotransporter)\", \n .     \" ammonium transporter, Amt family;  exosome complex component CSL4\", \n .     \" MFS transporter, NNP family, nitrate/nitrite transporter\", \n .     \" potassium channel\", \" MFS transporter, NNP family, nitrate/nitrite transporter\"\n .     ))))",
      "3. BiocGenerics:::replaceSlots(x, rowRanges = local({\n .     r <- rowRanges(x)\n .     mcols(r) <- value\n .     r\n . }), check = FALSE)",
      "4. unsafe_replaceSlots(object, ...)",
      "5. (function() {\n .     r <- rowRanges(x)\n .     mcols(r) <- value\n .     r\n . })()",
      "6. `mcols<-`(`*tmp*`, value = value)",
      "7. `mcols<-`(`*tmp*`, value = value)",
      "8. `elementMetadata<-`(x, ..., value = value)",
      "9. `elementMetadata<-`(x, ..., value = value)",
      "10. .local(x, ..., value = value)",
      "11. stop(k, \" rows in value to replace \", n, \"rows\")"
     ]
    }
   ],
   "source": [
    "\n",
    "#add kegg annotation as metadata to rows\n",
    "add.rowData <- function(dds, ko_def){\n",
    "    salmon.orfs <- rownames(dds)\n",
    "    head(salmon.orfs)\n",
    "    ko <- filter(ko_def, (ko_def$orfs %in% salmon.orfs)==TRUE)\n",
    "\n",
    "        #remove enzymes when shown\n",
    "    ko$name <- gsub(\" \\\\[EC.*\\\\]\", \"\", ko$name)\n",
    "    print(head(ko))\n",
    "        # take ko's defined by name and symbol and collaps by orf's so each orf is shown once\n",
    "    ko.sum <- ko %>% \n",
    "                     group_by(orfs)%>% \n",
    "                     summarize('ko_id'=str_flatten(ko_id, collapse='; '),\n",
    "                              'symbol'=str_flatten(symbol, collaps=\"; \"), \n",
    "                                'name'= str_flatten(name, collaps=\"; \"))\n",
    "\n",
    "        # create df of orf's not included in kegg annotation table\n",
    "    ko.na <- data.frame('orfs'=salmon.orfs[(salmon.orfs %in% ko.sum$orfs)==FALSE], 'ko_id'=NA, 'symbol'=NA, 'name'=NA)\n",
    "    full.anno <- rbind(ko.sum, ko.na ) %>% select(!orfs)\n",
    "    mcols(dds) <- cbind(mcols(dds), full.anno)\n",
    "    print(head(mcols(dds)))\n",
    "}\n",
    "\n",
    "\n",
    "mcols(dds4) <- add.rowData(dds4, ko4_def)\n",
    "mcols(dds8) <- add.rowData(dds8, ko8_def)\n",
    "mcols(dds6) <- add.rowData(dds6, ko6_def)\n",
    "mcols(dds13) <- add.rowData(dds13, ko13_def)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d2332c3",
   "metadata": {},
   "source": [
    "# running process for orfs with annotations only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bec803fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#add kegg annotation as metadata to rows\n",
    "select.ko <- function(dds, ko_def){\n",
    "    salmon.orfs <- rownames(dds)\n",
    "    ko <- filter(ko_def, (ko_def$orfs %in% salmon.orfs)==TRUE)\n",
    "\n",
    "        #remove enzymes when shown\n",
    "    ko$name <- gsub(\" \\\\[EC.*\\\\]\", \"\", ko$name)\n",
    "    \n",
    "        # take ko's defined by name and symbol and collaps by orf's so each orf is shown once\n",
    "    ko.sum <- ko %>% \n",
    "                     group_by(orfs)%>% \n",
    "                     summarize('ko_id'=str_flatten(ko_id, collapse='; '),\n",
    "                              'symbol'=str_flatten(symbol, collaps=\"; \"), \n",
    "                                'name'= str_flatten(name, collaps=\"; \"))\n",
    "    #dds.ko <- dds[(salmon.orfs %in% ko.sum$orfs)==TRUE, ]\n",
    "    #ko.data <- ko.sum %>% select(!orfs)\n",
    "    #mcols(dds.ko) <- cbind(mcols(dds.ko), ko.data)\n",
    "    #dds.ko\n",
    "}\n",
    "\n",
    "ko.sum4 <- select.ko(dds4, ko4_def)\n",
    "dds4.ko <- select.ko(dds4, ko4_def)\n",
    "dds8.ko <- select.ko(dds8, ko8_def)\n",
    "dds6.ko <- select.ko(dds6, ko6_def)\n",
    "dds13.ko <- select.ko(dds13, ko13_def)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f976821",
   "metadata": {},
   "source": [
    "### 2.2 VSD Normalize counts \n",
    "---\n",
    "Order rows by variance across treatment. Top rows will have highest variance in normalized counts between treatments. Save in vsd folder. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "744d73cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"saving 04\"\n",
      "[1] \"saving 08\"\n",
      "[1] \"saving 06\"\n",
      "[1] \"saving 13\"\n"
     ]
    }
   ],
   "source": [
    "vsd.norm <- function(dds){\n",
    "    vst(dds,blind=FALSE)\n",
    "    }\n",
    "\n",
    "#full vsd deseq2 objects:\n",
    "vsd4 <- vsd.norm(dds4)\n",
    "vsd8 <- vsd.norm(dds8)\n",
    "vsd6 <- vsd.norm(dds6)\n",
    "vsd13 <- vsd.norm(dds13)\n",
    "\n",
    "## order the df's by decreasing variance. top rows have highest varience between \n",
    "## samples. write dataframe\n",
    "write.vsd <- function(vsd, org){\n",
    "    vsd <- assay(vsd)\n",
    "    vsd_order <- order(rowVars(vsd), decreasing=T)\n",
    "    vsd_new <- vsd[vsd_order, ]\n",
    "    print(paste('saving',org,sep=' '))\n",
    "    vsd_new <- as.data.frame(vsd_new) %>% rownames_to_column(\"orfs\")\n",
    "    write.csv(vsd_new, paste('./vsd_files/', org, \"vsd.csv\", sep=\"\"), row.names=FALSE)\n",
    "}\n",
    "\n",
    "write.vsd(vsd4, \"04\")\n",
    "write.vsd(vsd8, \"08\")\n",
    "write.vsd(vsd6, \"06\")\n",
    "write.vsd(vsd13, \"13\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "4e888fcf",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in pheatmap(vsd_dist_mat, clustering_distance_rows = vsd_dist, clustering_distance_cols = vsd_dist, : could not find function \"pheatmap\"\n",
     "output_type": "error",
     "traceback": [
      "Error in pheatmap(vsd_dist_mat, clustering_distance_rows = vsd_dist, clustering_distance_cols = vsd_dist, : could not find function \"pheatmap\"\nTraceback:\n",
      "1. vsd.dist.mat(vsd4)"
     ]
    }
   ],
   "source": [
    "#distance matrix \n",
    "vsd.dist.mat <- function(vsd){\n",
    "    vsd_dist <- dist(t(assay(vsd)))\n",
    "    vsd_dist_mat <- as.matrix(vsd_dist)\n",
    "    rownames(vsd_dist_mat) <- paste(vsd$id)\n",
    "    pheatmap(vsd_dist_mat,\n",
    "            clustering_distance_rows = vsd_dist,\n",
    "            clustering_distance_cols = vsd_dist,\n",
    "            col = colors)\n",
    "    }\n",
    "\n",
    "dist_mat4 <- vsd.dist.mat(vsd4)\n",
    "dist_mat8 <- vsd.dist.mat(vsd8)\n",
    "dist_mat6 <- vsd.dist.mat(vsd6)\n",
    "dist_mat13 <- vsd.dist.mat(vsd13)\n",
    "\n",
    "grid.arrange(grobs=list(\n",
    "    dist_mat4[[4]],\n",
    "    dist_mat8[[4]], \n",
    "    dist_mat6[[4]], \n",
    "    dist_mat13[[4]]),\n",
    "    ncol=2, nrow=2, common.legend=TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5b3e46f",
   "metadata": {},
   "source": [
    "## 2.3 Differential Expression from DESeq\n",
    "---\n",
    "Run the differential expression. \n",
    "\n",
    "Extract the results using res() and specifying the contrast, or which treatments we wish to compare.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b1905ec4-f496-41de-b9c4-854b83dcbc63",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "de4 <- DESeq(dds4)\n",
    "de8 <- DESeq(dds8)\n",
    "de6 <- DESeq(dds6)\n",
    "de13 <- DESeq(dds13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f311e5d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "de4 <- DESeq(dds4)\n",
    "de8 <- DESeq(dds8)\n",
    "de6 <- DESeq(dds6)\n",
    "de13 <- DESeq(dds13)\n",
    "\n",
    "#contrasts\n",
    "HvL <- c(\"treatment\", \"High_Iron\", \"Low_Iron\")\n",
    "AvL <- c(\"treatment\", \"Add_Back\", \"Low_Iron\")\n",
    "AvH <- c(\"treatment\", \"Add_Back\", \"High_Iron\")\n",
    "\n",
    "#results from Low Iron vs Iron ammendment\n",
    "AvL4 <- results(de4, contrast=AvL, tidy=TRUE)\n",
    "AvL8 <- results(de8, contrast=AvL, tidy=T)\n",
    "AvL6 <- results(de6, contrast=AvL, tidy=TRUE)\n",
    "AvL13 <- results(de13, contrast=AvL, tidy=TRUE)\n",
    "\n",
    "#results from High Iron vs Low Iron\n",
    "HvL8  <- results(de8, contrast=HvL, tidy=TRUE)\n",
    "HvL6  <- results(de6, contrast=HvL,  tidy=TRUE)\n",
    "HvL13  <- results(de13,contrast=HvL, tidy=TRUE)\n",
    "\n",
    "#results from High Iron vs Iron ammendment\n",
    "AvH8  <- results(de8, contrast=AvH, tidy=TRUE)\n",
    "AvH6  <- results(de6, contrast=AvH,  tidy=TRUE)\n",
    "AvH13  <- results(de13, contrast=AvH, tidy=TRUE)\n",
    "\n",
    "\n",
    "#sum(res$padj < 0.05, na.rm=T) #sum up how many adjusted p-values are significant\n",
    "#nrow(res)\n",
    "#nrow(emap) ##??????how is the rows of emap > res?? Is this because res only has diff expressed contigs? \n",
    "\n",
    "colnames(AvL4)[1]  <- \"orfs\"\n",
    "colnames(AvL8)[1]  <- \"orfs\"\n",
    "colnames(AvL6)[1]  <- \"orfs\"\n",
    "colnames(AvL13)[1]  <- \"orfs\"\n",
    "\n",
    "colnames(HvL8)[1]  <- \"orfs\"\n",
    "colnames(HvL6)[1]  <- \"orfs\"\n",
    "colnames(HvL13)[1]  <- \"orfs\"\n",
    "\n",
    "colnames(AvH8)[1]  <- \"orfs\"\n",
    "colnames(AvH6)[1]  <- \"orfs\"\n",
    "colnames(AvH13)[1]  <- \"orfs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7caacfa0-6a61-435d-a459-91293f39ef42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orfs</th><th scope=col>baseMean</th><th scope=col>log2FoldChange</th><th scope=col>lfcSE</th><th scope=col>stat</th><th scope=col>pvalue</th><th scope=col>padj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>NODE_10000_length_2060_cov_11.205335_g5364_i0.p1</td><td>17.815316</td><td> 0.4643168</td><td>0.9386188</td><td> 0.4946809</td><td>0.620825381</td><td>0.90208279</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>NODE_10001_length_2060_cov_5.581278_g5365_i0.p1 </td><td>15.834006</td><td> 0.1955311</td><td>1.1688008</td><td> 0.1672921</td><td>0.867140223</td><td>0.97272842</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>NODE_10002_length_2059_cov_37.920947_g5366_i0.p1</td><td>85.381955</td><td>-1.6125740</td><td>0.4985678</td><td>-3.2344127</td><td>0.001218932</td><td>0.03408861</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>NODE_10002_length_2059_cov_37.920947_g5366_i0.p2</td><td> 7.518903</td><td>-0.3956059</td><td>1.2030317</td><td>-0.3288408</td><td>0.742276024</td><td>        NA</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NODE_10003_length_2059_cov_28.617321_g5366_i1.p1</td><td>25.249582</td><td>-1.2007758</td><td>0.8319401</td><td>-1.4433440</td><td>0.148923602</td><td>0.55676320</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>NODE_10004_length_2059_cov_25.532729_g4469_i1.p1</td><td>21.393023</td><td> 1.4248841</td><td>0.6729055</td><td> 2.1175101</td><td>0.034216582</td><td>0.27807968</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & orfs & baseMean & log2FoldChange & lfcSE & stat & pvalue & padj\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & NODE\\_10000\\_length\\_2060\\_cov\\_11.205335\\_g5364\\_i0.p1 & 17.815316 &  0.4643168 & 0.9386188 &  0.4946809 & 0.620825381 & 0.90208279\\\\\n",
       "\t2 & NODE\\_10001\\_length\\_2060\\_cov\\_5.581278\\_g5365\\_i0.p1  & 15.834006 &  0.1955311 & 1.1688008 &  0.1672921 & 0.867140223 & 0.97272842\\\\\n",
       "\t3 & NODE\\_10002\\_length\\_2059\\_cov\\_37.920947\\_g5366\\_i0.p1 & 85.381955 & -1.6125740 & 0.4985678 & -3.2344127 & 0.001218932 & 0.03408861\\\\\n",
       "\t4 & NODE\\_10002\\_length\\_2059\\_cov\\_37.920947\\_g5366\\_i0.p2 &  7.518903 & -0.3956059 & 1.2030317 & -0.3288408 & 0.742276024 &         NA\\\\\n",
       "\t5 & NODE\\_10003\\_length\\_2059\\_cov\\_28.617321\\_g5366\\_i1.p1 & 25.249582 & -1.2007758 & 0.8319401 & -1.4433440 & 0.148923602 & 0.55676320\\\\\n",
       "\t6 & NODE\\_10004\\_length\\_2059\\_cov\\_25.532729\\_g4469\\_i1.p1 & 21.393023 &  1.4248841 & 0.6729055 &  2.1175101 & 0.034216582 & 0.27807968\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | orfs &lt;chr&gt; | baseMean &lt;dbl&gt; | log2FoldChange &lt;dbl&gt; | lfcSE &lt;dbl&gt; | stat &lt;dbl&gt; | pvalue &lt;dbl&gt; | padj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | NODE_10000_length_2060_cov_11.205335_g5364_i0.p1 | 17.815316 |  0.4643168 | 0.9386188 |  0.4946809 | 0.620825381 | 0.90208279 |\n",
       "| 2 | NODE_10001_length_2060_cov_5.581278_g5365_i0.p1  | 15.834006 |  0.1955311 | 1.1688008 |  0.1672921 | 0.867140223 | 0.97272842 |\n",
       "| 3 | NODE_10002_length_2059_cov_37.920947_g5366_i0.p1 | 85.381955 | -1.6125740 | 0.4985678 | -3.2344127 | 0.001218932 | 0.03408861 |\n",
       "| 4 | NODE_10002_length_2059_cov_37.920947_g5366_i0.p2 |  7.518903 | -0.3956059 | 1.2030317 | -0.3288408 | 0.742276024 |         NA |\n",
       "| 5 | NODE_10003_length_2059_cov_28.617321_g5366_i1.p1 | 25.249582 | -1.2007758 | 0.8319401 | -1.4433440 | 0.148923602 | 0.55676320 |\n",
       "| 6 | NODE_10004_length_2059_cov_25.532729_g4469_i1.p1 | 21.393023 |  1.4248841 | 0.6729055 |  2.1175101 | 0.034216582 | 0.27807968 |\n",
       "\n"
      ],
      "text/plain": [
       "  orfs                                             baseMean  log2FoldChange\n",
       "1 NODE_10000_length_2060_cov_11.205335_g5364_i0.p1 17.815316  0.4643168    \n",
       "2 NODE_10001_length_2060_cov_5.581278_g5365_i0.p1  15.834006  0.1955311    \n",
       "3 NODE_10002_length_2059_cov_37.920947_g5366_i0.p1 85.381955 -1.6125740    \n",
       "4 NODE_10002_length_2059_cov_37.920947_g5366_i0.p2  7.518903 -0.3956059    \n",
       "5 NODE_10003_length_2059_cov_28.617321_g5366_i1.p1 25.249582 -1.2007758    \n",
       "6 NODE_10004_length_2059_cov_25.532729_g4469_i1.p1 21.393023  1.4248841    \n",
       "  lfcSE     stat       pvalue      padj      \n",
       "1 0.9386188  0.4946809 0.620825381 0.90208279\n",
       "2 1.1688008  0.1672921 0.867140223 0.97272842\n",
       "3 0.4985678 -3.2344127 0.001218932 0.03408861\n",
       "4 1.2030317 -0.3288408 0.742276024         NA\n",
       "5 0.8319401 -1.4433440 0.148923602 0.55676320\n",
       "6 0.6729055  2.1175101 0.034216582 0.27807968"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(AvL4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "43162d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many DE's per treatment-sample?\n",
    "avl4.perc=(nrow(filter(AvL4,(padj < 0.05)==T))/nrow(de4))*100\n",
    "avl4=nrow(filter(AvL4,(padj < 0.05)==T))\n",
    "\n",
    "avl8.perc=(nrow(filter(AvL8,(padj < 0.05)==T))/nrow(de8))*100\n",
    "avl8=nrow(filter(AvL8,(padj < 0.05)==T))\n",
    "\n",
    "avl6.perc=(nrow(filter(AvL6,(padj < 0.05)==T))/nrow(de6))*100\n",
    "avl6=nrow(filter(AvL6,(padj < 0.05)==T))\n",
    "\n",
    "avl13.perc=(nrow(filter(AvL13,(padj < 0.05)==T))/nrow(de13))*100\n",
    "avl13=nrow(filter(AvL13,(padj < 0.05)==T))\n",
    "\n",
    "hvl8.perc=(nrow(filter(HvL8,(padj < 0.05)==T))/nrow(de8))*100\n",
    "hvl8=nrow(filter(HvL8,(padj < 0.05)==T))\n",
    "\n",
    "hvl6.perc=(nrow(filter(HvL6,(padj < 0.05)==T))/nrow(de6))*100\n",
    "hvl6=nrow(filter(HvL6,(padj < 0.05)==T))\n",
    "\n",
    "hvl13.perc=(nrow(filter(HvL13,(padj < 0.05)==T))/nrow(de13))*100\n",
    "hvl13=nrow(filter(HvL13,(padj < 0.05)==T))\n",
    "\n",
    "avh8.perc=(nrow(filter(AvH8,(padj < 0.05)==T))/nrow(de8))*100\n",
    "avh8=nrow(filter(AvH8,(padj < 0.05)==T))\n",
    "\n",
    "avh6.perc=(nrow(filter(AvH6,(padj < 0.05)==T))/nrow(de6))*100\n",
    "avh6=nrow(filter(AvH6,(padj < 0.05)==T))\n",
    "\n",
    "avh13.perc=(nrow(filter(AvH13,(padj < 0.05)==T))/nrow(de13))*100\n",
    "avh13=nrow(filter(AvH13,(padj < 0.05)==T))\n",
    "\n",
    "de_output = data.frame('Organism'=c('C. closterium 4', 'C. closterium 8', 'G. oceanica', 'G. huxleyi'),\n",
    "                       'Iron amendment vs low iron'=c(avl4,avl8,avl6,avl13),\n",
    "                       'High iron vs low iron'=c('NA', hvl8,hvl6,hvl13),\n",
    "                       'Iron amendment vs high iron'=c('NA',avh8,avh6,avh13))\n",
    "\n",
    "de_output_perc = data.frame('Organism'=c('C. closterium 4', 'C. closterium 8', 'G. oceanica', 'G. huxleyi'),\n",
    "                       'Iron amendment vs low iron'=c(avl4.perc,avl8.perc,avl6.perc,avl13.perc),\n",
    "                       'High iron vs low iron'=c('NA', hvl8.perc,hvl6.perc,hvl13.perc),\n",
    "                       'Iron amendment vs high iron'=c('NA',avh8.perc,avh6.perc,avh13.perc)) \n",
    "\n",
    "write.csv(de_output, './de_res_files/de_output.csv', row.names=F)\n",
    "write.csv(de_output_perc, './de_res_files/de_output_perc.csv', row.names=F)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "600371d8",
   "metadata": {},
   "source": [
    "## Logfold 2 shrinkage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "18018ff5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "lfc4.AvL <- lfcShrink(de4,  contrast=c(\"treatment\", \"Add_Back\", \"Low_Iron\"), type='ashr')\n",
    "lfc4.AvL <- lfc4.AvL %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc4.AvL, \"./de_res_files/lfc4.AvL.csv\", row.names=F)\n",
    "\n",
    "lfc8.AvL <- lfcShrink(de8, contrast=c(\"treatment\", \"Add_Back\", \"Low_Iron\"), type='ashr')\n",
    "lfc8.AvL <- lfc8.AvL %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc8.AvL, \"./de_res_files/lfc8.AvL.csv\", row.names=F)\n",
    "\n",
    "lfc6.AvL <- lfcShrink(de6, contrast=c(\"treatment\", \"Add_Back\", \"Low_Iron\"), type='ashr')\n",
    "lfc6.AvL <- lfc6.AvL %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc6.AvL, \"./de_res_files/lfc6.AvL.csv\", row.names=F)\n",
    "\n",
    "lfc13.AvL <- lfcShrink(de13,  contrast=c(\"treatment\", \"Add_Back\", \"Low_Iron\"), type='ashr')\n",
    "lfc13.AvL <- lfc13.AvL %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc13.AvL, \"./de_res_files/lfc13.AvL.csv\", row.names=F)\n",
    "\n",
    "lfc8.HvL <- lfcShrink(de8, contrast=c(\"treatment\", \"High_Iron\", \"Low_Iron\"), type='ashr')\n",
    "lfc8.HvL <- lfc8.HvL %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc8.HvL, \"./de_res_files/lfc8.HvL.csv\", row.names=F)\n",
    "\n",
    "lfc6.HvL <- lfcShrink(de6, contrast=c(\"treatment\", \"High_Iron\", \"Low_Iron\"), type='ashr')\n",
    "lfc6.HvL <- lfc6.HvL %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc6.HvL, \"./de_res_files/lfc6.HvL.csv\", row.names=F)\n",
    "\n",
    "lfc13.HvL <- lfcShrink(de13,  contrast=c(\"treatment\", \"High_Iron\", \"Low_Iron\"), type='ashr')\n",
    "lfc13.HvL <- lfc13.HvL %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc13.HvL, \"./de_res_files/lfc13.HvL.csv\", row.names=F)\n",
    "\n",
    "lfc8.AvH <- lfcShrink(de8, contrast=c(\"treatment\", \"Add_Back\", \"High_Iron\"), type='ashr')\n",
    "lfc8.AvH <- lfc8.AvH %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc8.AvH, \"./de_res_files/lfc8.AvH.csv\", row.names=F)\n",
    "\n",
    "lfc6.AvH <- lfcShrink(de6, contrast=c(\"treatment\", \"Add_Back\", \"High_Iron\"), type='ashr')\n",
    "lfc6.AvH <- lfc6.AvH %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc6.AvH, \"./de_res_files/lfc6.AvH.csv\", row.names=F)\n",
    "\n",
    "lfc13.AvH <- lfcShrink(de13,  contrast=c(\"treatment\", \"Add_Back\", \"High_Iron\"), type='ashr')\n",
    "lfc13.AvH <- lfc13.AvH %>% as.data.frame() %>% rownames_to_column(\"orfs\") \n",
    "write.csv(lfc13.AvH, \"./de_res_files/lfc13.AvH.csv\", row.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffdbed7d",
   "metadata": {},
   "source": [
    "## 2.4 Subset results tables by significance and save\n",
    "df's made into a list and looped through to order rows by adjusted p value and pull out significant rows. saving both files. Sorting the _res files positive or negative log2fold change will enable me to get the up regulated and down regulated genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "888d9e80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"contrast  HvL  for  4  not found\"\n",
      "[1] \"contrast  AvH  for  4  not found\"\n"
     ]
    }
   ],
   "source": [
    "df.names <- list()\n",
    "samples <- c('4','8','6','13')\n",
    "contrasts <- c('AvL','HvL','AvH')\n",
    "\n",
    "for (s in samples){\n",
    "    for (c in contrasts){\n",
    "        if (exists(paste(c,s, sep=\"\")) == FALSE){\n",
    "            print(paste('contrast ', c, ' for ', s, ' not found'))\n",
    "            next}\n",
    "        df.names <- append(df.names, paste(s,c,sep=''))\n",
    "    }\n",
    "}\n",
    "\n",
    "res.ls <- list(AvL4, AvL8, AvL6, AvL13, HvL8, HvL6, HvL13, AvH8, AvH6, AvH13)\n",
    "names(res.ls) <- df.names\n",
    "\n",
    "res.ls <- lapply(res.ls, function(df){   #order each df in list by p.value\n",
    "    arrange(df, padj)\n",
    "})\n",
    "\n",
    "#save each df ordered by p.value\n",
    "walk2(res.ls, paste0(\"./de_res_files/\", names(res.ls), \"_res.csv\", sep=\"\"), write.csv,row.names=F)\n",
    "\n",
    "res.ls.sig <- lapply(res.ls, function(df){   #pull out significant DE's\n",
    "    filter(df, padj<=0.05)\n",
    "})\n",
    "\n",
    "#save significant de's for each df\n",
    "walk2(res.ls.sig, paste0(\"./de_res_files/\", names(res.ls), \"sig_res.csv\", sep=\"\"), write.csv,row.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0c32042",
   "metadata": {},
   "source": [
    "# Glimma MA plots\n",
    "\n",
    "The Glimma MDS contains two main components:\n",
    "\n",
    "1. a plot showing two MDS dimensions, and\n",
    "2. a plot of the eigenvalues of each dimension\n",
    "\n",
    "The Glimma MA plot contains two main components:\n",
    "\n",
    "1. a plot of summary statistics across all genes that have been tested, and\n",
    "2. a plot of gene expression from individual samples for a given gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b84bdc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "glimmaMDS(dds4.ko)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dd7f1f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "de4.ko <- DESeq(dds4.ko)\n",
    "de8.ko <- DESeq(dds8.ko)\n",
    "de6.ko <- DESeq(dds6.ko)\n",
    "de13.ko <- DESeq(dds13.ko)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d2129dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ma.glimma <- function(de){\n",
    "    #plotting function removes orfs with a pvalue of na, its not updating the anno df properly, \n",
    "    #so here I calculated those which will be removed\n",
    "    #and make a df with the annotations already removed\n",
    "    r <- as.data.frame(results(de))\n",
    "    complete.genes <- (complete.cases(r))\n",
    "    res.comp <- r[complete.genes, ]\n",
    "    x <- de[complete.genes, ]\n",
    "   \n",
    "    glimmaMA(de,\n",
    "            #status=NULL,                                    # vector indicated status of each gene\n",
    "             groups= colData(de)$treatment,    #vector length of ncol dds, shows categories of samples \n",
    "             anno=mcols(x),                                         #df with nrow of de object with rows containing gene annotations\n",
    "             display.columns=c('symbol','name'),  #character vector with names from anno from which to display \n",
    "             main='MA Plot', \n",
    "             xlab='logCPM', \n",
    "             ylab='log2FC')\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe264a94",
   "metadata": {},
   "outputs": [],
   "source": [
    "ma.glimma(de4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "932260ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "ma.glimma(de8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffc190fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "ma.glimma(de6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "344bca4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ma.glimma(de13)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19a18b86-f770-4232-befd-4f36218a362a",
   "metadata": {},
   "source": [
    "# Make table of % kegg annotated orfs and other functional annotation info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "96f79e8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"% orfs annotated\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Organism</th><th scope=col>Anotated.orfs....</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>C. closterium 4</td><td>45.27044</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>C. closterium 8</td><td>46.33550</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>G. oceanica    </td><td>66.00849</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>G. huxleyi     </td><td>70.26608</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & Organism & Anotated.orfs....\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & C. closterium 4 & 45.27044\\\\\n",
       "\t2 & C. closterium 8 & 46.33550\\\\\n",
       "\t3 & G. oceanica     & 66.00849\\\\\n",
       "\t4 & G. huxleyi      & 70.26608\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 2\n",
       "\n",
       "| <!--/--> | Organism &lt;chr&gt; | Anotated.orfs.... &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 1 | C. closterium 4 | 45.27044 |\n",
       "| 2 | C. closterium 8 | 46.33550 |\n",
       "| 3 | G. oceanica     | 66.00849 |\n",
       "| 4 | G. huxleyi      | 70.26608 |\n",
       "\n"
      ],
      "text/plain": [
       "  Organism        Anotated.orfs....\n",
       "1 C. closterium 4 45.27044         \n",
       "2 C. closterium 8 46.33550         \n",
       "3 G. oceanica     66.00849         \n",
       "4 G. huxleyi      70.26608         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ko4_ls=read.csv('../kegg_names/ko4_ls.csv')\n",
    "ko8_ls=read.csv('../kegg_names/ko8_ls.csv')\n",
    "ko6_ls=read.csv('../kegg_names/ko6_ls.csv')\n",
    "ko13_ls=read.csv('../kegg_names/ko13_ls.csv')\n",
    "\n",
    "\n",
    "#make sure to REMOVE ORFS NOT COUNTED BY SALMON!\n",
    "ko_annotated=data.frame('Organism'=c('C. closterium 4', 'C. closterium 8', 'G. oceanica', 'G. huxleyi'),\n",
    "                        'Anotated orfs (%)'=c(\n",
    "                            (nrow(filter(ko4_ls, ko_iteration=='ko:_1'))/nrow(de4)*100),\n",
    "                            (nrow(filter(ko8_ls, ko_iteration=='ko:_1'))/nrow(de8)*100),\n",
    "                            (nrow(filter(ko6_ls, ko_iteration=='ko:_1'))/nrow(de6)*100),\n",
    "                            (nrow(filter(ko13_ls, ko_iteration=='ko:_1'))/nrow(de13)*100)))\n",
    "head(ko_annotated)\n",
    "write.csv(ko_annotated, './de_res_files/ko_anotat_de.csv', row.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d8645be",
   "metadata": {},
   "source": [
    "# Now repeat steps but summarize to 'kegg gene level'\n",
    "## 1. Make df of raw counts which are summed to kegg level.\n",
    "#### Read in the list of ORF-to-Ko's for each sample \n",
    "This list repeates orfs when multiple ko's were assigned to a single orf by eggnog. Using the 'orf' column from the organism specific ko list, we want to map the counts to each row of the orf-to-ko list. This will automatically repeat the counts for each repeated orf in the list and allow us to sum orfs with matching ko's later. \n",
    "\n",
    "<b/> Remember, not all rows from the counts table will have a ko assigned and thus will not appear in the orf-to-ko list. Additionally, not all orfs annotated by eggnog were counted by salmon. Thus we must remove any rows from the orf-to-ko table for which orfs are not found </b>\n",
    "\n",
    "Because some orfs had multiple ko assignments the merging relationship will be many-to-one, many orf-ko rows matching to one counts row. \"Each row in x (orf-to-ko list) matches at most 1 row in y (counts table).\"\n",
    "\n",
    "Once the two tables are merged, we can group by ko_id and sum counts which have the same ko_id. The result is a new raw counts table (matrix) which we can used in deseq. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b0702e78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \" # unique ko_ids should equal number rows of ko-summed counts.\"\n",
      "[1] \"# unique ko_ids = \"\n",
      "[1] 4243\n",
      "[1] \" and # ko-summed counts = \"\n",
      "[1] 4243\n",
      "[1] \" # unique ko_ids should equal number rows of ko-summed counts.\"\n",
      "[1] \"# unique ko_ids = \"\n",
      "[1] 4436\n",
      "[1] \" and # ko-summed counts = \"\n",
      "[1] 4436\n",
      "[1] \" # unique ko_ids should equal number rows of ko-summed counts.\"\n",
      "[1] \"# unique ko_ids = \"\n",
      "[1] 5374\n",
      "[1] \" and # ko-summed counts = \"\n",
      "[1] 5374\n",
      "[1] \" # unique ko_ids should equal number rows of ko-summed counts.\"\n",
      "[1] \"# unique ko_ids = \"\n",
      "[1] 5270\n",
      "[1] \" and # ko-summed counts = \"\n",
      "[1] 5270\n"
     ]
    }
   ],
   "source": [
    "ko.def=read.csv('../kegg_names/ko_def.csv')\n",
    "sum.kegg <- function(org, counts.raw){\n",
    "    ko_df= read.csv(paste('../kegg_names/ko', org,'_ls.csv', sep=''))\n",
    "    #ko_df = organsim-specific orf-to-ko list\n",
    "    ko_df <- select(ko_df, c('orfs', 'ko_id'))\n",
    "    \n",
    "    # make raw counts matrix into a tibble\n",
    "    counts <- as_tibble(counts.raw$counts, rownames = \"orfs\")\n",
    "    \n",
    "    # remove orfs that eggnog annotated but salmon did not count\n",
    "    ko <- filter(ko_df, (ko_df$orfs %in% counts$orfs)==TRUE)\n",
    "    \n",
    "    # remove orfs that did not have a matching kegg annotation\n",
    "    #counts.ko <- filter(counts, (counts$orfs%in%ko_df$orfs)==TRUE)\n",
    "\n",
    "    \n",
    "    # merge the two, so orfs and thier counts are repeated when they match to multiple ko's\n",
    "    b <- left_join(x=ko, y=counts, by='orfs', relationship=\"many-to-one\")\n",
    "    if(all(ko$orfs%in% counts$orfs)==FALSE){\n",
    "        print(paste(\"ERROR: merged counts and ko_ids should be the same length as \n",
    "                    the ko_id df length of merged counts is \", nrow(b), \" and ko_ids is: \", \n",
    "                    nrow(ko), sep=''))\n",
    "    }\n",
    "\n",
    "    #group by ko_id and sum counts for each ko_id\n",
    "    b <- b %>% select(!orfs) %>% \n",
    "        group_by(ko_id) %>% \n",
    "        summarize(across(everything(), sum)) %>%\n",
    "        column_to_rownames(\"ko_id\") %>% \n",
    "        as.matrix\n",
    "    \n",
    "    mode(b) <- 'integer'\n",
    "    print(' # unique ko_ids should equal number rows of ko-summed counts.')\n",
    "    print('# unique ko_ids = ')\n",
    "    print(length(unique(ko$ko_id)))\n",
    "    print(' and # ko-summed counts = ')\n",
    "    print(nrow(b))\n",
    "    b\n",
    "    }\n",
    "\n",
    "kcounts_4 <- sum.kegg('4', counts_4)\n",
    "kcounts_8 <- sum.kegg('8', counts_8)\n",
    "kcounts_6 <- sum.kegg('6', counts_6)\n",
    "kcounts_13 <- sum.kegg('13', counts_13)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae074186",
   "metadata": {},
   "source": [
    "## 2. Run DeSeq2 on kegg-summed counts\n",
    "Using DESeqDataSetFromMatrix, read in summed counts matrix.\n",
    "\n",
    "Next, add the ko name and symbol to the metadata for each deseq object. Run the DE and VST normalization "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e23ed91",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## create deseq object from matrix\n",
    "k.dds <- function(k.counts, metadata,n){\n",
    "    #n = lowest number of reps in any treatment\n",
    "   dds <-  DESeqDataSetFromMatrix(\n",
    "       countData = k.counts, \n",
    "       colData = metadata, \n",
    "       design=~treatment)\n",
    "    dds$treatment <- relevel(dds$treatment, ref = \"Low_Iron\")\n",
    "    keep <- rowSums(counts(dds) >=10) >= n #filter out rows with too low expression\n",
    "    dds <- dds[keep, ]\n",
    "    dds}\n",
    "\n",
    "k.dds4 <- k.dds(kcounts_4, metadata_4, 2)\n",
    "k.dds8 <- k.dds(kcounts_8, metadata_8, 2)\n",
    "k.dds6 <- k.dds(kcounts_6, metadata_6, 3)\n",
    "k.dds13 <- k.dds(kcounts_13, metadata_13, 3)\n",
    "\n",
    "add.rowData <- function(dds, ko_def){\n",
    "    d <- data.frame('ko_id'= rownames(dds))\n",
    "        #filter ko's which are present in the dds df\n",
    "    v <- ko_def[(ko_def$ko_id%in%d$ko_id)==TRUE, c('ko_id', 'symbol', 'name')]\n",
    "        #because counts are summed by ko_id, I need to take only the unique ko's from anno\n",
    "    anno <- distinct(v)\n",
    "    all(rownames(dds)==anno$ko_id)\n",
    "    #reorder annotation table to same row order as dds\n",
    "    anno <- anno[match(rownames(dds), anno$ko_id),]\n",
    "    all(rownames(dds)==anno$ko_id)\n",
    "\n",
    "    mcols(dds) <- cbind(mcols(dds), anno)\n",
    "    print(mcols(dds))\n",
    "}\n",
    "\n",
    "#mcols(k.dds4) <- add.rowData(k.dds4, ko4_def)\n",
    "#mcols(k.dds8) <- add.rowData(k.dds8, ko8_def)\n",
    "#mcols(k.dds6) <- add.rowData(k.dds6, ko6_def)\n",
    "#mcols(k.dds13) <- add.rowData(k.dds13, ko13_def)\n",
    "#mcols(k.dds13)\n",
    "    \n",
    "## vsd normalize and save ordered by variance\n",
    "vsd.norm <- function(dds){\n",
    "    varianceStabilizingTransformation(dds,blind=FALSE)\n",
    "    }\n",
    "k.vsd4 <- vsd.norm(k.dds4)\n",
    "k.vsd8 <- vsd.norm(k.dds8)\n",
    "k.vsd6 <- vsd.norm(k.dds6)\n",
    "k.vsd13 <- vsd.norm(k.dds13)\n",
    "#k.vsd4 <- varianceStabilizingTransformation(k.dds4,blind = FALSE)\n",
    "write.vsd.k <- function(vsd, org){\n",
    "    vsd <- assay(vsd)\n",
    "    vsd_order <- order(rowVars(vsd), decreasing=T)\n",
    "    vsd_new <- vsd[vsd_order, ]\n",
    "    write.csv(vsd_new, paste('./vsd_files/', org, \"vsd.k.csv\", sep=\"\"))\n",
    "}\n",
    "\n",
    "write.vsd.k(k.vsd4, '04')\n",
    "write.vsd.k(k.vsd8, '08')\n",
    "write.vsd.k(k.vsd6, '06')\n",
    "write.vsd.k(k.vsd13, '13')\n",
    "\n",
    "## run differential expression analysis\n",
    "de4.k <- DESeq(k.dds4)\n",
    "de8.k <- DESeq(k.dds8)\n",
    "de6.k <- DESeq(k.dds6)\n",
    "de13.k <- DESeq(k.dds13)\n",
    "\n",
    "#extract results\n",
    "HvL <- c(\"treatment\", \"High_Iron\", \"Low_Iron\")\n",
    "AvL <- c(\"treatment\", \"Add_Back\", \"Low_Iron\")\n",
    "AvH <- c(\"treatment\", \"Add_Back\", \"High_Iron\")\n",
    "\n",
    "#results from Low Iron vs Iron ammendment\n",
    "AvL4.k <- results(de4.k, contrast=AvL, tidy=TRUE)\n",
    "AvL8.k <- results(de8.k, contrast=AvL, tidy=T)\n",
    "AvL6.k <- results(de6.k, contrast=AvL, tidy=TRUE)\n",
    "AvL13.k <- results(de13.k, contrast=AvL, tidy=TRUE)\n",
    "\n",
    "#results from Low Iron vs Iron ammendment\n",
    "AvL4.k <- results(de4.k, contrast=AvL, tidy=TRUE)\n",
    "AvL8.k <- results(de8.k, contrast=AvL, tidy=T)\n",
    "AvL6.k <- results(de6.k, contrast=AvL, tidy=TRUE)\n",
    "AvL13.k <- results(de13.k, contrast=AvL, tidy=TRUE)\n",
    "\n",
    "#results from High Iron vs Low Iron\n",
    "HvL8.k  <- results(de8.k, contrast=HvL, tidy=TRUE)\n",
    "HvL6.k  <- results(de6.k, contrast=HvL,  tidy=TRUE)\n",
    "HvL13.k  <- results(de13.k,contrast=HvL, tidy=TRUE)\n",
    "\n",
    "#results from High Iron vs Iron ammendment\n",
    "AvH8.k  <- results(de8.k, contrast=AvH, tidy=TRUE)\n",
    "AvH6.k  <- results(de6.k, contrast=AvH,  tidy=TRUE)\n",
    "AvH13.k  <- results(de13.k, contrast=AvH, tidy=TRUE)\n",
    "\n",
    "#save deseq results\n",
    "\n",
    "AvL4.k %>% na.omit() %>% write.csv('./de_res_files/AvL4.k.csv', row.names=F)\n",
    "AvL8.k%>% na.omit() %>% write.csv('./de_res_files/AvL8.k.csv', row.names=F)\n",
    "AvL6.k %>% na.omit() %>% write.csv('./de_res_files/AvL6.k.csv', row.names=F)\n",
    "AvL13.k %>% na.omit() %>% write.csv('./de_res_files/AvL13.k.csv', row.names=F)\n",
    "\n",
    "AvH8.k %>% na.omit() %>% write.csv('./de_res_files/AvH8.k.csv', row.names=F)\n",
    "AvH6.k %>% na.omit() %>% write.csv('./de_res_files/AvH6.k.csv', row.names=F)\n",
    "AvH13.k %>% na.omit() %>% write.csv('./de_res_files/AvH13.k.csv', row.names=F)\n",
    "\n",
    "HvL8.k %>% na.omit() %>% write.csv('./de_res_files/HvL8.k.csv', row.names=F)\n",
    "HvL6.k %>% na.omit() %>% write.csv('./de_res_files/HvL6.k.csv', row.names=F)\n",
    "HvL13.k %>% na.omit() %>% write.csv('./de_res_files/HvL13.k.csv', row.names=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7a4f0e5b-5b6c-4fcc-8144-9edb9fc01d6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 8 rows and 5 columns\n",
       "                   id     isolate treatment         rep sizeFactor\n",
       "          <character> <character>  <factor> <character>  <numeric>\n",
       "add_backB   add_backB          08 Add_Back            B   2.045848\n",
       "add_backC   add_backC          08 Add_Back            C   1.056302\n",
       "pFe21_9A     pFe21_9A          08 Low_Iron            A   0.599845\n",
       "pFe19A         pFe19A          08 High_Iron           A   0.713617\n",
       "pFe19B         pFe19B          08 High_Iron           B   0.592784\n",
       "pFe19C         pFe19C          08 High_Iron           C   1.221984\n",
       "pFe21_9B     pFe21_9B          08 Low_Iron            B   1.350000\n",
       "pFe21_9C     pFe21_9C          08 Low_Iron            C   1.705215"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colData(de8.k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "628b0472",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Organism</th><th scope=col>Iron.amendment.vs.low.iron</th><th scope=col>High.iron.vs.low.iron</th><th scope=col>Iron.amendment.vs.high.iron</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>C. closterium 4</td><td> 206</td><td>NA  </td><td>NA </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>C. closterium 8</td><td> 830</td><td>300 </td><td>947</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>G. oceanica    </td><td>1291</td><td>1381</td><td>221</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>G. huxleyi     </td><td>1154</td><td>1132</td><td>396</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & Organism & Iron.amendment.vs.low.iron & High.iron.vs.low.iron & Iron.amendment.vs.high.iron\\\\\n",
       "  & <chr> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & C. closterium 4 &  206 & NA   & NA \\\\\n",
       "\t2 & C. closterium 8 &  830 & 300  & 947\\\\\n",
       "\t3 & G. oceanica     & 1291 & 1381 & 221\\\\\n",
       "\t4 & G. huxleyi      & 1154 & 1132 & 396\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 4\n",
       "\n",
       "| <!--/--> | Organism &lt;chr&gt; | Iron.amendment.vs.low.iron &lt;int&gt; | High.iron.vs.low.iron &lt;chr&gt; | Iron.amendment.vs.high.iron &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | C. closterium 4 |  206 | NA   | NA  |\n",
       "| 2 | C. closterium 8 |  830 | 300  | 947 |\n",
       "| 3 | G. oceanica     | 1291 | 1381 | 221 |\n",
       "| 4 | G. huxleyi      | 1154 | 1132 | 396 |\n",
       "\n"
      ],
      "text/plain": [
       "  Organism        Iron.amendment.vs.low.iron High.iron.vs.low.iron\n",
       "1 C. closterium 4  206                       NA                   \n",
       "2 C. closterium 8  830                       300                  \n",
       "3 G. oceanica     1291                       1381                 \n",
       "4 G. huxleyi      1154                       1132                 \n",
       "  Iron.amendment.vs.high.iron\n",
       "1 NA                         \n",
       "2 947                        \n",
       "3 221                        \n",
       "4 396                        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# how many DE's per treatment-sample?\n",
    "\n",
    "avl4.perc.k=(nrow(filter(AvL4.k,(padj < 0.05)==T))/nrow(de4.k))*100\n",
    "avl4.k=nrow(filter(AvL4.k,(padj < 0.05)==T))\n",
    "\n",
    "avl8.perc.k=(nrow(filter(AvL8.k,(padj < 0.05)==T))/nrow(de8.k))*100\n",
    "avl8.k=nrow(filter(AvL8.k,(padj < 0.05)==T))\n",
    "\n",
    "avl6.perc.k=(nrow(filter(AvL6.k,(padj < 0.05)==T))/nrow(de6.k))*100\n",
    "avl6.k=nrow(filter(AvL6.k,(padj < 0.05)==T))\n",
    "\n",
    "avl13.perc.k=(nrow(filter(AvL13.k,(padj < 0.05)==T))/nrow(de13.k))*100\n",
    "avl13.k=nrow(filter(AvL13.k,(padj < 0.05)==T))\n",
    "\n",
    "\n",
    "hvl8.perc.k=(nrow(filter(HvL8.k,(padj < 0.05)==T))/nrow(de8.k))*100\n",
    "hvl8.k=nrow(filter(HvL8.k,(padj < 0.05)==T))\n",
    "\n",
    "hvl6.perc.k=(nrow(filter(HvL6.k,(padj < 0.05)==T))/nrow(de6.k))*100\n",
    "hvl6.k=nrow(filter(HvL6.k,(padj < 0.05)==T))\n",
    "\n",
    "hvl13.perc.k=(nrow(filter(HvL13.k,(padj < 0.05)==T))/nrow(de13.k))*100\n",
    "hvl13.k=nrow(filter(HvL13.k,(padj < 0.05)==T))\n",
    "\n",
    "\n",
    "avh8.perc.k=(nrow(filter(AvH8.k,(padj < 0.05)==T))/nrow(de8.k))*100\n",
    "avh8.k=nrow(filter(AvH8.k,(padj < 0.05)==T))\n",
    "\n",
    "avh6.perc.k=(nrow(filter(AvH6.k,(padj < 0.05)==T))/nrow(de6.k))*100\n",
    "avh6.k=nrow(filter(AvH6.k,(padj < 0.05)==T))\n",
    "\n",
    "avh13.perc.k=(nrow(filter(AvH13.k,(padj < 0.05)==T))/nrow(de13.k))*100\n",
    "avh13.k=nrow(filter(AvH13.k,(padj < 0.05)==T))\n",
    "\n",
    "de_ko_output = data.frame('Organism'=c('C. closterium 4', 'C. closterium 8', 'G. oceanica', 'G. huxleyi'),\n",
    "                       'Iron amendment vs low iron'=c(avl4.k,avl8.k,avl6.k,avl13.k),\n",
    "                       'High iron vs low iron'=c('NA', hvl8.k,hvl6.k,hvl13.k),\n",
    "                       'Iron amendment vs high iron'=c('NA',avh8.k,avh6.k,avh13.k))\n",
    "\n",
    "de_ko_output_perc = data.frame('Organism'=c('C. closterium 4', 'C. closterium 8', 'G. oceanica', 'G. huxleyi'),\n",
    "                       'Iron amendment vs low iron'=c(avl4.perc.k,avl8.perc.k,avl6.perc.k,avl13.perc.k),\n",
    "                       'High iron vs low iron'=c('NA', hvl8.perc.k,hvl6.perc.k,hvl13.perc.k),\n",
    "                       'Iron amendment vs high iron'=c('NA',avh8.perc.k,avh6.perc.k,avh13.perc.k)) \n",
    "\n",
    "write.csv(de_ko_output, './de_res_files/de_ko_output.csv', row.names=F)\n",
    "write.csv(de_ko_output_perc, './de_res_files/de_ko_output_perc.csv', row.names=F)\n",
    "\n",
    "head(de_ko_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "1ec925b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mExpected 2 pieces. Missing pieces filled with `NA` in 17 rows [1, 2, 3, 5, 7, 8, 9, 10, 13, 17, 22, 24,\n",
      "26, 27, 28, 29, 30].”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mExpected 2 pieces. Missing pieces filled with `NA` in 15 rows [2, 3, 4, 8, 9, 11, 12, 14, 15, 16, 20, 24,\n",
      "26, 29, 30].”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mExpected 2 pieces. Missing pieces filled with `NA` in 14 rows [1, 4, 5, 6, 7, 8, 11, 12, 20, 21, 22, 27,\n",
      "29, 30].”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mExpected 2 pieces. Missing pieces filled with `NA` in 18 rows [1, 3, 5, 6, 7, 10, 11, 12, 15, 17, 18, 20,\n",
      "21, 22, 23, 24, 26, 29].”\n"
     ]
    }
   ],
   "source": [
    "#Save info for top 30 VSD normalized orfs with highest variance across samples\n",
    "\n",
    "top30vsd.info <- function(k.vsd, ko_def, org){\n",
    "    #order vsd matrix by variance across samples\n",
    "    vsd <- assay(k.vsd)\n",
    "    var_order <- order(rowVars(vsd), decreasing=T)\n",
    "    vsd_new <- vsd[var_order, ]\n",
    "    vsd_new <- as.data.frame(vsd_new) %>% rownames_to_column(\"ko_id\") \n",
    "    #select top 30 orfs, these have the highest variance\n",
    "    vsd30 <- vsd_new[1:30,]\n",
    "    #merge with name and info for ko id and create seperate column of enzyme info\n",
    "    vsd30 <- left_join(vsd30, ko.def, by = 'ko_id') %>% select('ko_id', 'symbol', 'name')\n",
    "    vsd30 <- vsd30 %>% separate(name, c('name', 'enzyme'), \"\\\\[EC:\")  \n",
    "    vsd30$enzyme <- str_replace(vsd30$enzyme, \"\\\\]\", \"\")\n",
    "    vsd30\n",
    "    write.csv(vsd30, paste(\"./vsd_files/vsd.30.\", org, \".csv\", sep=''), row.names=F)\n",
    "}\n",
    "\n",
    "top30vsd.info(k.vsd4, ko4_def, \"4\")\n",
    "top30vsd.info(k.vsd8, ko8_def, \"8\")\n",
    "top30vsd.info(k.vsd6, ko6_def, \"6\")\n",
    "top30vsd.info(k.vsd13, ko13_def, \"13\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "43f4ce13",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n",
      "using 'ashr' for LFC shrinkage. If used in published research, please cite:\n",
      "    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.\n",
      "    https://doi.org/10.1093/biostatistics/kxw041\n",
      "\n"
     ]
    }
   ],
   "source": [
    "lfc4.k.AvL <- lfcShrink(de4.k,  contrast=c(\"treatment\", \"Add_Back\", \"Low_Iron\"), type='ashr')\n",
    "lfc4.k.AvL <- lfc4.k.AvL %>% as.data.frame() %>% rownames_to_column(\"ko_id\") \n",
    "lfc8.k.AvL <- lfcShrink(de8.k, contrast=c(\"treatment\", \"Add_Back\", \"Low_Iron\"), type='ashr')\n",
    "lfc8.k.AvL <- lfc8.k.AvL %>% as.data.frame() %>% rownames_to_column(\"ko_id\") \n",
    "lfc6.k.AvL <- lfcShrink(de6.k, contrast=c(\"treatment\", \"Add_Back\", \"Low_Iron\"), type='ashr')\n",
    "lfc6.k.AvL <- lfc6.k.AvL %>% as.data.frame() %>% rownames_to_column(\"ko_id\") \n",
    "lfc13.k.AvL <- lfcShrink(de13.k,  contrast=c(\"treatment\", \"Add_Back\", \"Low_Iron\"), type='ashr')\n",
    "lfc13.k.AvL <- lfc13.k.AvL %>% as.data.frame() %>% rownames_to_column(\"ko_id\") \n",
    "\n",
    "\n",
    "lfc8.k.HvL <- lfcShrink(de8.k, contrast=c(\"treatment\", \"High_Iron\", \"Low_Iron\"), type='ashr')\n",
    "lfc8.k.HvL <- lfc8.k.HvL %>% as.data.frame() %>% rownames_to_column(\"ko_id\") \n",
    "lfc6.k.HvL <- lfcShrink(de6.k, contrast=c(\"treatment\", \"High_Iron\", \"Low_Iron\"), type='ashr')\n",
    "lfc6.k.HvL <- lfc6.k.HvL %>% as.data.frame() %>% rownames_to_column(\"ko_id\") \n",
    "lfc13.k.HvL <- lfcShrink(de13.k,  contrast=c(\"treatment\", \"High_Iron\", \"Low_Iron\"), type='ashr')\n",
    "lfc13.k.HvL <- lfc13.k.HvL %>% as.data.frame() %>% rownames_to_column(\"ko_id\") \n",
    "\n",
    "lfc8.k.AvH <- lfcShrink(de8.k, contrast=c(\"treatment\", \"Add_Back\", \"High_Iron\"), type='ashr')\n",
    "lfc8.k.AvH <- lfc8.k.AvH %>% as.data.frame() %>% rownames_to_column(\"ko_id\") \n",
    "lfc6.k.AvH <- lfcShrink(de6.k, contrast=c(\"treatment\", \"Add_Back\", \"High_Iron\"), type='ashr')\n",
    "lfc6.k.AvH <- lfc6.k.AvH %>% as.data.frame() %>% rownames_to_column(\"ko_id\")\n",
    "lfc13.k.AvH <- lfcShrink(de13.k,  contrast=c(\"treatment\", \"Add_Back\", \"High_Iron\"), type='ashr')\n",
    "lfc13.k.AvH <- lfc13.k.AvH %>% as.data.frame() %>% rownames_to_column(\"ko_id\") \n",
    "\n",
    "write.csv(lfc4.k.AvL, \"./de_res_files/lfc4.k.AvL.csv\", row.names=F)\n",
    "write.csv(lfc8.k.AvL, \"./de_res_files/lfc8.k.AvL.csv\", row.names=F)\n",
    "write.csv(lfc6.k.AvL, \"./de_res_files/lfc6.k.AvL.csv\", row.names=F)\n",
    "write.csv(lfc13.k.AvL, \"./de_res_files/lfc13.k.AvL.csv\", row.names=F)\n",
    "write.csv(lfc8.k.HvL, \"./de_res_files/lfc8.k.HvL.csv\", row.names=F)\n",
    "write.csv(lfc6.k.HvL, \"./de_res_files/lfc6.k.HvL.csv\", row.names=F)\n",
    "write.csv(lfc13.k.HvL, \"./de_res_files/lfc13.k.HvL.csv\", row.names=F)\n",
    "write.csv(lfc8.k.AvH, \"./de_res_files/lfc8.k.AvH.csv\", row.names=F)\n",
    "write.csv(lfc6.k.AvH, \"./de_res_files/lfc6.k.AvH.csv\", row.names=F)\n",
    "write.csv(lfc13.k.AvH, \"./de_res_files/lfc13.k.AvH.csv\", row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fdca0def-7132-4560-8259-86aa91f206ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n",
      "\u001b[1m\u001b[22mJoining with `by = join_by(ko_id)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ko_id</th><th scope=col>baseMean</th><th scope=col>log2FoldChange</th><th scope=col>lfcSE</th><th scope=col>pvalue</th><th scope=col>padj</th><th scope=col>symbol</th><th scope=col>name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>K00525</td><td>1073.92643</td><td>-3.904022</td><td>0.2805200</td><td>8.315263e-46</td><td>1.930804e-42</td><td> E1.17.4.1A, nrdA, nrdE</td><td> ribonucleoside-diphosphate reductase alpha chain [EC:1.17.4.1]</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>K10734</td><td>  25.89301</td><td>-3.212495</td><td>0.5927836</td><td>1.303292e-09</td><td>1.061840e-07</td><td> GINS3                 </td><td> GINS complex subunit 3                                        </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>K10808</td><td> 501.63986</td><td>-3.196746</td><td>0.2288487</td><td>1.401861e-45</td><td>2.170081e-42</td><td> RRM2                  </td><td> ribonucleoside-diphosphate reductase subunit M2 [EC:1.17.4.1] </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>K18710</td><td>  22.68003</td><td>-3.160268</td><td>0.8150535</td><td>4.954375e-07</td><td>2.130381e-05</td><td> SLBP                  </td><td> histone RNA hairpin-binding protein                           </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>K22156</td><td>  56.82509</td><td>-3.085140</td><td>0.4034052</td><td>9.545449e-16</td><td>1.528588e-13</td><td> JADE3                 </td><td> protein Jade-3                                                </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>K14291</td><td>  83.10115</td><td>-3.078023</td><td>0.3364991</td><td>3.207594e-21</td><td>8.762392e-19</td><td> PHAX                  </td><td> phosphorylated adapter RNA export protein                     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & ko\\_id & baseMean & log2FoldChange & lfcSE & pvalue & padj & symbol & name\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & K00525 & 1073.92643 & -3.904022 & 0.2805200 & 8.315263e-46 & 1.930804e-42 &  E1.17.4.1A, nrdA, nrdE &  ribonucleoside-diphosphate reductase alpha chain {[}EC:1.17.4.1{]}\\\\\n",
       "\t2 & K10734 &   25.89301 & -3.212495 & 0.5927836 & 1.303292e-09 & 1.061840e-07 &  GINS3                  &  GINS complex subunit 3                                        \\\\\n",
       "\t3 & K10808 &  501.63986 & -3.196746 & 0.2288487 & 1.401861e-45 & 2.170081e-42 &  RRM2                   &  ribonucleoside-diphosphate reductase subunit M2 {[}EC:1.17.4.1{]} \\\\\n",
       "\t4 & K18710 &   22.68003 & -3.160268 & 0.8150535 & 4.954375e-07 & 2.130381e-05 &  SLBP                   &  histone RNA hairpin-binding protein                           \\\\\n",
       "\t5 & K22156 &   56.82509 & -3.085140 & 0.4034052 & 9.545449e-16 & 1.528588e-13 &  JADE3                  &  protein Jade-3                                                \\\\\n",
       "\t6 & K14291 &   83.10115 & -3.078023 & 0.3364991 & 3.207594e-21 & 8.762392e-19 &  PHAX                   &  phosphorylated adapter RNA export protein                     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 8\n",
       "\n",
       "| <!--/--> | ko_id &lt;chr&gt; | baseMean &lt;dbl&gt; | log2FoldChange &lt;dbl&gt; | lfcSE &lt;dbl&gt; | pvalue &lt;dbl&gt; | padj &lt;dbl&gt; | symbol &lt;chr&gt; | name &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | K00525 | 1073.92643 | -3.904022 | 0.2805200 | 8.315263e-46 | 1.930804e-42 |  E1.17.4.1A, nrdA, nrdE |  ribonucleoside-diphosphate reductase alpha chain [EC:1.17.4.1] |\n",
       "| 2 | K10734 |   25.89301 | -3.212495 | 0.5927836 | 1.303292e-09 | 1.061840e-07 |  GINS3                  |  GINS complex subunit 3                                         |\n",
       "| 3 | K10808 |  501.63986 | -3.196746 | 0.2288487 | 1.401861e-45 | 2.170081e-42 |  RRM2                   |  ribonucleoside-diphosphate reductase subunit M2 [EC:1.17.4.1]  |\n",
       "| 4 | K18710 |   22.68003 | -3.160268 | 0.8150535 | 4.954375e-07 | 2.130381e-05 |  SLBP                   |  histone RNA hairpin-binding protein                            |\n",
       "| 5 | K22156 |   56.82509 | -3.085140 | 0.4034052 | 9.545449e-16 | 1.528588e-13 |  JADE3                  |  protein Jade-3                                                 |\n",
       "| 6 | K14291 |   83.10115 | -3.078023 | 0.3364991 | 3.207594e-21 | 8.762392e-19 |  PHAX                   |  phosphorylated adapter RNA export protein                      |\n",
       "\n"
      ],
      "text/plain": [
       "  ko_id  baseMean   log2FoldChange lfcSE     pvalue       padj        \n",
       "1 K00525 1073.92643 -3.904022      0.2805200 8.315263e-46 1.930804e-42\n",
       "2 K10734   25.89301 -3.212495      0.5927836 1.303292e-09 1.061840e-07\n",
       "3 K10808  501.63986 -3.196746      0.2288487 1.401861e-45 2.170081e-42\n",
       "4 K18710   22.68003 -3.160268      0.8150535 4.954375e-07 2.130381e-05\n",
       "5 K22156   56.82509 -3.085140      0.4034052 9.545449e-16 1.528588e-13\n",
       "6 K14291   83.10115 -3.078023      0.3364991 3.207594e-21 8.762392e-19\n",
       "  symbol                 \n",
       "1  E1.17.4.1A, nrdA, nrdE\n",
       "2  GINS3                 \n",
       "3  RRM2                  \n",
       "4  SLBP                  \n",
       "5  JADE3                 \n",
       "6  PHAX                  \n",
       "  name                                                           \n",
       "1  ribonucleoside-diphosphate reductase alpha chain [EC:1.17.4.1]\n",
       "2  GINS complex subunit 3                                        \n",
       "3  ribonucleoside-diphosphate reductase subunit M2 [EC:1.17.4.1] \n",
       "4  histone RNA hairpin-binding protein                           \n",
       "5  protein Jade-3                                                \n",
       "6  phosphorylated adapter RNA export protein                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ranked4.AvL = lfc4.k.AvL[order(abs(lfc4.k.AvL$log2FoldChange), decreasing=TRUE),]\n",
    "ranked4.AvL = left_join(ranked4.AvL, ko.def)\n",
    "\n",
    "\n",
    "ranked8.AvL = lfc8.k.AvL[order(abs(lfc8.k.AvL$log2FoldChange), decreasing=TRUE),]\n",
    "ranked8.AvL = left_join(ranked8.AvL, ko.def)\n",
    "\n",
    "ranked8.HvL = lfc8.k.HvL[order(abs(lfc8.k.HvL$log2FoldChange), decreasing=TRUE),]\n",
    "ranked8.HvL = left_join(ranked8.HvL, ko.def)\n",
    "\n",
    "ranked8.AvH = lfc8.k.AvH[order(abs(lfc8.k.AvH$log2FoldChange), decreasing=TRUE),]\n",
    "ranked8.AvH = left_join(ranked8.AvH, ko.def)\n",
    "\n",
    "\n",
    "ranked6.AvL = lfc6.k.AvL[order(abs(lfc6.k.AvL$log2FoldChange), decreasing=TRUE),]\n",
    "ranked6.AvL = left_join(ranked6.AvL, ko.def)\n",
    "\n",
    "ranked6.HvL = lfc6.k.HvL[order(abs(lfc6.k.HvL$log2FoldChange), decreasing=TRUE),]\n",
    "ranked6.HvL = left_join(ranked6.HvL, ko.def)\n",
    "\n",
    "ranked6.AvH = lfc6.k.AvH[order(abs(lfc6.k.AvH$log2FoldChange), decreasing=TRUE),]\n",
    "ranked6.AvH = left_join(ranked6.AvH, ko.def)\n",
    "\n",
    "\n",
    "ranked13.AvL = lfc13.k.AvL[order(abs(lfc13.k.AvL$log2FoldChange), decreasing=TRUE),]\n",
    "ranked13.AvL = left_join(ranked13.AvL, ko.def)\n",
    "\n",
    "ranked13.HvL = lfc13.k.HvL[order(abs(lfc13.k.HvL$log2FoldChange), decreasing=TRUE),]\n",
    "ranked13.HvL = left_join(ranked13.HvL, ko.def)\n",
    "\n",
    "ranked13.AvH = lfc13.k.AvH[order(abs(lfc13.k.AvH$log2FoldChange), decreasing=TRUE),]\n",
    "ranked13.AvH = left_join(ranked13.AvH, ko.def)\n",
    "\n",
    "head(ranked13.AvH)\n",
    "write.csv(ranked4.AvL, \"./de_res_files/ranked4.AvL.csv\", row.names=F)\n",
    "write.csv(ranked8.AvL, \"./de_res_files/ranked8.AvL.csv\", row.names=F)\n",
    "write.csv(ranked6.AvL, \"./de_res_files/ranked6.AvL.csv\", row.names=F)\n",
    "write.csv(ranked13.AvL, \"./de_res_files/ranked13.AvL.csv\", row.names=F)\n",
    "write.csv(ranked8.AvH, \"./de_res_files/ranked8.AvH.csv\", row.names=F)\n",
    "write.csv(ranked6.AvH, \"./de_res_files/ranked6.AvH.csv\", row.names=F)\n",
    "write.csv(ranked13.AvH, \"./de_res_files/ranked13.AvH.csv\", row.names=F)\n",
    "write.csv(ranked8.HvL, \"./de_res_files/ranked8.HvL.csv\", row.names=F)\n",
    "write.csv(ranked6.HvL, \"./de_res_files/ranked6.HvL.csv\", row.names=F)\n",
    "write.csv(ranked13.HvL, \"./de_res_files/ranked13.HvL.csv\", row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "36f6da91",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in file(file, \"rt\"):\n",
      "“cannot open file '\n",
      "': No such file or directory”\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in file(file, \"rt\"): cannot open the connection\n",
     "output_type": "error",
     "traceback": [
      "Error in file(file, \"rt\"): cannot open the connection\nTraceback:\n",
      "1. read.csv((module.url), sep = \"\\t\", header = F)",
      "2. read.table(file = file, header = header, sep = sep, quote = quote, \n .     dec = dec, fill = fill, comment.char = comment.char, ...)",
      "3. file(file, \"rt\")"
     ]
    }
   ],
   "source": [
    "module.url <- \"https://rest.kegg.jp/link/module/ko04050\"\n",
    "    module.url <- getURL(module.url)\n",
    "    module.link <- read.csv((module.url), sep = '\\t', header = F)  #find brite categories associated with ko\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1f860e3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "\n",
       "\n"
      ],
      "text/plain": [
       "list()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "ERROR",
     "evalue": "Error in read.table(file = file, header = header, sep = sep, quote = quote, : no lines available in input\n",
     "output_type": "error",
     "traceback": [
      "Error in read.table(file = file, header = header, sep = sep, quote = quote, : no lines available in input\nTraceback:\n",
      "1. read.csv(textConnection(name.table), sep = \"\\t\", header = F)",
      "2. read.table(file = file, header = header, sep = sep, quote = quote, \n .     dec = dec, fill = fill, comment.char = comment.char, ...)",
      "3. stop(\"no lines available in input\")"
     ]
    }
   ],
   "source": [
    "ko.30 <- list(vsd4.30$ko_id)\n",
    "\n",
    "module.list <- list()\n",
    "for (i in ko.30){\n",
    "    module.url <-  paste0(\"https://rest.kegg.jp/link/module/\",i,sep='')\n",
    "    module.url <- getURL(module.url)\n",
    "    module.link <- read.csv(textConnection(module.url), sep = '\\t', header = F)  #find brite categories associated with ko\n",
    "    module.link$V2 <-  str_replace(module.link$V2, pattern = \"md:\" , replacement = '')\n",
    "    module.link <- list(module.link$V2)\n",
    "    append(module.list, module.link)\n",
    "}\n",
    "\n",
    "module.list\n",
    "#for (i in module.list){\n",
    "    name.url <- paste0(\"https://rest.kegg.jp/find/module/\", 'ko00195', sep='')\n",
    "    name.table <- getURL(name.url)\n",
    "   name.table <-  read.csv(textConnection(name.table), sep='\\t', header = F)\n",
    "    print(name.table)\n",
    "    #append(final, name.table)\n",
    "#}\n",
    "\n",
    "pathways <- list()\n",
    "for (i in ko.30){\n",
    "    pathway.url <-  paste0(\"https://rest.kegg.jp/link/pathway/\",i,sep='')\n",
    "    pathway.url <- getURL(pathway.url)\n",
    "    pathway.link <- read.csv(textConnection(pathway.url), sep = '\\t', header = F)  #find brite categories associated with ko\n",
    "    pathway.link$V2 <-  str_replace(pathway.link$V2, pattern = \"path:\" , replacement = '')\n",
    "    pathway.list <- list(pathway.link$V2)\n",
    "    print(pathway.list)\n",
    "    append(pathways, pathway.list)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "for (i in pathway.list){\n",
    "    name.url <- paste0(\"https://rest.kegg.jp/find/pathway/\", i, sep='')\n",
    "    name.table <- getURL(name.url)\n",
    "   name.table <-  read.csv(textConnection(name.table), sep='\\t', header = F)\n",
    "    print(name.table)\n",
    "    #append(final, name.table)\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe578036",
   "metadata": {},
   "source": [
    "# Glimma MA plots\n",
    "\n",
    "The Glimma MDS contains two main components:\n",
    "\n",
    "1. a plot showing two MDS dimensions, and\n",
    "2. a plot of the eigenvalues of each dimension\n",
    "\n",
    "The Glimma MA plot contains two main components:\n",
    "\n",
    "1. a plot of summary statistics across all genes that have been tested, and\n",
    "2. a plot of gene expression from individual samples for a given gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0edd00f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "glimmaMDS(k.dds6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9cb40f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ma.glimma <- function(de, lfc.shrink){\n",
    "    #plotting function removes orfs with a pvalue of na, its not updating the anno df properly, \n",
    "    #so here I calculated those which will be removed\n",
    "    #and make a df with the annotations already removed\n",
    "    r <- as.data.frame(results(de))\n",
    "    complete.genes <- (complete.cases(r))\n",
    "    res.comp <- r[complete.genes, ]\n",
    "    x <- de[complete.genes, ]\n",
    "    anno <- mcols(x)\n",
    "\n",
    "    glimmaMA(de,\n",
    "            #status=NULL,                                    # vector indicated status of each gene\n",
    "             groups= colData(de)$treatment,    #vector length of ncol dds, shows categories of samples \n",
    "             anno=anno,                                         #df with nrow of de object with rows containing gene annotations\n",
    "             #display.columns=c('symbol','name'),  #character vector with names from anno from which to display \n",
    "                 #type of transformation used\n",
    "                                                                            #default is logcpm\n",
    "             #status.cols=NULL,                           #vector of 3, indicateing CSS strings for color associated with \n",
    "                                                                          #status of -1, 0, 1\n",
    "            #sample.cols=NULL,                     #character vector of length ncol(counts) with CSS strings for \n",
    "                                                                        #colors for each sample\n",
    "             main='MA Plot', \n",
    "             xlab='logCPM', \n",
    "             ylab='log2FC')\n",
    "            #html=NULL,                               #character string for naming HTML file for exporting widget ending in .html)\n",
    "    }\n",
    "ma.glimma(de4.k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2562659a",
   "metadata": {},
   "outputs": [],
   "source": [
    "lfc13.k.HvL <- lfcShrink(de13.k,  contrast=c(\"treatment\", \"High_Iron\", \"Low_Iron\"), type='ashr')\n",
    "mcols(lfc13.k.HvL)\n",
    "mcols(de13.k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69062923",
   "metadata": {},
   "outputs": [],
   "source": [
    "d <- data.frame('ko_id'= rownames(k.dds6))\n",
    "\n",
    "v <- ko6_def%>% filter((ko_id %in% dds$ko_id)==TRUE) %>% select(ko_id, symbol, name)\n",
    "anno <- distinct(v)\n",
    "all(rownames(k.dds6)==anno$ko_id)\n",
    "#reorder annotation table to same row order as dds\n",
    "anno <- anno[match(rownames(k.dds6), anno$ko_id),]\n",
    "all(rownames(k.dds6)==anno$ko_id)\n",
    "\n",
    "mcols(k.dds6) <- cbind(mcols(k.dds6), anno)\n",
    "mcols(k.dds6)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d599de3a",
   "metadata": {},
   "source": [
    "## Now we have ordered vsd file and results file (for each comparison) for both transcript level and counts summarized at kegg annotation level. \n",
    "#### Taking these files we will create plots. \n",
    "## It might be best to use transcript level for ma plots and 'gene' level for clusterProfiler?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
