{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9171cae8",
   "metadata": {},
   "source": [
    "Read in tables with significant DEGs and mapped annotations like GO or KEGG. Create vector of Kegg id's to match them to a descriptive name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a66141cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "library('tidyr')\n",
    "library('ggplot2')\n",
    "library('stringr')\n",
    "library('dplyr')\n",
    "library(\"RCurl\")\n",
    "library('tibble')\n",
    "library('KEGGREST')\n",
    "library('RCurl')\n",
    "#library('enrichplot')\n",
    "#library('data.table')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "070bb3bf",
   "metadata": {},
   "source": [
    "# Read in kegg annotations from eggnog\n",
    "## make a list of all unique kegg_ko's from all annotations \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66403180",
   "metadata": {},
   "outputs": [],
   "source": [
    "get.eggnog <- function(org){\n",
    "    dir2  <- paste(\"/work/nclab/lucy/SAB/Annotation/eggnog/\", org, sep=\"\")\n",
    "    \n",
    "    emap <- read.csv(file.path(dir2, paste(org,\".emapper.annotations\", sep=\"\")),\n",
    "                 sep = \"\\t\",\n",
    "                 comment.char = \"#\",\n",
    "                 header = FALSE,\n",
    "                 na.strings = \"-\")\n",
    "    \n",
    "    colnames(emap) <- c(\n",
    "        'orfs', 'seed_ortholog', 'evalue', 'score', 'eggNOG_OGs', 'max_annot_lvl', 'COG_category', 'Description', \n",
    "        'Preferred_name', 'GOs', 'EC', 'ko_id', 'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction', \n",
    "        'KEGG_rclass',  'BRITE', 'KEGG_TC', 'CAZy', 'BiGG_Reaction', 'PFAMs')\n",
    "    \n",
    "    emap[, c(\"orfs\", \"seed_ortholog\", \"EC\",\"GOs\",\"ko_id\",'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction', \n",
    "       'KEGG_rclass',  'BRITE', 'KEGG_TC', 'CAZy', 'BiGG_Reaction',\"PFAMs\")]\n",
    "    }\n",
    "\n",
    "emap4 <- get.eggnog('04')\n",
    "emap8 <- get.eggnog('08')\n",
    "emap6 <- get.eggnog('06')\n",
    "emap13 <- get.eggnog('13')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1453b0ea-27d5-42aa-9a2a-73ac054804c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "length(na.omit(emap4$ko_id))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f41c467",
   "metadata": {},
   "source": [
    "Create df for each organism with annotation of choice, removing the rows with NA's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf728468",
   "metadata": {},
   "outputs": [],
   "source": [
    "get.anno <- function(emap, anno){\n",
    "    #extract query column and annotation of choice, ex: pfams\n",
    "    vars <- c(\"orfs\", anno)\n",
    "    df <- select(emap, all_of(vars))\n",
    "    df <- df %>% filter(is.na(df[,ncol(df)])==FALSE)\n",
    "    }\n",
    "\n",
    "pfam4 <- get.anno(emap4, \"PFAMs\")\n",
    "pfam8 <- get.anno(emap8, \"PFAMs\")\n",
    "pfam6 <- get.anno(emap6, \"PFAMs\")\n",
    "pfam13 <- get.anno(emap13, \"PFAMs\")\n",
    "\n",
    "ko4_ls <- get.anno(emap4,\"ko_id\")\n",
    "ko8_ls <- get.anno(emap8,\"ko_id\")\n",
    "ko6_ls <- get.anno(emap6,\"ko_id\")\n",
    "ko13_ls <- get.anno(emap13,\"ko_id\")\n",
    "\n",
    "go4 <- get.anno(emap4, \"GOs\")\n",
    "go8 <- get.anno(emap8, \"GOs\")\n",
    "go6 <- get.anno(emap6, \"GOs\")\n",
    "go13 <- get.anno(emap13, \"GOs\")\n",
    "\n",
    "paths4 <- get.anno(emap4, c('KEGG_Module', 'KEGG_Pathway','BRITE'))\n",
    "paths8 <- get.anno(emap8, c('KEGG_Module', 'KEGG_Pathway','BRITE'))\n",
    "paths6 <- get.anno(emap6, c('KEGG_Module', 'KEGG_Pathway','BRITE'))\n",
    "paths13 <- get.anno(emap13, c('KEGG_Module', 'KEGG_Pathway','BRITE'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7722c6ee-90a0-43e1-9ede-d9707ad2f4b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('# kos')\n",
    "dim(ko4_ls)\n",
    "dim(ko8_ls)\n",
    "dim(ko6_ls)\n",
    "dim(ko13_ls)\n",
    "print('# pfams')\n",
    "dim(pfam4)\n",
    "dim(pfam8)\n",
    "dim(pfam6)\n",
    "dim(pfam13)\n",
    "print('# gos')\n",
    "dim(go4)\n",
    "dim(go8)\n",
    "dim(go6)\n",
    "dim(go13)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8aa3b7a",
   "metadata": {},
   "source": [
    "## Create a ko-to-orf mapping table\n",
    "#### Split columns with mulitple annotations and pivot into a long table with repeating orfs\n",
    "This will give us a 3 column table with columns 'orfs', 'ko_id', and 'ko_iteration'; which repeats orfs for rows from original table with multiple eggnog assignments. The ko_iteration increases from 1 to n for n eggnog assignmnet for a particular orf. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1440559",
   "metadata": {},
   "outputs": [],
   "source": [
    "split_into_multiple <- function(column, pattern = \",\", into_prefix){\n",
    "    #adapted from post on stack overflow\n",
    "    cols <- str_split_fixed(column, pattern, n = Inf)\n",
    "  # Replace empty matrix indicies with NA's \n",
    "    cols[which(cols == \"\")] <- NA\n",
    "  # turn matrix into a table with unique but arbitraty column names  \n",
    "    cols <- as_tibble(cols, .name_repair = make.names)\n",
    "  # where m = # columns in tibble 'cols'\n",
    "    m <- dim(cols)[2]\n",
    "  # assign column names as 'into_prefix_1', 'into_prefix_2', ..., 'into_prefix_m' \n",
    "    names(cols) <- paste(into_prefix, 1:m, sep = \"_\")\n",
    "    print('# of values in matrix w/o NA:')\n",
    "    print(sum(!is.na(cols)))\n",
    "    return(cols)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "clean_ko <- function(df, org){\n",
    "  # split up ko_id's into multiple columns naming each column \n",
    "  # ko:_1 to ko:_n \n",
    "  # (remember at this point we have the same number of rows but \n",
    "  # far more columns)\n",
    "    ko_iterations <- split_into_multiple(df$ko_id, \",\", \"ko:\")\n",
    "  # select the orfs column from original df and bind to to \n",
    "  # split columns \n",
    "    df = df %>% select(orfs) %>% bind_cols(ko_iterations)\n",
    "  # now combine all ko:_n columns so all ko_id's \n",
    "  # become one column called ko_id and each column name, \n",
    "  # 'ko:_1'...'ko:_n' becomes one column called 'ko_iteration'. \n",
    "  # This tells us how many ko_id's were assigned to a particular orf\n",
    "  # the column 'orfs' will repeat values for rows with > 1 ko_iteration\n",
    "  # Drop the values from the matrix which were NA\n",
    "    df_clean = pivot_longer(df, cols = !orfs, values_drop_na = T,\n",
    "                            values_to = 'ko_id', names_to = 'ko_iteration')\n",
    "  # finally, clean up the df a bit by removing the ko_id prefixes 'ko:'\n",
    "  # we now have a 3 column table with orfs repeated in for each instance \n",
    "  # a ko_id was assigned to it by eggnog!\n",
    "    df_clean$ko_id <- gsub('ko:', '', df_clean$ko_id)\n",
    "  # the final df should have same number of rows as sum(!na(cols))\n",
    "    print('# rows in final df:')\n",
    "    print(nrow(df_clean))\n",
    "    write.csv(df_clean, paste('../kegg_names/ko',org, '_ls.csv', sep=''), row.names=F)\n",
    "    df_clean\n",
    "    }\n",
    "\n",
    "ko4 <- clean_ko(ko4_ls,'4')\n",
    "ko6 <- clean_ko(ko6_ls,'6')\n",
    "ko8 <- clean_ko(ko8_ls,'8')\n",
    "ko13 <- clean_ko(ko13_ls,'13')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ec7c4cb-f4aa-41de-a8d8-b516a18ec31b",
   "metadata": {},
   "source": [
    "Create a list of all ko's found across all organisms. We will use the list to find the name and symbol of the ko_id's and map it back to the organism-specific tables made above.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "562e1fe0-6593-4267-98ce-475f45b74de6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a df of all unique ko's found for all annotations\n",
    "all.ko <- bind_rows(ko4, ko6,ko8,ko13) %>% select(ko_id)\n",
    "\n",
    "all.ko <- distinct(all.ko, ko_id, .keep_all = T)\n",
    "#so that bash will read my file correctly, I need to start with ko= \n",
    "#and surround all ko_id's by a single quote\n",
    "kk=c('ko= ')\n",
    "q=c('\"')\n",
    "all.ko=rbind(kk, q, all.ko, q)\n",
    "#write to a .txt file with ko_id's separated by a space\n",
    "\n",
    "write.table(all.ko, \"../kegg_names/all.ko.txt\", quote=F, sep=\" \", eol=\" \", row.names = FALSE, col.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d390efbb-d4cc-422c-992a-8f9d5851ba9e",
   "metadata": {},
   "source": [
    "# a total of 7218 rows, or ko_id's "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df2c4ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in table with name and symbol matches to kegg ko's\n",
    "#make tidy\n",
    "ko_def <- read.delim(\"../kegg_names/ko_pathways.txt\", head=F,sep = \";\")\n",
    "id <- str_extract(ko_def$V1,'K[[:digit:]]*')\n",
    "sym <- str_remove(ko_def$V1, 'ko:K[[:digit:]]*')\n",
    "ko_def <- data.frame(ko_id = id, symbol=sym, name=ko_def$V2)\n",
    "isip=data.frame(ko_id=c('isip1a','isip_2a','isip_3'), \n",
    "                symbol=c('isip_1a','isip_2a','isip_3'),\n",
    "                name=c('Iron stress induced protein 1a',\n",
    "                       'Iron stress induced protein 2a', 'Iron stress induced protein 3'))\n",
    "ko_def=bind_rows(ko_def, isip)\n",
    "print(paste(nrow(ko_def), \"    =   all unique ko's found across organisms\"))\n",
    "write.csv(ko_def,'../kegg_names/ko_def.csv', row.names=F)\n",
    "tail(ko_def)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cca5c9",
   "metadata": {},
   "source": [
    "# Now we have tidy df of query and annotation for mapping back to counts. We need to add the ISIP annotations though. \n",
    "## Read in isip isoform annotations and make a tidy df of all isip isoforms for each organism\n",
    "Check that isip's were also counted by Salmon, remove any which were not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5863d89e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## read in the files\n",
    "dir <- \"/work/nclab/lucy/SAB/Annotation/eggnog/isip/pep_hit/\"\n",
    "\n",
    "read.isip <- function(org, prot){\n",
    "    f = t(read.delim(file=paste(dir, org, 'isip', prot,'_hits.txt', sep=\"\"),\n",
    "                 sep = ' ',header = F))\n",
    "    if (nrow(f) >=1){\n",
    "    f <- data.frame(\"orfs\"=f, \n",
    "                    \"ko_iteration\"=\"ko:_1\",\n",
    "                    \"ko_id\"=paste(\"isip_\", prot, sep=\"\"))}\n",
    "}\n",
    "\n",
    "i41 <- read.isip(\"04\", \"1\")\n",
    "i41a <- read.isip(\"04\", \"1a\")\n",
    "i41b <- read.isip(\"04\", \"1b\")\n",
    "i42a <- read.isip(\"04\", \"2a\")\n",
    "i43 <- read.isip(\"04\", \"3\")\n",
    "isip4 <- rbind(i41a, i41b, i42a, i43)\n",
    "isip4 <- distinct(isip4, orfs, .keep_all = T)\n",
    "\n",
    "i81 <- read.isip(\"08\", \"1\")\n",
    "i81a <- read.isip(\"08\", \"1a\")\n",
    "i81b <- read.isip(\"08\", \"1b\")\n",
    "i82a <- read.isip(\"08\", \"2a\")\n",
    "i83 <- read.isip(\"08\", \"3\")\n",
    "isip8 <- rbind( i81a, i81b, i82a, i83)  \n",
    "isip8 <- distinct(isip8, orfs, .keep_all = T)\n",
    "\n",
    "\n",
    "i61 <- read.isip(\"06\", \"1\")\n",
    "i61a <- read.isip(\"06\", \"1a\")\n",
    "i61b <- read.isip(\"06\", \"1b\")\n",
    "i62a <- read.isip(\"06\", \"2a\")\n",
    "i63 <- read.isip(\"06\", \"3\")\n",
    "isip6 <- rbind(i61a, i61b, i62a, i63)\n",
    "isip6 <- distinct(isip6, orfs, .keep_all = T)\n",
    "\n",
    "i132a <- read.isip(\"13\", \"2a\")\n",
    "i133 <- read.isip(\"13\", \"3\")\n",
    "isip13 <- rbind(i132a, i133)\n",
    "isip13 <- distinct(isip13, orfs, .keep_all = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebed5576",
   "metadata": {},
   "source": [
    "## Create df with ko_id and names mapped to ORF's of each organism\n",
    "add the isip proteins to ko dataframe, checking not to repeat if the orf was already annotated. (should I replace annotation with isip?)\n",
    "\n",
    "1. Merge full ko definition list with ko's in organism using many-to-one merge relationship.  Organism's ko df have repeated ko_id, as the same annotation could be matched to multiple ORFs. So many rows in organisms's df may match to one row in full ko definition df. The column ko_id is named the same for both objects and will be used to merge the two dataframes. For this we use dplyr's left-join with relationship=many-to-one. \n",
    "\n",
    "2. Isip df and ko df have column query in common. We create a list of ORF's that the isip df and ko df have in common. isip rows are added to ko df for those ORFs not in common."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28e2fde7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "combine_save <- function(isip, ko_df, org){\n",
    "    \n",
    "    ko_fin <- bind_rows(ko_df, isip)\n",
    " \n",
    "    write.csv(ko_fin, paste(\"../kegg_names/ko\", org,\"_ls.csv\", sep=\"\"), row.names=F)\n",
    "    print(nrow(ko_fin))\n",
    "    print(nrow(ko_df)+nrow(isip))\n",
    "}\n",
    "\n",
    "\n",
    "ko4_def <- combine_save(isip4, ko4, \"4\")\n",
    "ko8_def <- combine_save(isip8, ko8, \"8\")\n",
    "ko6_def <- combine_save(isip6, ko6, \"6\")\n",
    "ko13_def <- combine_save(isip13, ko13, \"13\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0caa3fe4",
   "metadata": {},
   "source": [
    "# Now we have a df with all possible ko annotations, including the isip proteins, associated with orf's \n",
    "\n",
    "## this will be now used in creating figures. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "841cb6b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "emap.ko4 <- emap4[,c('orfs', 'KEGG_ko')]\n",
    "ko4_def$name <- gsub(\" \\\\[EC.*\\\\]\", \"\", ko4_def$name)\n",
    "ko.sum <- ko4_def %>% group_by(orfs)%>% summarize('symbol'=str_flatten(symbol, collaps=\"; \"), 'name'= str_flatten(name, collaps=\"; \"))\n",
    "ko.na <- data.frame('orfs'=emap.ko4[(emap.ko4$orfs %in% ko.sum$orfs)==FALSE, 'orfs'], 'symbol'=NA, 'name'=NA)\n",
    "full.anno <- rbind(ko.sum, ko.na )\n",
    "nrow(emap4)\n",
    "nrow(full.anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d0b7509",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in table with name and symbol matches to kegg ko's\n",
    "#make tidy\n",
    "br_def <- read.csv(\"../kegg_names/brite_def.csv\", header = FALSE)\n",
    "head(br_def)\n",
    "id <- str_extract(br_def$V1,'K[[:digit:]]*')\n",
    "sym <- str_remove(br_def$V1, 'ko:K[[:digit:]]*')\n",
    "br_def <- data.frame(br_id = id, symbol=sym, name=br_def$V2)\n",
    "\n",
    "#write.csv(ko_def,\"../kegg_names/br_def.csv\", row.names=F)\n",
    "\n",
    "print(paste(nrow(br_def), \"     =   all unique BRITEs found across organisms\"))\n",
    "dim(br_def)\n",
    "head(br_def)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6db8a3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_br <- function(df){\n",
    "    df <- df %>% bind_cols(split_into_multiple(.$BRITE, \",\", \"br:\")) %>%\n",
    "           select(\"orfs\", starts_with(\"br:_\", ))\n",
    "paste(head(df))\n",
    "    df_clean <- pivot_longer(df, cols = !orfs, names_to = 'br_iteration',\n",
    "                         values_to ='br_id' , values_drop_na=TRUE)\n",
    "        df_clean$br_id <- gsub('br:', '', df_clean$br_id)\n",
    "    df_clean\n",
    "    }\n",
    "\n",
    "clean_mods <- function(df){\n",
    "    df <- df %>% bind_cols(split_into_multiple(.$KEGG_Module, \",\", \"mo:\")) %>%\n",
    "           select(\"orfs\", starts_with(\"mo:_\", ))\n",
    "paste(head(df))\n",
    "    df_clean <- pivot_longer(df, cols = !orfs, names_to = 'mo_iteration',\n",
    "                         values_to ='mo_id' , values_drop_na=TRUE)\n",
    "        df_clean$mo_id <- gsub('mo:', '', df_clean$mo_id)\n",
    "    df_clean\n",
    "    }\n",
    "\n",
    "clean_paths <- function(df){\n",
    "    df <- df %>% bind_cols(split_into_multiple(.$KEGG_Pathway, \",\", \"path:\")) %>%\n",
    "           select(\"orfs\", starts_with(\"path:_\", ))\n",
    "paste(head(df))\n",
    "    df_clean <- pivot_longer(df, cols = !orfs, names_to = 'path_iteration',\n",
    "                         values_to ='path_id' , values_drop_na=TRUE)\n",
    "        df_clean$path_id <- gsub('path:', '', df_clean$path_id)\n",
    "    df_clean\n",
    "    }\n",
    "\n",
    "br4 <- clean_br(paths4)\n",
    "br8 <- clean_br(paths8)\n",
    "br6 <- clean_br(paths6)\n",
    "br13 <- clean_br(paths13)\n",
    "\n",
    "all.br <- bind_rows(br4, br6, br8, br13) %>% select(br_id)\n",
    "all.br <- distinct(all.br, br_id)\n",
    "#write to a .txt file with path_id's separated by a space \n",
    " write.table(paste(\"brite=\", paste(all.br$br_id, sep=\" \", collapse = \" \")), \"all.brite.txt\", row.names = FALSE, col.names = FALSE)\n",
    "\n",
    "\n",
    "mo4 <- clean_mods(paths4)\n",
    "mo8 <- clean_mods(paths8)\n",
    "mo6 <- clean_mods(paths6)\n",
    "mo13 <- clean_mods(paths13)\n",
    "\n",
    "all.mo <- bind_rows(mo4, mo6, mo8, mo13) %>% select(mo_id)\n",
    "all.mo <- distinct(all.mo, mo_id)\n",
    "#write to a .txt file with path_id's separated by a space \n",
    " write.table(paste(\"module=\", paste(all.mo$mo_id, sep=\" \", collapse = \" \")), \"all.modules.txt\", row.names = FALSE, col.names = FALSE)\n",
    "\n",
    "\n",
    "pa4 <- clean_paths(paths4)\n",
    "pa8 <- clean_paths(paths8)\n",
    "pa6 <- clean_paths(paths6)\n",
    "pa13 <- clean_paths(paths13)\n",
    "\n",
    "all.path <- bind_rows(pa4, pa6,pa8,pa13) %>% select(path_id)\n",
    "all.path <- distinct(all.path, path_id)\n",
    "#write to a .txt file with path_id's separated by a space \n",
    " write.table(paste(\"pathway=\", paste(all.path$path_id, sep=\" \", collapse = \" \")), \"all.path.txt\", row.names = FALSE, col.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd3ea4dd",
   "metadata": {},
   "source": [
    "# Using Kegg api for Kegg annotations \n",
    "Imported csv files of modules, paths, and individual ko's, which are not associated with a path or module, have associated subcategories and broad categories which I am interested in. To get the module or path name, all of the ko's and thier names and symbols from the module or path code, we loop through each module/path code and using Kegg api, extract this information. Because pathways and modules are associated with one annother, we will keep these tables seperate\n",
    "## After running each loop, we should have a dataframe for each original file with a columns:\n",
    "- ko_id\n",
    "- name\n",
    "- symbol\n",
    "- pathway/module\n",
    "- sub_category\n",
    "- broad_category\n",
    "## Extract all ko's from tables which appear in at least one sample\n",
    "We will use the resulting tables for subsetting data later in making heat maps. We can create heat maps based on pathway, module, or category. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7c99e65-77c5-48fa-ab8d-92d05727f870",
   "metadata": {},
   "outputs": [],
   "source": [
    "mod_man=read.csv('../kegg_names//subcategories_ko.csv')[,1:3]\n",
    "head(mod_man)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfe19184-1b23-48ed-aba4-b89b66adc975",
   "metadata": {},
   "outputs": [],
   "source": [
    "#1. for manually made modules ko's\n",
    "## find name and symbol for each\n",
    "## add name and symbol to dataframe\n",
    "mod_man=read.csv('../kegg_names//subcategories_ko.csv')[,1:3]\n",
    "dim(mod_man)\n",
    "url=c(\"https://rest.kegg.jp/\")\n",
    "symbol=list()\n",
    "ko_id=list()\n",
    "name=list()\n",
    "\n",
    "for (i in mod_man$ko_id){\n",
    "    #m=c(keggFind('ko',i))\n",
    "    m = getURL(paste(url,'find/ko/', i, sep=''))\n",
    "    m = str_remove(m, 'ko:K[[:digit:]]{5}\\t')\n",
    "    ko_id=c(ko_id,i)\n",
    "    symbol=c(symbol, str_remove(str_extract(string = m, pattern = '.+.*;'),';'))\n",
    "    name=c(name,str_remove(str_extract(string = m, pattern = '; [[:alnum:]]+.*'),'; '))\n",
    "    }\n",
    "\n",
    "mod_ko=data.frame('ko_id'=unlist(ko_id),\n",
    "                  'symbol'= unlist(symbol),\n",
    "                  'name'= str_remove(unlist(name),'\\n'))\n",
    "mod_man=left_join(mod_man,mod_ko, by='ko_id')\n",
    "mod_man=mutate(mod_man, Path=NA, path_name=mod_man$sub_category)\n",
    "head(mod_man)\n",
    "tail(mod_man)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bfbca2a-d18c-44b6-824b-d1a4549418d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#2. Kegg modules\n",
    "## Get name for each module\n",
    "## add module name to dataframe\n",
    "## get list of all ko's for each module\n",
    "## get ko name and symbol for each module\n",
    "sub_module=read.csv(\"../kegg_names/subcategories_module.csv\")\n",
    "nrow(sub_module)\n",
    "url=c(\"https://rest.kegg.jp/\")\n",
    "\n",
    "find.mod=list()\n",
    "for (i in sub_module$Module){\n",
    "    #p=keggFind('module', i)\n",
    "    p=getURL(paste(url,'find/module/', i, sep=''))\n",
    "    find.mod=unname(c(find.mod,str_remove(p, \", .*\")))\n",
    "    #module=unname(c(module,str_remove(p, \", .*\"))) #str_extract(p, '[[:alpha:]]+.*'))\n",
    "}\n",
    "\n",
    "module_name=data.frame('module_name'=str_remove(find.mod,'md:M[[:digit:]]{5}\\t'))\n",
    "sub_mod=bind_cols(sub_module, module_name)\n",
    "colnames(sub_mod)[1]='Module'\n",
    "\n",
    "p=list()\n",
    "for (i in sub_mod$Module){\n",
    "    link.ko=getURL(paste(url,'link/ko/', i, sep=''))\n",
    "    p=c(p,link.ko)\n",
    "    #p=c(p,keggLink('ko', i))\n",
    "}\n",
    "\n",
    "p=data.frame(ko_id=unlist(str_extract_all(p,'ko:K[[:digit:]]+')), \n",
    "               module=(unlist(str_extract_all(p, 'M[[:digit:]]{5}'))))\n",
    "\n",
    "symbol=list()\n",
    "name=list()\n",
    "for (i in p$ko_id){\n",
    "     g=getURL(paste(url,'find/ko/', i, sep=''))\n",
    "   # g=keggFind('ko', i)\n",
    "    symbol=c(symbol,str_remove(str_extract(unname(g), '\\t.+;'), ';'))\n",
    "    name=c(name, str_remove(str_extract(unname(g),'; .*'), '; '))\n",
    "}\n",
    "\n",
    "mod_ko=data.frame('Module'=p$module,\n",
    "                  'ko_id'=p$ko_id,\n",
    "                  'symbol'=str_remove(unlist(symbol),'\\t'),\n",
    "                  'name'=unlist(name))\n",
    "sub_mod=left_join(mod_ko, sub_mod, by='Module')\n",
    "head(sub_mod)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1e3cf27-d732-4eb3-89d1-662c3ee20975",
   "metadata": {},
   "source": [
    "## Pulling out pathways used for heatmaps in paper to make loop faster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "951e4edf-2ca0-40b2-a097-382db7bbbcac",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_cat_path=read.csv('../kegg_names/subcategories_path.csv')\n",
    "\n",
    "heatMapPath = filter(sub_cat_path, sub_category %in% c(\n",
    "    'Carbon fixation in photosynthetic organisms', 'Photosynthesis', \n",
    "    'Photosynthesis-antenna proteins', 'Nitrogen metabolism', 'Carotenoid biosynthesis'))\n",
    "heatMapPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "979b4944-1ce8-4727-bcc9-0209a3350c31",
   "metadata": {},
   "outputs": [],
   "source": [
    "#3. pathways\n",
    "## get name for each pathway\n",
    "## add pathway name to dataframe\n",
    "## get list of all ko's for each pathway\n",
    "## get name and symbol for each ko\n",
    "\n",
    "\n",
    "url=c(\"https://rest.kegg.jp/\")\n",
    "find.path=list()\n",
    "for (i in heatMapPath$Pathway){\n",
    "    p=getURL(paste(url,'find/pathway/', i, sep=''))\n",
    "    find.path=unname(c(find.path,p))\n",
    "}\n",
    "\n",
    "path_name=data.frame('path_name'=str_remove(find.path,'path:map[[:digit:]]{5}\\t'))\n",
    "path_name$path_name=str_remove(path_name$path_name, '\\n')\n",
    "head(path_name)\n",
    "sub_path=bind_cols(heatMapPath, path_name)\n",
    "colnames(sub_path)[1]='Path'\n",
    "\n",
    "p=list()\n",
    "for (i in sub_path$Path){\n",
    "    link.ko=getURL(paste(url,'link/ko/', i, sep=''))\n",
    "    #link.ko=str_extract(p, 'ko:K[[:digit:]]+')\n",
    "    p=c(p,link.ko)\n",
    "}\n",
    "\n",
    "p=data.frame(ko_id=unlist(str_extract_all(p,'K[[:digit:]]+')), \n",
    "               path=(unlist(str_extract_all(p, 'map[[:digit:]]{5}')))) \n",
    "symbol=list()\n",
    "name=list()\n",
    "for (i in p$ko_id){\n",
    "    g=getURL(paste(url,'find/ko/', i, sep=''))\n",
    "    symbol=c(symbol,str_remove(str_extract(unname(g), '\\t.+;'), ';'))\n",
    "    name=c(name, str_remove(str_extract(unname(g),'; .*'), '; '))\n",
    "}\n",
    "path_ko=data.frame('Path'=p$path,\n",
    "                   'ko_id'=p$ko_id,\n",
    "                   'symbol'=str_remove(unlist(symbol),'\\t'),\n",
    "                   'name'= str_remove(unlist(name), '\\\\[EC:.*\\\\]'))\n",
    "\n",
    "sub_path=left_join(path_ko, sub_path, by='Path', relationship='many-to-many')\n",
    "\n",
    "head(sub_path)\n",
    "dim(sub_path)\n",
    "\n",
    "## add fld genes \n",
    "flavodoxin=data.frame(Path='map00195', ko_id=c('K03839','K03840','K21567','K00528'), \n",
    "                      symbol=c('fldA, nifF, isiB','fldB','fnr','fpr'), \n",
    "                      name=c('flavodoxin I','flavodoxin II','ferredoxin/flavodoxin---NADP+ reductase',\n",
    "                            'ferredoxin/flavodoxin---NADP+ reductase'), sub_category='Photosynthesis',\n",
    "                     broad_category='photosynthesis', path_name='Photosynthesis')\n",
    "sub_path=bind_rows(sub_path, flavodoxin) \n",
    "\n",
    "tail(sub_path)\n",
    "dim(sub_path)\n",
    "filter(sub_path, path_name=='Photosynthesis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e20c6ea9-c068-4428-9bf1-62dc1ea5e833",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_path$path_name[sub_path$path_name=='Photosynthesis - antenna proteins']='Photosynthesis'\n",
    "\n",
    "sub_path[str_detect(sub_path$symbol, \"DUF\"), 'sub_category'] =  \"NADH dehydrogenase\"\n",
    "sub_path[str_detect(sub_path$symbol, \"COX|cox|CYC\"), 'sub_category'] = \"cytochrome c\"\n",
    "sub_path[str_detect(sub_path$symbol, \"ATP.*V\"),'sub_category'] = \"V-Type ATP-ase\"\n",
    "sub_path[str_detect(sub_path$symbol, \"ATP.*F\"), 'sub_category'] = \"F-Type ATP-ase\"\n",
    "\n",
    "heat.path=(filter(sub_path, (ko_id %in% ko_def$ko_id)==T))\n",
    "\n",
    "heat.path$name= str_replace_all(heat.path$name, c('light-harvesting complex I '='LHCA ','light-harvesting complex II'='LHCB',  \n",
    "                                 'chlorophyll'='Chl','photosystem I '='PSI ', 'photosystem II'='PSII',\n",
    "                                 'F-type .* subunit '='F-Type ATP-ase ', 'isip_'='Iron starvation induced protein ',\n",
    "                                 'fructose-bisphosphate aldolase'='FBA', 'ribulose-bisphosphate carboxylase'='RuBisCO',\n",
    "                                               'MFS transporter, NNP family, '=''))\n",
    "\n",
    "urea_cycle = filter(ko_def, ko_id%in%c('K00611','K01940','K01755','K01476'))\n",
    "urea_cycle = mutate(urea_cycle, 'sub_category'='Urea cycle', 'broad_category'='Nitrogen metabolism','path_name'='Nitrogen metabolism', 'Path'='map00910')\n",
    "heat.path = full_join(heat.path, urea_cycle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95c9ac8a-1e1e-4c73-950a-136add025d5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "heat.path[str_detect(heat.path$symbol, 'MDH1'), 'name'] = 'malate dehydrogenase 1'\n",
    "heat.path[str_detect(heat.path$symbol, 'MDH2'), 'name'] = 'malate dehydrogenase 2'\n",
    "heat.path[str_detect(heat.path$symbol, 'maeB|ppdK'), 'sub_category'] = 'CAM light'\n",
    "heat.path[str_detect(heat.path$symbol, 'MDH1|MDH2|mdh|ppc'), 'sub_category'] = 'CAM dark'\n",
    "heat.path[heat.path$ko_id %in% \n",
    "          c('K00855','K00927','K01100','K01601','K01623','K01624','K01783',\n",
    "            'K01803','K01807','K01808','K02446','K03841','K11532','K00134',\n",
    "            'K00615'), 'sub_category'] = 'Calvin cycle' \n",
    "heat.path$sub_category = str_replace(heat.path$sub_category, 'Carbon fixation in photosynthetic organisms', 'C4 Dicarboxilic acid cycle')\n",
    "heat.path$name = str_remove(heat.path$name, '.(phosphorylating).')\n",
    "filter(heat.path, path_name=='Carbon fixation in photosynthetic organisms')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dae01e7-953c-4123-af59-69cfdd846cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "heat.path[str_detect(heat.path$name, 'carbonic anhydrase'), 'sub_category'] = 'Carbonic anhydrase'\n",
    "heat.path[str_detect(heat.path$name, 'nitrate/nitrite transport'), 'sub_category'] = 'Nitrogen transporters'\n",
    "heat.path[str_detect(heat.path$name, 'glutamate|glutamine'), 'sub_category'] = 'GS/GOGAT and GDH'\n",
    "heat.path[str_detect(heat.path$name, 'nitrite reductase'), 'sub_category'] = 'Nitrite reductase'\n",
    "heat.path[str_detect(heat.path$name, 'nitrate reductase'), 'sub_category'] = 'Nitrite reductase'\n",
    "heat.path[str_detect(heat.path$symbol, 'CPS1'), 'sub_category'] = 'Urea cycle'\n",
    "heat.path$name = str_remove(heat.path$name, '\\\\[.*\\\\]')\n",
    "filter(heat.path, path_name=='Nitrogen metabolism')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb6ef7b5-6a4d-4990-a298-cfa272404f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "heat.path[str_detect(heat.path$symbol, 'psa'), 'sub_category'] = 'PSI'\n",
    "heat.path[str_detect(heat.path$symbol, 'psb'), 'sub_category'] = 'PSII'\n",
    "heat.path[str_detect(heat.path$symbol, 'LHCA'), 'sub_category'] = 'LHCA'\n",
    "heat.path[str_detect(heat.path$symbol, 'LHCB'), 'sub_category'] = 'LHCB'\n",
    "heat.path$sub_category = str_replace(heat.path$sub_category, 'Photosynthesis', 'Electron transport chain')\n",
    "heat.path[str_detect(heat.path$sub_category, 'F-Type ATP-ase'), 'path_name'] = 'F-Type ATP-ase'\n",
    "\n",
    "filter(heat.path, path_name=='Photosynthesis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f467086d-d727-498f-a04b-cd348849b944",
   "metadata": {},
   "outputs": [],
   "source": [
    "heat.path[str_detect(heat.path$name, 'violaxanthin|zeaxanthin'), 'path_name'] = 'Photosynthesis'\n",
    "heat.path[str_detect(heat.path$name, 'violaxanthin|zeaxanthin'), 'sub_category'] = 'Xanthophyll cycle'\n",
    "\n",
    "filter(heat.path, path_name=='Photosynthesis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9f64a83-9399-41ca-8c97-9b2911b9dde9",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique(heat.path$path_name)\n",
    "write.csv(heat.path, '../kegg_names/pathwaysHeatMap.csv', row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4085b63a-58e3-4ccb-ae0c-776ed0bad228",
   "metadata": {},
   "outputs": [],
   "source": [
    "all(heat.path$ko_id %in% ko_def$ko_id)\n",
    "dim(heat.path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43b7eb78-fade-422b-b861-f0e5024f3365",
   "metadata": {},
   "source": [
    "## Combine kegg module and manual module tables "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23de9c26-07d9-460e-9328-31e3192c0521",
   "metadata": {},
   "source": [
    "### Now I have mod/path code and name, sub and broad categorie and ko_id, name and symbol for each\n",
    "using the ko_id find matches in my data, some ko's might be repeated in different categories so make sure they are repeated in each category\n",
    "\n",
    "## in the end I want a df with links of ko's in my data, name, symbol, path/module code and name, subcategory, broad category\n",
    "i can use this when refering to data for plotting\n",
    "## Make sure to check that genes i am interested in like lhp and ferredoxin actually showed up! check differences in length between original list and those in data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "107d7dc0-d423-4aad-ac2d-7584ed554a1a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ko4=read.csv('../kegg_names/ko4_ls.csv')\n",
    "ko8=read.csv('../kegg_names/ko8_ls.csv')\n",
    "ko6=read.csv('../kegg_names/ko6_ls.csv')\n",
    "ko13=read.csv('../kegg_names/ko13_ls.csv')\n",
    "head(ko13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aba63206-ccbe-415e-906a-2ab41fa52dac",
   "metadata": {},
   "outputs": [],
   "source": [
    "a=unique(ko4$ko_id)\n",
    "a=a[a%in%heat.path$ko_id]\n",
    "\n",
    "b=unique(ko8$ko_id)\n",
    "b=b[b%in%heat.path$ko_id]\n",
    "\n",
    "c=unique(ko6$ko_id)\n",
    "c=c[c%in%heat.path$ko_id]\n",
    "\n",
    "d=unique(ko13$ko_id)\n",
    "d=d[d%in%heat.path$ko_id]\n",
    "\n",
    "\n",
    "length(b)\n",
    "length(c)\n",
    "length(d)\n",
    "length(a)\n",
    "x=union(a,b)\n",
    "y=union(c,d)\n",
    "z=union(x,y)\n",
    "\n",
    "length(x)\n",
    "length(y)\n",
    "length(z)\n",
    "st=heat.path[(heat.path$ko_id)%in%z,]\n",
    "nrow(st)\n",
    "length(unique(heat.path$ko_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "766da7ee-2603-43d4-acb0-625ea4b20fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "mod.path=function(df,ko){\n",
    "    inner_join(x = ko, y=df,by = 'ko_id', relationship='many-to-many')\n",
    "    }\n",
    "\n",
    "path4=mod.path(sub_path, ko4)\n",
    "path8=mod.path(sub_path, ko8)\n",
    "path6=mod.path(sub_path, ko6)\n",
    "path13=mod.path(sub_path, ko13)\n",
    "\n",
    "write.csv(path4,'../kegg_names/path4.csv',row.names=FALSE)\n",
    "write.csv(path8,'../kegg_names/path8.csv',row.names=FALSE)\n",
    "write.csv(path6,'../kegg_names/path6.csv',row.names=FALSE)\n",
    "write.csv(path13,'../kegg_names/path13.csv',row.names=FALSE)\n",
    "\n",
    "sub_path=bind_rows(path4,path8,path6,path13)\n",
    "sub_path=sub_path[,-c(1:2)] %>% distinct()\n",
    "\n",
    "write.csv(all.path,'../kegg_names/all.path.csv', row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1904f47d-08c0-428b-948a-35196b161fb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mod4=mod.path(sub_mod_all, ko4)\n",
    "mod8=mod.path(sub_mod_all, ko8)\n",
    "mod6=mod.path(sub_mod_all, ko6)\n",
    "mod13=mod.path(sub_mod_all, ko13)\n",
    "\n",
    "write.csv(mod4,'../kegg_names/mod4.csv',row.names=FALSE)\n",
    "write.csv(mod8,'../kegg_names/mod8.csv',row.names=FALSE)\n",
    "write.csv(mod6,'../kegg_names/mod6.csv',row.names=FALSE)\n",
    "write.csv(mod13,'../kegg_names/mod13.csv',row.names=FALSE)\n",
    "\n",
    "#now rewriting sub_cat table so only genes found in my organisms show up\n",
    "all.mod=bind_rows(mod4,mod8,mod6,mod13)\n",
    "all.mod=all.mod[,-c(1:2)] %>% distinct()\n",
    "\n",
    "write.csv(all.mod,'../kegg_names/all.mod.csv', row.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f66db89-72ab-42a6-a435-1031081bd95b",
   "metadata": {},
   "source": [
    "----\n",
    "# Old way of extracting pathways of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9118847",
   "metadata": {},
   "outputs": [],
   "source": [
    "## List of pathways of interest with ko-id's associated. \n",
    "## find orfs and counts within pathways and plot on heatmap\n",
    "## add name of pathway in df by joinging pathways dataframes\n",
    "\n",
    "select.paths <- data.frame(Name=c('Nitrogen Metabolism', \n",
    "                                'Photosynthesis', \n",
    "                                'Oxidative Phosphorylation', \n",
    "                                'Photosynthesis-antenna Proteins', \n",
    "                                'Carbon Fixation',\n",
    "                               'Methan Metabolism', \n",
    "                                'Sulfur Metabolism',\n",
    "                                'Citrate Cycle', \n",
    "                                'Glycolysis', \n",
    "                                'ABC Transport', \n",
    "                                'Phosphotransferase System'),\n",
    "                      path_id=c('map00910', 'map00195', 'map00190','map00196', 'map00710',\n",
    "                              'map00680', 'map00920', 'map00020', 'map00010', 'map02010', 'map02060'))\n",
    "#keggLink('ko', 'map00910')\n",
    "\n",
    "code <- list(select.paths$path_id)\n",
    "head(code)\n",
    "for (i in code){\n",
    "   #print(keggList(i))\n",
    "    pathways <- as.matrix(keggLink('ko', i))\n",
    "}\n",
    "pathways <- data.frame('ko_id'=pathways) %>% rownames_to_column('path_id')  \n",
    "pathways$path_id <- gsub('path.', '', pathways$path_id)\n",
    "pathways$ko_id <- gsub('ko:','', pathways$ko_id)\n",
    "pathways$path_id <- str_remove(pathways$path_id, '\\\\.[[:digit:]]+')\n",
    "pathways <- pathways %>% group_by(path_id)\n",
    "\n",
    "select.pathways <- left_join(pathways, select.paths, by='path_id')\n",
    "write.csv(select.pathways, '../kegg_names/select.pathways.csv', row.names=FALSE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
