{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9171cae8",
   "metadata": {},
   "source": [
    "# Kegg annotation tables \n",
    "\n",
    "- This notebook will use output from Eggnogmapper to create key tables used to sum ORF counts to the 'Kegg Ko gene level' and to translate Ko IDs into gene names for visualization. \n",
    "---\n",
    "## Prepare Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "569719fe-78a8-4e44-8744-a2c9eb9cf706",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ────────────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ──────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "\n",
      "Attaching package: ‘RCurl’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    complete\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library('tidyverse')\n",
    "library('tibble') \n",
    "library('KEGGREST')\n",
    "library('RCurl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "070bb3bf",
   "metadata": {},
   "source": [
    "# 1. Read in eggnog annotation tables\n",
    "---\n",
    "The `get.eggnog` function reads in eggnog annotation tables. Changing the directory and file name to the organism number (4,8,6 or 13) is specific to how I organized my files. The only columns of interest are subset from the annotation table.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "66403180",
   "metadata": {},
   "outputs": [],
   "source": [
    "get.eggnog <- function(org){\n",
    "    emap <- read.csv(paste('../expression_analysis/eggnog_annotations/',org,\".emapper.annotations\", sep=\"\"),\n",
    "                 sep = \"\\t\",\n",
    "                 comment.char = \"#\",\n",
    "                 header = FALSE,\n",
    "                 na.strings = \"-\")\n",
    "    \n",
    "    colnames(emap) <- c(\n",
    "        'orfs', 'seed_ortholog', 'evalue', 'score', 'eggNOG_OGs', 'max_annot_lvl', 'COG_category', 'Description', \n",
    "        'Preferred_name', 'GOs', 'EC', 'ko_id', 'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction', \n",
    "        'KEGG_rclass',  'BRITE', 'KEGG_TC', 'CAZy', 'BiGG_Reaction', 'PFAMs')\n",
    "    \n",
    "    emap[, c(\"orfs\", \"seed_ortholog\", \"EC\",\"GOs\",\"ko_id\",'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction', \n",
    "       'KEGG_rclass',  'BRITE', 'KEGG_TC', 'CAZy', 'BiGG_Reaction',\"PFAMs\")]\n",
    "    }\n",
    "\n",
    "emap4 <- get.eggnog('04')\n",
    "emap8 <- get.eggnog('08')\n",
    "emap6 <- get.eggnog('06')\n",
    "emap13 <- get.eggnog('13')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f41c467",
   "metadata": {},
   "source": [
    "### 1.2 Create df for each organism with annotation of choice, removing the rows with NA's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bf728468",
   "metadata": {},
   "outputs": [],
   "source": [
    "get.anno <- function(emap, anno){\n",
    "    #extract query column and annotation of choice, ex: pfams\n",
    "    vars <- c(\"orfs\", anno)\n",
    "    df <- select(emap, all_of(vars))\n",
    "    df <- df %>% filter(is.na(df[,ncol(df)])==FALSE)\n",
    "    }\n",
    "\n",
    "ko4_ls <- get.anno(emap4,\"ko_id\")\n",
    "ko8_ls <- get.anno(emap8,\"ko_id\")\n",
    "ko6_ls <- get.anno(emap6,\"ko_id\")\n",
    "ko13_ls <- get.anno(emap13,\"ko_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7b85a4ef-cb81-406c-8760-b6e772427773",
   "metadata": {},
   "outputs": [],
   "source": [
    "pfam4_ls = get.anno(emap4, 'PFAMs')\n",
    "pfam8_ls = get.anno(emap8, 'PFAMs')\n",
    "pfam6_ls = get.anno(emap6, 'PFAMs')\n",
    "pfam13_ls = get.anno(emap13, 'PFAMs')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8aa3b7a",
   "metadata": {},
   "source": [
    "# 2. Create a ko-to-orf mapping table\n",
    "---\n",
    "The ko-to-orf mapping table will associate each ORF to a Kegg Ko, and can be used as a key. Because some ORFs were assigned multiple Ko's, these will be split into multiple columns with the `split into multiple` function adapted from a stack overflow post. ORFs with multiple Ko's will appear as many times as a Ko was assigned and the ko_iteration numbers from 1-n ko's assigned. \n",
    "Then, `clean_ko` pivots the table into three columns, ORF, Ko, and Ko_iteration, dropping the Ko prefix 'Ko:'.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c1440559",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"# of values in matrix w/o NA: 14281\"\n",
      "[1] \"# rows in final df:  14281\"\n",
      "[1] \"# of values in matrix w/o NA: 35251\"\n",
      "[1] \"# rows in final df:  35251\"\n",
      "[1] \"# of values in matrix w/o NA: 12765\"\n",
      "[1] \"# rows in final df:  12765\"\n",
      "[1] \"# of values in matrix w/o NA: 30911\"\n",
      "[1] \"# rows in final df:  30911\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>orfs</th><th scope=col>ko_iteration</th><th scope=col>ko_id</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>NODE_10001_length_2060_cov_5.581278_g5365_i0.p1 </td><td>ko:_1</td><td>K00914</td></tr>\n",
       "\t<tr><td>NODE_10002_length_2059_cov_37.920947_g5366_i0.p1</td><td>ko:_1</td><td>K06911</td></tr>\n",
       "\t<tr><td>NODE_10003_length_2059_cov_28.617321_g5366_i1.p1</td><td>ko:_1</td><td>K06911</td></tr>\n",
       "\t<tr><td>NODE_10009_length_2059_cov_5.150050_g5369_i0.p1 </td><td>ko:_1</td><td>K21797</td></tr>\n",
       "\t<tr><td>NODE_10010_length_2058_cov_58.211083_g5370_i0.p1</td><td>ko:_1</td><td>K18038</td></tr>\n",
       "\t<tr><td>NODE_10011_length_2058_cov_52.328967_g4722_i3.p3</td><td>ko:_1</td><td>K03020</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 3\n",
       "\\begin{tabular}{lll}\n",
       " orfs & ko\\_iteration & ko\\_id\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t NODE\\_10001\\_length\\_2060\\_cov\\_5.581278\\_g5365\\_i0.p1  & ko:\\_1 & K00914\\\\\n",
       "\t NODE\\_10002\\_length\\_2059\\_cov\\_37.920947\\_g5366\\_i0.p1 & ko:\\_1 & K06911\\\\\n",
       "\t NODE\\_10003\\_length\\_2059\\_cov\\_28.617321\\_g5366\\_i1.p1 & ko:\\_1 & K06911\\\\\n",
       "\t NODE\\_10009\\_length\\_2059\\_cov\\_5.150050\\_g5369\\_i0.p1  & ko:\\_1 & K21797\\\\\n",
       "\t NODE\\_10010\\_length\\_2058\\_cov\\_58.211083\\_g5370\\_i0.p1 & ko:\\_1 & K18038\\\\\n",
       "\t NODE\\_10011\\_length\\_2058\\_cov\\_52.328967\\_g4722\\_i3.p3 & ko:\\_1 & K03020\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 3\n",
       "\n",
       "| orfs &lt;chr&gt; | ko_iteration &lt;chr&gt; | ko_id &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| NODE_10001_length_2060_cov_5.581278_g5365_i0.p1  | ko:_1 | K00914 |\n",
       "| NODE_10002_length_2059_cov_37.920947_g5366_i0.p1 | ko:_1 | K06911 |\n",
       "| NODE_10003_length_2059_cov_28.617321_g5366_i1.p1 | ko:_1 | K06911 |\n",
       "| NODE_10009_length_2059_cov_5.150050_g5369_i0.p1  | ko:_1 | K21797 |\n",
       "| NODE_10010_length_2058_cov_58.211083_g5370_i0.p1 | ko:_1 | K18038 |\n",
       "| NODE_10011_length_2058_cov_52.328967_g4722_i3.p3 | ko:_1 | K03020 |\n",
       "\n"
      ],
      "text/plain": [
       "  orfs                                             ko_iteration ko_id \n",
       "1 NODE_10001_length_2060_cov_5.581278_g5365_i0.p1  ko:_1        K00914\n",
       "2 NODE_10002_length_2059_cov_37.920947_g5366_i0.p1 ko:_1        K06911\n",
       "3 NODE_10003_length_2059_cov_28.617321_g5366_i1.p1 ko:_1        K06911\n",
       "4 NODE_10009_length_2059_cov_5.150050_g5369_i0.p1  ko:_1        K21797\n",
       "5 NODE_10010_length_2058_cov_58.211083_g5370_i0.p1 ko:_1        K18038\n",
       "6 NODE_10011_length_2058_cov_52.328967_g4722_i3.p3 ko:_1        K03020"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "split_into_multiple <- function(column, pattern = \",\", into_prefix){\n",
    "    #adapted from post on stack overflow\n",
    "    cols <- str_split_fixed(column, pattern, n = Inf)\n",
    "  # Replace empty matrix indicies with NA's \n",
    "    cols[which(cols == \"\")] <- NA\n",
    "  # turn matrix into a table with unique but arbitraty column names  \n",
    "    cols <- as_tibble(cols, .name_repair = make.names)\n",
    "  # where m = # columns in tibble 'cols'\n",
    "    m <- dim(cols)[2]\n",
    "  # assign column names as 'into_prefix_1', 'into_prefix_2', ..., 'into_prefix_m' \n",
    "    names(cols) <- paste(into_prefix, 1:m, sep = \"_\")\n",
    "    print(paste('# of values in matrix w/o NA:',sum(!is.na(cols)),sep=' '))\n",
    "    return(cols)\n",
    "}\n",
    "clean_ko <- function(df, org, into_prefix, value_col='ko_id',name_col='ko_iteration'){\n",
    "  # split up ko_id's into multiple columns naming each column \n",
    "  # ko:_1 to ko:_n \n",
    "  # (remember at this point we have the same number of rows but \n",
    "  # far more columns)\n",
    "    anno_iterations <- split_into_multiple(df[,2], \",\", into_prefix)\n",
    "  # select the orfs column from original df and bind to to \n",
    "  # split columns \n",
    "    df = df %>% select(orfs) %>% bind_cols(anno_iterations)\n",
    "  # now combine all ko:_n columns so all ko_id's \n",
    "  # become one column called ko_id and each column name, \n",
    "  # 'ko:_1'...'ko:_n' becomes one column called 'ko_iteration'. \n",
    "  # This tells us how many ko_id's were assigned to a particular orf\n",
    "  # the column 'orfs' will repeat values for rows with > 1 ko_iteration\n",
    "  # Drop the values from the matrix which were NA\n",
    "    df_clean = pivot_longer(df, cols = !orfs, values_drop_na = T,\n",
    "                            values_to = value_col, names_to = name_col)\n",
    "  # finally, clean up the df a bit by removing the ko_id prefixes 'ko:'\n",
    "  # we now have a 3 column table with orfs repeated in for each instance \n",
    "  # a ko_id was assigned to it by eggnog!\n",
    "    if(str_detect(colnames(df_clean[3]),'ko_id')==T){\n",
    "    df_clean$ko_id <- gsub(into_prefix, '', df_clean$ko_id)\n",
    "        }\n",
    "  # the final df should have same number of rows as sum(!na(cols))\n",
    "    print(paste('# rows in final df:',nrow(df_clean), sep='  '))\n",
    "    #write.csv(df_clean, paste('../kegg_names/ko',org, '_ls.csv', sep=''), row.names=F)\n",
    "    df_clean\n",
    "    }\n",
    "\n",
    "ko4 <- clean_ko(ko4_ls,'4','ko:')\n",
    "ko6 <- clean_ko(ko6_ls,'6','ko:')\n",
    "ko8 <- clean_ko(ko8_ls,'8','ko:')\n",
    "ko13 <- clean_ko(ko13_ls,'13','ko:')\n",
    "head(ko4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0d8cd705-031e-4f22-97fc-91703f2b27f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"# of values in matrix w/o NA: 44311\"\n",
      "[1] \"# rows in final df:  44311\"\n",
      "[1] \"# of values in matrix w/o NA: 37925\"\n",
      "[1] \"# rows in final df:  37925\"\n",
      "[1] \"# of values in matrix w/o NA: 68828\"\n",
      "[1] \"# rows in final df:  68828\"\n",
      "[1] \"# of values in matrix w/o NA: 61554\"\n",
      "[1] \"# rows in final df:  61554\"\n"
     ]
    }
   ],
   "source": [
    "pfam4 = clean_ko(pfam4_ls, '4','pfam', value_col = 'pfam_id',name_col = 'pfam_iteration')\n",
    "pfam8 = clean_ko(pfam8_ls, '8','pfam', value_col = 'pfam_id',name_col = 'pfam_iteration')\n",
    "pfam6 = clean_ko(pfam6_ls, '6','pfam', value_col = 'pfam_id',name_col = 'pfam_iteration')\n",
    "pfam13 = clean_ko(pfam13_ls, '13','pfam', value_col = 'pfam_id',name_col = 'pfam_iteration')\n",
    "\n",
    "write.csv(pfam4, '../expression_analysis/eggnog_annotations/pfam4.csv', row.names=F)\n",
    "write.csv(pfam8, '../expression_analysis/eggnog_annotations/pfam8.csv', row.names=F)\n",
    "write.csv(pfam6, '../expression_analysis/eggnog_annotations/pfam6.csv', row.names=F)\n",
    "write.csv(pfam13, '../expression_analysis/eggnog_annotations/pfam13.csv', row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b16d6e16-30e3-4f18-b89c-f37813885f2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rho4 = filter(pfam4, pfam_id =='7tm_1') %>% mutate('org'='4')\n",
    "rho6 = filter(pfam6, pfam_id =='7tm_1')%>% mutate('org'='6')\n",
    "rho13 = filter(pfam13, pfam_id =='7tm_1')%>% mutate('org'='13')\n",
    "rho_pfam = bind_rows(rho4,rho6,rho13)\n",
    "write.csv(rho_pfam, '../expression_analysis/eggnog_annotations/rho_pfam.csv', row.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cca5c9",
   "metadata": {},
   "source": [
    "## 2.1 Add ISIP transcript ORFs \n",
    "The tables created above have tidy df of query (ORF) and annotation (Ko_id) for mapping to counts, but need the ISIP annotations. To do this I: \n",
    "- Read in ISIP isoform annotations\n",
    "- Make a tidy df of all ISIP isoforms for each organism\n",
    "- Check that ISIP's were also counted by Salmon, remove any which were not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5863d89e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orfs</th><th scope=col>ko_iteration</th><th scope=col>ko_id</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>V1</th><td>NODE_11480_length_1968_cov_823.584169_g5802_i0.p2 </td><td>ko:_1</td><td>isip_2a</td></tr>\n",
       "\t<tr><th scope=row>V2</th><td>NODE_12355_length_1900_cov_867.879037_g5802_i1.p2 </td><td>ko:_1</td><td>isip_2a</td></tr>\n",
       "\t<tr><th scope=row>V3</th><td>NODE_14509_length_1755_cov_775.599287_g5802_i2.p2 </td><td>ko:_1</td><td>isip_2a</td></tr>\n",
       "\t<tr><th scope=row>V4</th><td>NODE_1485_length_3787_cov_81.809908_g739_i0.p1    </td><td>ko:_1</td><td>isip_2a</td></tr>\n",
       "\t<tr><th scope=row>V5</th><td>NODE_19097_length_1503_cov_1567.323776_g9985_i0.p2</td><td>ko:_1</td><td>isip_2a</td></tr>\n",
       "\t<tr><th scope=row>V6</th><td>NODE_29963_length_1047_cov_1763.823409_g3409_i1.p3</td><td>ko:_1</td><td>isip_2a</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & orfs & ko\\_iteration & ko\\_id\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tV1 & NODE\\_11480\\_length\\_1968\\_cov\\_823.584169\\_g5802\\_i0.p2  & ko:\\_1 & isip\\_2a\\\\\n",
       "\tV2 & NODE\\_12355\\_length\\_1900\\_cov\\_867.879037\\_g5802\\_i1.p2  & ko:\\_1 & isip\\_2a\\\\\n",
       "\tV3 & NODE\\_14509\\_length\\_1755\\_cov\\_775.599287\\_g5802\\_i2.p2  & ko:\\_1 & isip\\_2a\\\\\n",
       "\tV4 & NODE\\_1485\\_length\\_3787\\_cov\\_81.809908\\_g739\\_i0.p1     & ko:\\_1 & isip\\_2a\\\\\n",
       "\tV5 & NODE\\_19097\\_length\\_1503\\_cov\\_1567.323776\\_g9985\\_i0.p2 & ko:\\_1 & isip\\_2a\\\\\n",
       "\tV6 & NODE\\_29963\\_length\\_1047\\_cov\\_1763.823409\\_g3409\\_i1.p3 & ko:\\_1 & isip\\_2a\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | orfs &lt;chr&gt; | ko_iteration &lt;chr&gt; | ko_id &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| V1 | NODE_11480_length_1968_cov_823.584169_g5802_i0.p2  | ko:_1 | isip_2a |\n",
       "| V2 | NODE_12355_length_1900_cov_867.879037_g5802_i1.p2  | ko:_1 | isip_2a |\n",
       "| V3 | NODE_14509_length_1755_cov_775.599287_g5802_i2.p2  | ko:_1 | isip_2a |\n",
       "| V4 | NODE_1485_length_3787_cov_81.809908_g739_i0.p1     | ko:_1 | isip_2a |\n",
       "| V5 | NODE_19097_length_1503_cov_1567.323776_g9985_i0.p2 | ko:_1 | isip_2a |\n",
       "| V6 | NODE_29963_length_1047_cov_1763.823409_g3409_i1.p3 | ko:_1 | isip_2a |\n",
       "\n"
      ],
      "text/plain": [
       "   orfs                                               ko_iteration ko_id  \n",
       "V1 NODE_11480_length_1968_cov_823.584169_g5802_i0.p2  ko:_1        isip_2a\n",
       "V2 NODE_12355_length_1900_cov_867.879037_g5802_i1.p2  ko:_1        isip_2a\n",
       "V3 NODE_14509_length_1755_cov_775.599287_g5802_i2.p2  ko:_1        isip_2a\n",
       "V4 NODE_1485_length_3787_cov_81.809908_g739_i0.p1     ko:_1        isip_2a\n",
       "V5 NODE_19097_length_1503_cov_1567.323776_g9985_i0.p2 ko:_1        isip_2a\n",
       "V6 NODE_29963_length_1047_cov_1763.823409_g3409_i1.p3 ko:_1        isip_2a"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## read in the files\n",
    "dir <- \"../expression_analysis/eggnog_annotations/\"\n",
    "\n",
    "read.isip <- function(org, prot){\n",
    "    f = t(read.delim(file=paste(dir, org, 'isip', prot,'_hits.txt', sep=\"\"),\n",
    "                 sep = ' ',header = F,))\n",
    "    if (nrow(f) >=1){\n",
    "    f <- data.frame(\"orfs\"=f, \n",
    "                    \"ko_iteration\"=\"ko:_1\",\n",
    "                    \"ko_id\"=paste(\"isip_\", prot, sep=\"\"))}\n",
    "}\n",
    "\n",
    "i41 <- read.isip(\"04\", \"1\")\n",
    "i41a <- read.isip(\"04\", \"1a\")\n",
    "i41b <- read.isip(\"04\", \"1b\")\n",
    "i42a <- read.isip(\"04\", \"2a\")\n",
    "i43 <- read.isip(\"04\", \"3\")\n",
    "isip4 <- rbind(i41a, i41b, i42a, i43)\n",
    "isip4 <- distinct(isip4, orfs, .keep_all = T)\n",
    "\n",
    "i81 <- read.isip(\"08\", \"1\")\n",
    "i81a <- read.isip(\"08\", \"1a\")\n",
    "i81b <- read.isip(\"08\", \"1b\")\n",
    "i82a <- read.isip(\"08\", \"2a\")\n",
    "i83 <- read.isip(\"08\", \"3\")\n",
    "isip8 <- rbind( i81a, i81b, i82a, i83)  \n",
    "isip8 <- distinct(isip8, orfs, .keep_all = T)\n",
    "\n",
    "\n",
    "i61 <- read.isip(\"06\", \"1\")\n",
    "i61a <- read.isip(\"06\", \"1a\")\n",
    "i61b <- read.isip(\"06\", \"1b\")\n",
    "i62a <- read.isip(\"06\", \"2a\")\n",
    "i63 <- read.isip(\"06\", \"3\")\n",
    "isip6 <- rbind(i61a, i61b, i62a, i63)\n",
    "isip6 <- distinct(isip6, orfs, .keep_all = T)\n",
    "\n",
    "i132a <- read.isip(\"13\", \"2a\")\n",
    "i133 <- read.isip(\"13\", \"3\")\n",
    "isip13 <- rbind(i132a, i133)\n",
    "isip13 <- distinct(isip13, orfs, .keep_all = T)\n",
    "\n",
    "head(isip13)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebed5576",
   "metadata": {},
   "source": [
    "Combine ISIP proteins and ko dataframes, checking not to repeat if the orf was already annotated.\n",
    "\n",
    "1. Merge full ko definition list with ko's in organism using many-to-one merge relationship.  Organism's ko df have repeated ko_id, as the same annotation could be matched to multiple ORFs. So many rows in organisms's df may match to one row in full ko definition df. The column ko_id is named the same for both objects and will be used to merge the two dataframes. For this we use dplyr's left-join with relationship=many-to-one. \n",
    "\n",
    "2. Isip df and ko df have column query in common. We create a list of ORF's that the isip df and ko df have in common. isip rows are added to ko df for those ORFs not in common."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "28e2fde7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 14312\n",
      "[1] 14312\n",
      "[1] 12787\n",
      "[1] 12787\n",
      "[1] 35262\n",
      "[1] 35262\n",
      "[1] 30920\n",
      "[1] 30920\n"
     ]
    }
   ],
   "source": [
    "\n",
    "combine_save <- function(isip, ko_df, org){\n",
    "    \n",
    "    ko_fin <- bind_rows(ko_df, isip)\n",
    " \n",
    "    write.csv(ko_fin, paste(\"../expression_analysis/kegg_files/ko\", org,\"_ls.csv\", sep=\"\"), row.names=F)\n",
    "    print(nrow(ko_fin))\n",
    "    print(nrow(ko_df)+nrow(isip))\n",
    "}\n",
    "\n",
    "\n",
    "ko4_def <- combine_save(isip4, ko4, \"4\")\n",
    "ko8_def <- combine_save(isip8, ko8, \"8\")\n",
    "ko6_def <- combine_save(isip6, ko6, \"6\")\n",
    "ko13_def <- combine_save(isip13, ko13, \"13\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b79a095a-58e6-4911-99b6-8fa877d556b8",
   "metadata": {},
   "source": [
    "# 3. Create a list of unique ko's found across all organisms. \n",
    "---\n",
    "Some formating the the list of unique ko's is made so that bash scripting can be used. This list is later used to find the name and symbol of the ko_id's and map it back to the organism-specific tables made above.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "562e1fe0-6593-4267-98ce-475f45b74de6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a df of all unique ko's found for all annotations\n",
    "all.ko <- bind_rows(ko4, ko6,ko8,ko13) %>% select(ko_id)\n",
    "\n",
    "all.ko <- distinct(all.ko, ko_id, .keep_all = T)\n",
    "#so that bash will read my file correctly, I need to start with ko= \n",
    "#and surround all ko_id's by a single quote\n",
    "kk=c('ko=')\n",
    "q=c('\"')nano \n",
    "all.ko=rbind(kk, q, all.ko, q)\n",
    "\n",
    "#write to a .txt file with ko_id's separated by a space\n",
    "write.table(all.ko, \"../expression_analysis/kegg_files/all.ko.txt\", quote=F, sep=\"\", eol=\" \", row.names = FALSE, col.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69dfdc59-41a8-4bb3-baca-e15a97f815bf",
   "metadata": {},
   "source": [
    "## 3.1 Run bash script `koNames.sh` in script folder\n",
    "a total of 7218 rows, or ko_id's \n",
    "\n",
    "## 3.2 Clean up ko_def table\n",
    "---\n",
    "Seperate the Ko ID, symbol, and name into different columns, add the ISIP genes and save. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "df2c4ac5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"7191     =   all unique ko's found across organisms\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ko_id</th><th scope=col>symbol</th><th scope=col>name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>7186</th><td>K11287 </td><td> TSPYL4</td><td> TSPY-like 4                  </td></tr>\n",
       "\t<tr><th scope=row>7187</th><td>K11288 </td><td> TSPYL5</td><td> TSPY-like 5                  </td></tr>\n",
       "\t<tr><th scope=row>7188</th><td>K11289 </td><td> TSPYL6</td><td> TSPY-like 6                  </td></tr>\n",
       "\t<tr><th scope=row>7189</th><td>isip1a </td><td>isip_1a</td><td>Iron stress induced protein 1a</td></tr>\n",
       "\t<tr><th scope=row>7190</th><td>isip_2a</td><td>isip_2a</td><td>Iron stress induced protein 2a</td></tr>\n",
       "\t<tr><th scope=row>7191</th><td>isip_3 </td><td>isip_3 </td><td>Iron stress induced protein 3 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & ko\\_id & symbol & name\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t7186 & K11287  &  TSPYL4 &  TSPY-like 4                  \\\\\n",
       "\t7187 & K11288  &  TSPYL5 &  TSPY-like 5                  \\\\\n",
       "\t7188 & K11289  &  TSPYL6 &  TSPY-like 6                  \\\\\n",
       "\t7189 & isip1a  & isip\\_1a & Iron stress induced protein 1a\\\\\n",
       "\t7190 & isip\\_2a & isip\\_2a & Iron stress induced protein 2a\\\\\n",
       "\t7191 & isip\\_3  & isip\\_3  & Iron stress induced protein 3 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | ko_id &lt;chr&gt; | symbol &lt;chr&gt; | name &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 7186 | K11287  |  TSPYL4 |  TSPY-like 4                   |\n",
       "| 7187 | K11288  |  TSPYL5 |  TSPY-like 5                   |\n",
       "| 7188 | K11289  |  TSPYL6 |  TSPY-like 6                   |\n",
       "| 7189 | isip1a  | isip_1a | Iron stress induced protein 1a |\n",
       "| 7190 | isip_2a | isip_2a | Iron stress induced protein 2a |\n",
       "| 7191 | isip_3  | isip_3  | Iron stress induced protein 3  |\n",
       "\n"
      ],
      "text/plain": [
       "     ko_id   symbol  name                          \n",
       "7186 K11287   TSPYL4  TSPY-like 4                  \n",
       "7187 K11288   TSPYL5  TSPY-like 5                  \n",
       "7188 K11289   TSPYL6  TSPY-like 6                  \n",
       "7189 isip1a  isip_1a Iron stress induced protein 1a\n",
       "7190 isip_2a isip_2a Iron stress induced protein 2a\n",
       "7191 isip_3  isip_3  Iron stress induced protein 3 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#read in table with name and symbol matches to kegg ko's\n",
    "#make tidy\n",
    "ko_def <- read.delim(\"../expression_analysis/kegg_files/ko_pathways.txt\", head=F,sep = \";\")\n",
    "id <- str_extract(ko_def$V1,'K[[:digit:]]*')\n",
    "sym <- str_remove(ko_def$V1, 'ko:K[[:digit:]]*')\n",
    "ko_def <- data.frame(ko_id = id, symbol=sym, name=ko_def$V2)\n",
    "isip=data.frame(ko_id=c('isip_1a','isip_2a','isip_3'), \n",
    "                symbol=c('isip_1a','isip_2a','isip_3'),\n",
    "                name=c('Iron stress induced protein 1a',\n",
    "                       'Iron stress induced protein 2a', 'Iron stress induced protein 3'))\n",
    "ko_def=bind_rows(ko_def, isip)\n",
    "print(paste(nrow(ko_def), \"    =   all unique ko's found across organisms\"))\n",
    "write.csv(ko_def,'../expression_analysis/kegg_files/ko_def.csv', row.names=F)\n",
    "tail(ko_def)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd3ea4dd",
   "metadata": {},
   "source": [
    "# Using Kegg api for Kegg annotations \n",
    "To create a Kegg pathway key with ko_id, name, and symbol, for the pathways of interest, use the kegg api starting with the Kegg pathway codes:\n",
    "- Photosynthesis\n",
    "- Photosynthesis - antenna proteins\n",
    "- Carbon fixation in photosynthetic organisms\n",
    "- Nitrogen metabolism\n",
    "- Carotenoid biosynthesis\n",
    "Get the pathway name and all of the ko's within that path byloop through each path code and using Kegg api, extract this information. \n",
    "Get the kegg ko name and symbol for each pathway by looping through each ko_id.\n",
    "## After running each loop, we should have a dataframe with columns:\n",
    "- ko_id\n",
    "- name\n",
    "- symbol\n",
    "- pathway/module\n",
    "- sub_category\n",
    "- broad_category\n",
    "## Add in genes not on Kegg, or not associated with a pathway\n",
    "After adding the additional genes, clean up the table by adding appropriate pathway names and shortening gene names for the heatmap.\n",
    "## Extract all ko's from tables which appear in at least one sample\n",
    "We will use the resulting tables for subsetting data later in making heat maps. We can create heat maps based on pathway, module, or category. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "979b4944-1ce8-4727-bcc9-0209a3350c31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Path</th><th scope=col>ko_id</th><th scope=col>symbol</th><th scope=col>name</th><th scope=col>sub_category</th><th scope=col>path_name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>259</th><td>map00906</td><td>K22492</td><td>CYP175A  </td><td>beta-carotene 3-hydroxylase            </td><td>Carotenoid biosynthesis</td><td>Carotenoid biosynthesis</td></tr>\n",
       "\t<tr><th scope=row>260</th><td>map00906</td><td>K22502</td><td>crtY     </td><td>lycopene beta-cyclase                  </td><td>Carotenoid biosynthesis</td><td>Carotenoid biosynthesis</td></tr>\n",
       "\t<tr><th scope=row>261</th><td>map00906</td><td>K23037</td><td>CRTS, ASY</td><td>beta-carotene 4-ketolase/3-hydroxylase </td><td>Carotenoid biosynthesis</td><td>Carotenoid biosynthesis</td></tr>\n",
       "\t<tr><th scope=row>262</th><td>map00906</td><td>K25072</td><td>crtNc    </td><td>4,4'-diapolycopenoate synthase         </td><td>Carotenoid biosynthesis</td><td>Carotenoid biosynthesis</td></tr>\n",
       "\t<tr><th scope=row>263</th><td>map00906</td><td>K25073</td><td>cruO     </td><td>1'-hydroxy-gamma-carotene C-4' ketolase</td><td>Carotenoid biosynthesis</td><td>Carotenoid biosynthesis</td></tr>\n",
       "\t<tr><th scope=row>264</th><td>map00906</td><td>K25074</td><td>crtU     </td><td>carotenoid chi-ring synthase           </td><td>Carotenoid biosynthesis</td><td>Carotenoid biosynthesis</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & Path & ko\\_id & symbol & name & sub\\_category & path\\_name\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t259 & map00906 & K22492 & CYP175A   & beta-carotene 3-hydroxylase             & Carotenoid biosynthesis & Carotenoid biosynthesis\\\\\n",
       "\t260 & map00906 & K22502 & crtY      & lycopene beta-cyclase                   & Carotenoid biosynthesis & Carotenoid biosynthesis\\\\\n",
       "\t261 & map00906 & K23037 & CRTS, ASY & beta-carotene 4-ketolase/3-hydroxylase  & Carotenoid biosynthesis & Carotenoid biosynthesis\\\\\n",
       "\t262 & map00906 & K25072 & crtNc     & 4,4'-diapolycopenoate synthase          & Carotenoid biosynthesis & Carotenoid biosynthesis\\\\\n",
       "\t263 & map00906 & K25073 & cruO      & 1'-hydroxy-gamma-carotene C-4' ketolase & Carotenoid biosynthesis & Carotenoid biosynthesis\\\\\n",
       "\t264 & map00906 & K25074 & crtU      & carotenoid chi-ring synthase            & Carotenoid biosynthesis & Carotenoid biosynthesis\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | Path &lt;chr&gt; | ko_id &lt;chr&gt; | symbol &lt;chr&gt; | name &lt;chr&gt; | sub_category &lt;chr&gt; | path_name &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 259 | map00906 | K22492 | CYP175A   | beta-carotene 3-hydroxylase             | Carotenoid biosynthesis | Carotenoid biosynthesis |\n",
       "| 260 | map00906 | K22502 | crtY      | lycopene beta-cyclase                   | Carotenoid biosynthesis | Carotenoid biosynthesis |\n",
       "| 261 | map00906 | K23037 | CRTS, ASY | beta-carotene 4-ketolase/3-hydroxylase  | Carotenoid biosynthesis | Carotenoid biosynthesis |\n",
       "| 262 | map00906 | K25072 | crtNc     | 4,4'-diapolycopenoate synthase          | Carotenoid biosynthesis | Carotenoid biosynthesis |\n",
       "| 263 | map00906 | K25073 | cruO      | 1'-hydroxy-gamma-carotene C-4' ketolase | Carotenoid biosynthesis | Carotenoid biosynthesis |\n",
       "| 264 | map00906 | K25074 | crtU      | carotenoid chi-ring synthase            | Carotenoid biosynthesis | Carotenoid biosynthesis |\n",
       "\n"
      ],
      "text/plain": [
       "    Path     ko_id  symbol    name                                   \n",
       "259 map00906 K22492 CYP175A   beta-carotene 3-hydroxylase            \n",
       "260 map00906 K22502 crtY      lycopene beta-cyclase                  \n",
       "261 map00906 K23037 CRTS, ASY beta-carotene 4-ketolase/3-hydroxylase \n",
       "262 map00906 K25072 crtNc     4,4'-diapolycopenoate synthase         \n",
       "263 map00906 K25073 cruO      1'-hydroxy-gamma-carotene C-4' ketolase\n",
       "264 map00906 K25074 crtU      carotenoid chi-ring synthase           \n",
       "    sub_category            path_name              \n",
       "259 Carotenoid biosynthesis Carotenoid biosynthesis\n",
       "260 Carotenoid biosynthesis Carotenoid biosynthesis\n",
       "261 Carotenoid biosynthesis Carotenoid biosynthesis\n",
       "262 Carotenoid biosynthesis Carotenoid biosynthesis\n",
       "263 Carotenoid biosynthesis Carotenoid biosynthesis\n",
       "264 Carotenoid biosynthesis Carotenoid biosynthesis"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>264</li><li>6</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 264\n",
       "\\item 6\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 264\n",
       "2. 6\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 264   6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#3. pathways\n",
    "## get name for each pathway\n",
    "## add pathway name to dataframe\n",
    "## get list of all ko's for each pathway\n",
    "## get name and symbol for each ko\n",
    "heatMapPath = data.frame('Pathway'=c('map00195','map00196','map00710','map00910','map00906'), \n",
    "                         sub_category=c('Photosynthesis','Photosynthesis-antenna proteins','Carbon fixation in photosynthetic organisms',\n",
    "                                       'Nitrogen metabolism','Carotenoid biosynthesis'))\n",
    "\n",
    "url=c(\"https://rest.kegg.jp/\")\n",
    "find.path=list()\n",
    "for (i in heatMapPath$Pathway){\n",
    "    p=getURL(paste(url,'find/pathway/', i, sep=''))\n",
    "    find.path=unname(c(find.path,p))\n",
    "}\n",
    "\n",
    "path_name=data.frame('path_name'=str_remove(find.path,'path:map[[:digit:]]{5}\\t'))\n",
    "path_name$path_name=str_remove(path_name$path_name, '\\n')\n",
    "\n",
    "sub_path=bind_cols(heatMapPath, path_name)\n",
    "colnames(sub_path)[1]='Path'\n",
    "\n",
    "p=list()\n",
    "for (i in sub_path$Path){\n",
    "    link.ko=getURL(paste(url,'link/ko/', i, sep=''))\n",
    "    #link.ko=str_extract(p, 'ko:K[[:digit:]]+')\n",
    "    p=c(p,link.ko)\n",
    "}\n",
    "\n",
    "p=data.frame(ko_id=unlist(str_extract_all(p,'K[[:digit:]]+')), \n",
    "               path=(unlist(str_extract_all(p, 'map[[:digit:]]{5}')))) \n",
    "symbol=list()\n",
    "name=list()\n",
    "for (i in p$ko_id){\n",
    "    g=getURL(paste(url,'find/ko/', i, sep=''))\n",
    "    symbol=c(symbol,str_remove(str_extract(unname(g), '\\t.+;'), ';'))\n",
    "    name=c(name, str_remove(str_extract(unname(g),'; .*'), '; '))\n",
    "}\n",
    "path_ko=data.frame('Path'=p$path,\n",
    "                   'ko_id'=p$ko_id,\n",
    "                   'symbol'=str_remove(unlist(symbol),'\\t'),\n",
    "                   'name'= str_remove(unlist(name), '\\\\[EC:.*\\\\]'))\n",
    "\n",
    "sub_path=left_join(path_ko, sub_path, by='Path', relationship='many-to-many')\n",
    "\n",
    "tail(sub_path)\n",
    "dim(sub_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "6d1233c9-2e29-4d76-b4c0-f0ff9f360af4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(Path, ko_id, symbol, name, sub_category, path_name)`\n"
     ]
    }
   ],
   "source": [
    "## add flavodoxin and ferredoxin genes \n",
    "flavodoxin=data.frame(Path='map00195', ko_id=c('K03839','K03840','K21567','K00528'), \n",
    "                      symbol=c('fldA, nifF, isiB','fldB','fnr','fpr'), \n",
    "                      name=c('flavodoxin I','flavodoxin II','ferredoxin/flavodoxin---NADP+ reductase',\n",
    "                            'ferredoxin/flavodoxin---NADP+ reductase'), sub_category='Photosynthesis',\n",
    "                      path_name='Photosynthesis')\n",
    "sub_path=bind_rows(sub_path, flavodoxin)  \n",
    "\n",
    "## add urea cycle genes\n",
    "urea_cycle = filter(ko_def, ko_id%in%c('K00611','K01940','K01755','K01476'))\n",
    "urea_cycle = mutate(urea_cycle, 'sub_category'='Urea cycle', 'path_name'='Nitrogen metabolism', 'Path'='map00910')\n",
    "sub_path = full_join(sub_path, urea_cycle)\n",
    "\n",
    "## add ISIP genes\n",
    "sub_path = bind_rows(sub_path, isip)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d45d44-4ec5-4607-ba41-25d68c454d38",
   "metadata": {},
   "source": [
    "### Filter pathway table for genes found across organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "236ac865-2880-4caa-ad72-4bc39f8bc3d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "heat.path=(filter(sub_path, (ko_id %in% ko_def$ko_id)==T))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8dc412e-accc-4745-acd3-569f810b3326",
   "metadata": {},
   "source": [
    "### Correct any pathway names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "e20c6ea9-c068-4428-9bf1-62dc1ea5e833",
   "metadata": {},
   "outputs": [],
   "source": [
    "heat.path$path_name[heat.path$path_name=='Photosynthesis - antenna proteins']='Photosynthesis'\n",
    "\n",
    "heat.path[str_detect(heat.path$symbol, 'ATP.*F'), 'path_name'] = 'F-Type ATP-ase'\n",
    "\n",
    "heat.path[str_detect(heat.path$name, 'violaxanthin|zeaxanthin|isip_.'), 'path_name'] = 'Photosynthesis'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9405bed-b881-45b8-b9b1-223775e839ef",
   "metadata": {},
   "source": [
    "### Correct subcategories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "95c9ac8a-1e1e-4c73-950a-136add025d5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "heat.path$sub_category = str_replace_all(heat.path$symbol, \n",
    "                                        c(\"DUF\"=\"NADH dehydrogenase\", \"COX|cox|CYC\"=\"cytochrome c\", \"ATP.*V\"=\"V-Type ATP-ase\",\"ATP.*F\"=\"F-Type ATP-ase\", \n",
    "                                          'maeB|ppdK'='CAM light', 'MDH1|MDH2|mdh|ppc'='CAM dark','CPS1'='Urea cycle','psa'='PSI', 'psb'='PSII', \n",
    "                                          'LHCA'='LHCA', 'LHCB'='LHCB','isip'='ISIP'))\n",
    "\n",
    "heat.path$sub_category = str_replace_all(heat.path$name, \n",
    "                                         c('carbonic anhydrase'='Carbonic anhydrase', 'nitrate/nitrite transport'='Nitrogen transporters', \n",
    "                                           'glutamate|glutamine'='GS/GOGAT and GDH', 'nitrite reductase'='Nitrite reductase', \n",
    "                                           'nitrate reductase'='Nitrate reductase', 'violaxanthin|zeaxanthin'='Xanthophyll cycle'))\n",
    "\n",
    "heat.path[heat.path$ko_id %in% \n",
    "          c('K00855','K00927','K01100','K01601','K01623','K01624','K01783',\n",
    "            'K01803','K01807','K01808','K02446','K03841','K11532','K00134',\n",
    "            'K00615'), 'sub_category'] = 'Calvin cycle' \n",
    "\n",
    "heat.path$sub_category = str_replace(heat.path$sub_category, 'Photosynthesis', 'Electron transport chain')\n",
    "\n",
    "heat.path$sub_category = str_replace(heat.path$sub_category, 'Carbon fixation in photosynthetic organisms', 'C4 Dicarboxilic acid cycle')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a497095-85fb-40b6-93e1-7020a0ca00b6",
   "metadata": {},
   "source": [
    "### Correct gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "3dae01e7-953c-4123-af59-69cfdd846cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "heat.path[str_detect(heat.path$symbol, 'MDH1'), 'name'] = 'malate dehydrogenase 1'\n",
    "heat.path[str_detect(heat.path$symbol, 'MDH2'), 'name'] = 'malate dehydrogenase 2'\n",
    "\n",
    "heat.path$name= str_replace_all(heat.path$name, c('light-harvesting complex I '='LHCA ','light-harvesting complex II'='LHCB',  \n",
    "                                 'chlorophyll'='Chl','photosystem I '='PSI ', 'photosystem II'='PSII',\n",
    "                                 'F-type .* subunit '='F-Type ATP-ase ', 'isip_'='Iron starvation induced protein ',\n",
    "                                 'fructose-bisphosphate aldolase'='FBA', 'ribulose-bisphosphate carboxylase'='RuBisCO',\n",
    "                                               'MFS transporter, NNP family, '=''))\n",
    "\n",
    "heat.path$name = str_remove(heat.path$name, '.(phosphorylating).')\n",
    "heat.path$name = str_remove(heat.path$name, '\\\\[.*\\\\]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "a6b0b0a9-a82e-4c6d-b768-dc5c2a8270fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'F-type H+-transporting ATPase subunit a'</li><li>'F-type H+-transporting ATPase subunit b'</li><li>'F-type H+/Na+-transporting ATPase subunit alpha '</li><li>'F-type H+/Na+-transporting ATPase subunit beta '</li><li>'F-type H+-transporting ATPase subunit delta'</li><li>'F-type H+-transporting ATPase subunit epsilon'</li><li>'F-type H+-transporting ATPase subunit gamma'</li><li>'apocytochrome f'</li><li>'cytochrome b6-f complex iron-sulfur subunit '</li><li>'ferredoxin'</li><li>'ferredoxin--NADP+ reductase '</li><li>'photosystem I P700 chlorophyll a apoprotein A1 '</li><li>'photosystem I P700 chlorophyll a apoprotein A2 '</li><li>'photosystem I subunit II'</li><li>'photosystem I subunit IV'</li><li>'photosystem I subunit X'</li><li>'photosystem II P680 reaction center D1 protein '</li><li>'photosystem II CP47 chlorophyll apoprotein'</li><li>'photosystem II CP43 chlorophyll apoprotein'</li><li>'photosystem II P680 reaction center D2 protein '</li><li>'photosystem II PsbM protein'</li><li>'photosystem II oxygen-evolving enhancer protein 1'</li><li>'photosystem II oxygen-evolving enhancer protein 2'</li><li>'photosystem II PsbU protein'</li><li>'photosystem II oxygen-evolving enhancer protein 3'</li><li>'photosystem II Psb27 protein'</li><li>'photosystem II 13kDa protein'</li><li>'cytochrome c6'</li><li>'light-harvesting complex I chlorophyll a/b binding protein 1'</li><li>'light-harvesting complex I chlorophyll a/b binding protein 2'</li><li>'light-harvesting complex I chlorophyll a/b binding protein 4'</li><li>'light-harvesting complex I chlorophyll a/b binding protein 5'</li><li>'light-harvesting complex II chlorophyll a/b binding protein 3'</li><li>'light-harvesting complex II chlorophyll a/b binding protein 5'</li><li>'light-harvesting complex II chlorophyll a/b binding protein 6'</li><li>'light-harvesting complex II chlorophyll a/b binding protein 7'</li><li>'malate dehydrogenase '</li><li>'malate dehydrogenase (decarboxylating) '</li><li>'malate dehydrogenase (oxaloacetate-decarboxylating)(NADP+) '</li><li>'Calvin cycle'</li><li>'alanine transaminase '</li><li>'pyruvate, orthophosphate dikinase '</li><li>'phosphoenolpyruvate carboxylase '</li><li>'phosphoenolpyruvate carboxykinase (ATP) '</li><li>'aspartate aminotransferase, cytoplasmic '</li><li>'aspartate aminotransferase, mitochondrial '</li><li>'GS/GOGAT and GDH dehydrogenase (NAD(P)+) '</li><li>'GS/GOGAT and GDH dehydrogenase (NADP+) '</li><li>'GS/GOGAT and GDH synthase (NADH) '</li><li>'GS/GOGAT and GDH synthase (NADPH) large chain '</li><li>'GS/GOGAT and GDH synthase (NADPH) small chain '</li><li>'GS/GOGAT and GDH synthase (ferredoxin) '</li><li>'Nitrite reductase (NADH) large subunit '</li><li>'ferredoxin-Nitrite reductase '</li><li>'Nitrite reductase (NO-forming) '</li><li>'assimilatory Nitrate reductase catalytic subunit '</li><li>'nitronate monooxygenase '</li><li>'formamidase '</li><li>'nitrilase '</li><li>'Carbonic anhydrase '</li><li>'cyanate lyase '</li><li>'GS/GOGAT and GDH synthetase '</li><li>'carbamoyl-phosphate synthase (ammonia) '</li><li>'MFS transporter, NNP family, Nitrogen transporterser'</li><li>'Nitrate reductase (NAD(P)H) '</li><li>'GS/GOGAT and GDH dehydrogenase '</li><li>'Nitrogen transporters system substrate-binding protein'</li><li>'Nitrogen transporters system permease protein'</li><li>'Nitrogen transporters system ATP-binding protein '</li><li>'Nitrogen transporters system ATP-binding protein'</li><li>'fungal nitric oxide reductase '</li><li>'zeta-carotene desaturase '</li><li>'15-cis-phytoene synthase '</li><li>'15-cis-phytoene desaturase '</li><li>'lycopene beta-cyclase '</li><li>'lycopene epsilon-cyclase '</li><li>'Xanthophyll cycle epoxidase '</li><li>'Xanthophyll cycle de-epoxidase '</li><li>'9-cis-epoxycarotenoid dioxygenase '</li><li>'(+)-abscisic acid 8\\'-hydroxylase '</li><li>'phytoene desaturase '</li><li>'capsanthin/capsorubin synthase '</li><li>'zeta-carotene isomerase '</li><li>'phytoene desaturase (3,4-didehydrolycopene-forming) '</li><li>'15-cis-phytoene synthase / lycopene beta-cyclase '</li><li>'9-cis-beta-carotene 9\\',10\\'-cleaving dioxygenase '</li><li>'carlactone synthase / all-trans-10\\'-apo-beta-carotenal 13,14-cleaving dioxygenase '</li><li>'4,4\\'-diapolycopenoate synthase '</li><li>'flavodoxin I'</li><li>'ferredoxin/flavodoxin---NADP+ reductase'</li><li>' argininosuccinate lyase [EC:4.3.2.1]'</li><li>' argininosuccinate synthase [EC:6.3.4.5]'</li><li>' ornithine carbamoyltransferase [EC:2.1.3.3]'</li><li>' arginase [EC:3.5.3.1]'</li><li>'Iron stress induced protein 1a'</li><li>'Iron stress induced protein 2a'</li><li>'Iron stress induced protein 3'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'F-type H+-transporting ATPase subunit a'\n",
       "\\item 'F-type H+-transporting ATPase subunit b'\n",
       "\\item 'F-type H+/Na+-transporting ATPase subunit alpha '\n",
       "\\item 'F-type H+/Na+-transporting ATPase subunit beta '\n",
       "\\item 'F-type H+-transporting ATPase subunit delta'\n",
       "\\item 'F-type H+-transporting ATPase subunit epsilon'\n",
       "\\item 'F-type H+-transporting ATPase subunit gamma'\n",
       "\\item 'apocytochrome f'\n",
       "\\item 'cytochrome b6-f complex iron-sulfur subunit '\n",
       "\\item 'ferredoxin'\n",
       "\\item 'ferredoxin--NADP+ reductase '\n",
       "\\item 'photosystem I P700 chlorophyll a apoprotein A1 '\n",
       "\\item 'photosystem I P700 chlorophyll a apoprotein A2 '\n",
       "\\item 'photosystem I subunit II'\n",
       "\\item 'photosystem I subunit IV'\n",
       "\\item 'photosystem I subunit X'\n",
       "\\item 'photosystem II P680 reaction center D1 protein '\n",
       "\\item 'photosystem II CP47 chlorophyll apoprotein'\n",
       "\\item 'photosystem II CP43 chlorophyll apoprotein'\n",
       "\\item 'photosystem II P680 reaction center D2 protein '\n",
       "\\item 'photosystem II PsbM protein'\n",
       "\\item 'photosystem II oxygen-evolving enhancer protein 1'\n",
       "\\item 'photosystem II oxygen-evolving enhancer protein 2'\n",
       "\\item 'photosystem II PsbU protein'\n",
       "\\item 'photosystem II oxygen-evolving enhancer protein 3'\n",
       "\\item 'photosystem II Psb27 protein'\n",
       "\\item 'photosystem II 13kDa protein'\n",
       "\\item 'cytochrome c6'\n",
       "\\item 'light-harvesting complex I chlorophyll a/b binding protein 1'\n",
       "\\item 'light-harvesting complex I chlorophyll a/b binding protein 2'\n",
       "\\item 'light-harvesting complex I chlorophyll a/b binding protein 4'\n",
       "\\item 'light-harvesting complex I chlorophyll a/b binding protein 5'\n",
       "\\item 'light-harvesting complex II chlorophyll a/b binding protein 3'\n",
       "\\item 'light-harvesting complex II chlorophyll a/b binding protein 5'\n",
       "\\item 'light-harvesting complex II chlorophyll a/b binding protein 6'\n",
       "\\item 'light-harvesting complex II chlorophyll a/b binding protein 7'\n",
       "\\item 'malate dehydrogenase '\n",
       "\\item 'malate dehydrogenase (decarboxylating) '\n",
       "\\item 'malate dehydrogenase (oxaloacetate-decarboxylating)(NADP+) '\n",
       "\\item 'Calvin cycle'\n",
       "\\item 'alanine transaminase '\n",
       "\\item 'pyruvate, orthophosphate dikinase '\n",
       "\\item 'phosphoenolpyruvate carboxylase '\n",
       "\\item 'phosphoenolpyruvate carboxykinase (ATP) '\n",
       "\\item 'aspartate aminotransferase, cytoplasmic '\n",
       "\\item 'aspartate aminotransferase, mitochondrial '\n",
       "\\item 'GS/GOGAT and GDH dehydrogenase (NAD(P)+) '\n",
       "\\item 'GS/GOGAT and GDH dehydrogenase (NADP+) '\n",
       "\\item 'GS/GOGAT and GDH synthase (NADH) '\n",
       "\\item 'GS/GOGAT and GDH synthase (NADPH) large chain '\n",
       "\\item 'GS/GOGAT and GDH synthase (NADPH) small chain '\n",
       "\\item 'GS/GOGAT and GDH synthase (ferredoxin) '\n",
       "\\item 'Nitrite reductase (NADH) large subunit '\n",
       "\\item 'ferredoxin-Nitrite reductase '\n",
       "\\item 'Nitrite reductase (NO-forming) '\n",
       "\\item 'assimilatory Nitrate reductase catalytic subunit '\n",
       "\\item 'nitronate monooxygenase '\n",
       "\\item 'formamidase '\n",
       "\\item 'nitrilase '\n",
       "\\item 'Carbonic anhydrase '\n",
       "\\item 'cyanate lyase '\n",
       "\\item 'GS/GOGAT and GDH synthetase '\n",
       "\\item 'carbamoyl-phosphate synthase (ammonia) '\n",
       "\\item 'MFS transporter, NNP family, Nitrogen transporterser'\n",
       "\\item 'Nitrate reductase (NAD(P)H) '\n",
       "\\item 'GS/GOGAT and GDH dehydrogenase '\n",
       "\\item 'Nitrogen transporters system substrate-binding protein'\n",
       "\\item 'Nitrogen transporters system permease protein'\n",
       "\\item 'Nitrogen transporters system ATP-binding protein '\n",
       "\\item 'Nitrogen transporters system ATP-binding protein'\n",
       "\\item 'fungal nitric oxide reductase '\n",
       "\\item 'zeta-carotene desaturase '\n",
       "\\item '15-cis-phytoene synthase '\n",
       "\\item '15-cis-phytoene desaturase '\n",
       "\\item 'lycopene beta-cyclase '\n",
       "\\item 'lycopene epsilon-cyclase '\n",
       "\\item 'Xanthophyll cycle epoxidase '\n",
       "\\item 'Xanthophyll cycle de-epoxidase '\n",
       "\\item '9-cis-epoxycarotenoid dioxygenase '\n",
       "\\item '(+)-abscisic acid 8\\textbackslash{}'-hydroxylase '\n",
       "\\item 'phytoene desaturase '\n",
       "\\item 'capsanthin/capsorubin synthase '\n",
       "\\item 'zeta-carotene isomerase '\n",
       "\\item 'phytoene desaturase (3,4-didehydrolycopene-forming) '\n",
       "\\item '15-cis-phytoene synthase / lycopene beta-cyclase '\n",
       "\\item '9-cis-beta-carotene 9\\textbackslash{}',10\\textbackslash{}'-cleaving dioxygenase '\n",
       "\\item 'carlactone synthase / all-trans-10\\textbackslash{}'-apo-beta-carotenal 13,14-cleaving dioxygenase '\n",
       "\\item '4,4\\textbackslash{}'-diapolycopenoate synthase '\n",
       "\\item 'flavodoxin I'\n",
       "\\item 'ferredoxin/flavodoxin---NADP+ reductase'\n",
       "\\item ' argininosuccinate lyase {[}EC:4.3.2.1{]}'\n",
       "\\item ' argininosuccinate synthase {[}EC:6.3.4.5{]}'\n",
       "\\item ' ornithine carbamoyltransferase {[}EC:2.1.3.3{]}'\n",
       "\\item ' arginase {[}EC:3.5.3.1{]}'\n",
       "\\item 'Iron stress induced protein 1a'\n",
       "\\item 'Iron stress induced protein 2a'\n",
       "\\item 'Iron stress induced protein 3'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'F-type H+-transporting ATPase subunit a'\n",
       "2. 'F-type H+-transporting ATPase subunit b'\n",
       "3. 'F-type H+/Na+-transporting ATPase subunit alpha '\n",
       "4. 'F-type H+/Na+-transporting ATPase subunit beta '\n",
       "5. 'F-type H+-transporting ATPase subunit delta'\n",
       "6. 'F-type H+-transporting ATPase subunit epsilon'\n",
       "7. 'F-type H+-transporting ATPase subunit gamma'\n",
       "8. 'apocytochrome f'\n",
       "9. 'cytochrome b6-f complex iron-sulfur subunit '\n",
       "10. 'ferredoxin'\n",
       "11. 'ferredoxin--NADP+ reductase '\n",
       "12. 'photosystem I P700 chlorophyll a apoprotein A1 '\n",
       "13. 'photosystem I P700 chlorophyll a apoprotein A2 '\n",
       "14. 'photosystem I subunit II'\n",
       "15. 'photosystem I subunit IV'\n",
       "16. 'photosystem I subunit X'\n",
       "17. 'photosystem II P680 reaction center D1 protein '\n",
       "18. 'photosystem II CP47 chlorophyll apoprotein'\n",
       "19. 'photosystem II CP43 chlorophyll apoprotein'\n",
       "20. 'photosystem II P680 reaction center D2 protein '\n",
       "21. 'photosystem II PsbM protein'\n",
       "22. 'photosystem II oxygen-evolving enhancer protein 1'\n",
       "23. 'photosystem II oxygen-evolving enhancer protein 2'\n",
       "24. 'photosystem II PsbU protein'\n",
       "25. 'photosystem II oxygen-evolving enhancer protein 3'\n",
       "26. 'photosystem II Psb27 protein'\n",
       "27. 'photosystem II 13kDa protein'\n",
       "28. 'cytochrome c6'\n",
       "29. 'light-harvesting complex I chlorophyll a/b binding protein 1'\n",
       "30. 'light-harvesting complex I chlorophyll a/b binding protein 2'\n",
       "31. 'light-harvesting complex I chlorophyll a/b binding protein 4'\n",
       "32. 'light-harvesting complex I chlorophyll a/b binding protein 5'\n",
       "33. 'light-harvesting complex II chlorophyll a/b binding protein 3'\n",
       "34. 'light-harvesting complex II chlorophyll a/b binding protein 5'\n",
       "35. 'light-harvesting complex II chlorophyll a/b binding protein 6'\n",
       "36. 'light-harvesting complex II chlorophyll a/b binding protein 7'\n",
       "37. 'malate dehydrogenase '\n",
       "38. 'malate dehydrogenase (decarboxylating) '\n",
       "39. 'malate dehydrogenase (oxaloacetate-decarboxylating)(NADP+) '\n",
       "40. 'Calvin cycle'\n",
       "41. 'alanine transaminase '\n",
       "42. 'pyruvate, orthophosphate dikinase '\n",
       "43. 'phosphoenolpyruvate carboxylase '\n",
       "44. 'phosphoenolpyruvate carboxykinase (ATP) '\n",
       "45. 'aspartate aminotransferase, cytoplasmic '\n",
       "46. 'aspartate aminotransferase, mitochondrial '\n",
       "47. 'GS/GOGAT and GDH dehydrogenase (NAD(P)+) '\n",
       "48. 'GS/GOGAT and GDH dehydrogenase (NADP+) '\n",
       "49. 'GS/GOGAT and GDH synthase (NADH) '\n",
       "50. 'GS/GOGAT and GDH synthase (NADPH) large chain '\n",
       "51. 'GS/GOGAT and GDH synthase (NADPH) small chain '\n",
       "52. 'GS/GOGAT and GDH synthase (ferredoxin) '\n",
       "53. 'Nitrite reductase (NADH) large subunit '\n",
       "54. 'ferredoxin-Nitrite reductase '\n",
       "55. 'Nitrite reductase (NO-forming) '\n",
       "56. 'assimilatory Nitrate reductase catalytic subunit '\n",
       "57. 'nitronate monooxygenase '\n",
       "58. 'formamidase '\n",
       "59. 'nitrilase '\n",
       "60. 'Carbonic anhydrase '\n",
       "61. 'cyanate lyase '\n",
       "62. 'GS/GOGAT and GDH synthetase '\n",
       "63. 'carbamoyl-phosphate synthase (ammonia) '\n",
       "64. 'MFS transporter, NNP family, Nitrogen transporterser'\n",
       "65. 'Nitrate reductase (NAD(P)H) '\n",
       "66. 'GS/GOGAT and GDH dehydrogenase '\n",
       "67. 'Nitrogen transporters system substrate-binding protein'\n",
       "68. 'Nitrogen transporters system permease protein'\n",
       "69. 'Nitrogen transporters system ATP-binding protein '\n",
       "70. 'Nitrogen transporters system ATP-binding protein'\n",
       "71. 'fungal nitric oxide reductase '\n",
       "72. 'zeta-carotene desaturase '\n",
       "73. '15-cis-phytoene synthase '\n",
       "74. '15-cis-phytoene desaturase '\n",
       "75. 'lycopene beta-cyclase '\n",
       "76. 'lycopene epsilon-cyclase '\n",
       "77. 'Xanthophyll cycle epoxidase '\n",
       "78. 'Xanthophyll cycle de-epoxidase '\n",
       "79. '9-cis-epoxycarotenoid dioxygenase '\n",
       "80. '(+)-abscisic acid 8\\'-hydroxylase '\n",
       "81. 'phytoene desaturase '\n",
       "82. 'capsanthin/capsorubin synthase '\n",
       "83. 'zeta-carotene isomerase '\n",
       "84. 'phytoene desaturase (3,4-didehydrolycopene-forming) '\n",
       "85. '15-cis-phytoene synthase / lycopene beta-cyclase '\n",
       "86. '9-cis-beta-carotene 9\\',10\\'-cleaving dioxygenase '\n",
       "87. 'carlactone synthase / all-trans-10\\'-apo-beta-carotenal 13,14-cleaving dioxygenase '\n",
       "88. '4,4\\'-diapolycopenoate synthase '\n",
       "89. 'flavodoxin I'\n",
       "90. 'ferredoxin/flavodoxin---NADP+ reductase'\n",
       "91. ' argininosuccinate lyase [EC:4.3.2.1]'\n",
       "92. ' argininosuccinate synthase [EC:6.3.4.5]'\n",
       "93. ' ornithine carbamoyltransferase [EC:2.1.3.3]'\n",
       "94. ' arginase [EC:3.5.3.1]'\n",
       "95. 'Iron stress induced protein 1a'\n",
       "96. 'Iron stress induced protein 2a'\n",
       "97. 'Iron stress induced protein 3'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"F-type H+-transporting ATPase subunit a\"                                           \n",
       " [2] \"F-type H+-transporting ATPase subunit b\"                                           \n",
       " [3] \"F-type H+/Na+-transporting ATPase subunit alpha \"                                  \n",
       " [4] \"F-type H+/Na+-transporting ATPase subunit beta \"                                   \n",
       " [5] \"F-type H+-transporting ATPase subunit delta\"                                       \n",
       " [6] \"F-type H+-transporting ATPase subunit epsilon\"                                     \n",
       " [7] \"F-type H+-transporting ATPase subunit gamma\"                                       \n",
       " [8] \"apocytochrome f\"                                                                   \n",
       " [9] \"cytochrome b6-f complex iron-sulfur subunit \"                                      \n",
       "[10] \"ferredoxin\"                                                                        \n",
       "[11] \"ferredoxin--NADP+ reductase \"                                                      \n",
       "[12] \"photosystem I P700 chlorophyll a apoprotein A1 \"                                   \n",
       "[13] \"photosystem I P700 chlorophyll a apoprotein A2 \"                                   \n",
       "[14] \"photosystem I subunit II\"                                                          \n",
       "[15] \"photosystem I subunit IV\"                                                          \n",
       "[16] \"photosystem I subunit X\"                                                           \n",
       "[17] \"photosystem II P680 reaction center D1 protein \"                                   \n",
       "[18] \"photosystem II CP47 chlorophyll apoprotein\"                                        \n",
       "[19] \"photosystem II CP43 chlorophyll apoprotein\"                                        \n",
       "[20] \"photosystem II P680 reaction center D2 protein \"                                   \n",
       "[21] \"photosystem II PsbM protein\"                                                       \n",
       "[22] \"photosystem II oxygen-evolving enhancer protein 1\"                                 \n",
       "[23] \"photosystem II oxygen-evolving enhancer protein 2\"                                 \n",
       "[24] \"photosystem II PsbU protein\"                                                       \n",
       "[25] \"photosystem II oxygen-evolving enhancer protein 3\"                                 \n",
       "[26] \"photosystem II Psb27 protein\"                                                      \n",
       "[27] \"photosystem II 13kDa protein\"                                                      \n",
       "[28] \"cytochrome c6\"                                                                     \n",
       "[29] \"light-harvesting complex I chlorophyll a/b binding protein 1\"                      \n",
       "[30] \"light-harvesting complex I chlorophyll a/b binding protein 2\"                      \n",
       "[31] \"light-harvesting complex I chlorophyll a/b binding protein 4\"                      \n",
       "[32] \"light-harvesting complex I chlorophyll a/b binding protein 5\"                      \n",
       "[33] \"light-harvesting complex II chlorophyll a/b binding protein 3\"                     \n",
       "[34] \"light-harvesting complex II chlorophyll a/b binding protein 5\"                     \n",
       "[35] \"light-harvesting complex II chlorophyll a/b binding protein 6\"                     \n",
       "[36] \"light-harvesting complex II chlorophyll a/b binding protein 7\"                     \n",
       "[37] \"malate dehydrogenase \"                                                             \n",
       "[38] \"malate dehydrogenase (decarboxylating) \"                                           \n",
       "[39] \"malate dehydrogenase (oxaloacetate-decarboxylating)(NADP+) \"                       \n",
       "[40] \"Calvin cycle\"                                                                      \n",
       "[41] \"alanine transaminase \"                                                             \n",
       "[42] \"pyruvate, orthophosphate dikinase \"                                                \n",
       "[43] \"phosphoenolpyruvate carboxylase \"                                                  \n",
       "[44] \"phosphoenolpyruvate carboxykinase (ATP) \"                                          \n",
       "[45] \"aspartate aminotransferase, cytoplasmic \"                                          \n",
       "[46] \"aspartate aminotransferase, mitochondrial \"                                        \n",
       "[47] \"GS/GOGAT and GDH dehydrogenase (NAD(P)+) \"                                         \n",
       "[48] \"GS/GOGAT and GDH dehydrogenase (NADP+) \"                                           \n",
       "[49] \"GS/GOGAT and GDH synthase (NADH) \"                                                 \n",
       "[50] \"GS/GOGAT and GDH synthase (NADPH) large chain \"                                    \n",
       "[51] \"GS/GOGAT and GDH synthase (NADPH) small chain \"                                    \n",
       "[52] \"GS/GOGAT and GDH synthase (ferredoxin) \"                                           \n",
       "[53] \"Nitrite reductase (NADH) large subunit \"                                           \n",
       "[54] \"ferredoxin-Nitrite reductase \"                                                     \n",
       "[55] \"Nitrite reductase (NO-forming) \"                                                   \n",
       "[56] \"assimilatory Nitrate reductase catalytic subunit \"                                 \n",
       "[57] \"nitronate monooxygenase \"                                                          \n",
       "[58] \"formamidase \"                                                                      \n",
       "[59] \"nitrilase \"                                                                        \n",
       "[60] \"Carbonic anhydrase \"                                                               \n",
       "[61] \"cyanate lyase \"                                                                    \n",
       "[62] \"GS/GOGAT and GDH synthetase \"                                                      \n",
       "[63] \"carbamoyl-phosphate synthase (ammonia) \"                                           \n",
       "[64] \"MFS transporter, NNP family, Nitrogen transporterser\"                              \n",
       "[65] \"Nitrate reductase (NAD(P)H) \"                                                      \n",
       "[66] \"GS/GOGAT and GDH dehydrogenase \"                                                   \n",
       "[67] \"Nitrogen transporters system substrate-binding protein\"                            \n",
       "[68] \"Nitrogen transporters system permease protein\"                                     \n",
       "[69] \"Nitrogen transporters system ATP-binding protein \"                                 \n",
       "[70] \"Nitrogen transporters system ATP-binding protein\"                                  \n",
       "[71] \"fungal nitric oxide reductase \"                                                    \n",
       "[72] \"zeta-carotene desaturase \"                                                         \n",
       "[73] \"15-cis-phytoene synthase \"                                                         \n",
       "[74] \"15-cis-phytoene desaturase \"                                                       \n",
       "[75] \"lycopene beta-cyclase \"                                                            \n",
       "[76] \"lycopene epsilon-cyclase \"                                                         \n",
       "[77] \"Xanthophyll cycle epoxidase \"                                                      \n",
       "[78] \"Xanthophyll cycle de-epoxidase \"                                                   \n",
       "[79] \"9-cis-epoxycarotenoid dioxygenase \"                                                \n",
       "[80] \"(+)-abscisic acid 8'-hydroxylase \"                                                 \n",
       "[81] \"phytoene desaturase \"                                                              \n",
       "[82] \"capsanthin/capsorubin synthase \"                                                   \n",
       "[83] \"zeta-carotene isomerase \"                                                          \n",
       "[84] \"phytoene desaturase (3,4-didehydrolycopene-forming) \"                              \n",
       "[85] \"15-cis-phytoene synthase / lycopene beta-cyclase \"                                 \n",
       "[86] \"9-cis-beta-carotene 9',10'-cleaving dioxygenase \"                                  \n",
       "[87] \"carlactone synthase / all-trans-10'-apo-beta-carotenal 13,14-cleaving dioxygenase \"\n",
       "[88] \"4,4'-diapolycopenoate synthase \"                                                   \n",
       "[89] \"flavodoxin I\"                                                                      \n",
       "[90] \"ferredoxin/flavodoxin---NADP+ reductase\"                                           \n",
       "[91] \" argininosuccinate lyase [EC:4.3.2.1]\"                                             \n",
       "[92] \" argininosuccinate synthase [EC:6.3.4.5]\"                                          \n",
       "[93] \" ornithine carbamoyltransferase [EC:2.1.3.3]\"                                      \n",
       "[94] \" arginase [EC:3.5.3.1]\"                                                            \n",
       "[95] \"Iron stress induced protein 1a\"                                                    \n",
       "[96] \"Iron stress induced protein 2a\"                                                    \n",
       "[97] \"Iron stress induced protein 3\"                                                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(heat.path$sub_category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9f64a83-9399-41ca-8c97-9b2911b9dde9",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique(heat.path$path_name)\n",
    "write.csv(heat.path, '../kegg_names/pathwaysHeatMap.csv', row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4085b63a-58e3-4ccb-ae0c-776ed0bad228",
   "metadata": {},
   "outputs": [],
   "source": [
    "all(heat.path$ko_id %in% ko_def$ko_id)\n",
    "dim(heat.path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
